<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Preamble | Omics Data Analysis" />
<meta property="og:type" content="book" />

<meta property="og:description" content="Course material for the Omics Data Analysis (WSBIM2122) course at UCLouvain." />


<meta name="author" content="Laurent Gatto" />

<meta name="date" content="2024-11-29" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="Course material for the Omics Data Analysis (WSBIM2122) course at UCLouvain.">

<title>Preamble | Omics Data Analysis</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>
<script>
function copy_link(id) {
  var dummy = document.createElement('input'),
  text = window.location.href.split(/[?#]/)[0] + '#' + id;
  document.body.appendChild(dummy);
  dummy.value = text;
  dummy.select();
  document.execCommand('copy');
  document.body.removeChild(dummy);
  
  var tooltip = document.getElementById(id + '-tooltip');
  tooltip.innerHTML = 'Copied!';
}

function reset_tooltip(id) {
  var tooltip = document.getElementById(id);
  tooltip.innerHTML = 'Copy link';
}
</script>
<link href="libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="libs/datatables-binding-0.33/datatables.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>
<script src="libs/plotly-binding-4.10.4/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-2.11.1/plotly-latest.min.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }

code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">Omics Data Analysis<p><p class="author">Laurent Gatto</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a id="active-page" href="index.html" id="toc-preamble">Preamble</a><ul class="toc-sections">
<li class="toc"><a href="#pre-requisites">Pre-requisites</a></li>
<li class="toc"><a href="#about-this-course-material">About this course material</a></li>
<li class="toc"><a href="#citation">Citation</a></li>
<li class="toc"><a href="#license">License</a></li>
<li class="toc"><a href="#setup">Setup</a></li>
</ul>
<a href="sec-rmd.html" id="toc-sec-rmd"><span class="toc-section-number">1</span> R markdown</a>
<a href="sec-linmod.html" id="toc-sec-linmod"><span class="toc-section-number">2</span> Linear models</a>
<a href="sec-hts.html" id="toc-sec-hts"><span class="toc-section-number">3</span> High throughput sequencing</a>
<a href="sec-rnaseq.html" id="toc-sec-rnaseq"><span class="toc-section-number">4</span> Differential expression analysis</a>
<a href="sec-design.html" id="toc-sec-design"><span class="toc-section-number">5</span> More complex experimental design</a>
<a href="sec-gsea.html" id="toc-sec-gsea"><span class="toc-section-number">6</span> Enrichment analyses</a>
<a href="sec-ms.html" id="toc-sec-ms"><span class="toc-section-number">7</span> Mass spectrometry</a>
<a href="sec-prot.html" id="toc-sec-prot"><span class="toc-section-number">8</span> Quantitative proteomics data analysis</a>
<a href="sec-ccl.html" id="toc-sec-ccl"><span class="toc-section-number">9</span> Conclusions</a>
<a href="sec-cli.html" id="toc-sec-cli"><span class="toc-section-number">10</span> Annex: the shell</a>
<a href="sec-git.html" id="toc-sec-git"><span class="toc-section-number">11</span> Annex: Collaborating with Git and GitHub</a>
<a href="sec-info.html" id="toc-sec-info"><span class="toc-section-number">12</span> Course and session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="preamble" class="section level1 unnumbered">
<h1>Preamble</h1>
<p>The <a href="https://uclouvain.be/cours-2021-wsbim2122.html">WSBIM2122</a> course
is a project-based course that teaches the analysis of quantitative
omics data. It is composed of three major parts:</p>
<ul>
<li><p>Introductory material about (advanced) usage of R markdown for
reports and slides, linear models, and utilisation and
interpretation of gene set enrichment and over-representation
analyses (GSEA).</p></li>
<li><p>Project 1: analysis and interpretation of RNA-Seq data, including
high throughput sequencing (HTS) data processing.</p></li>
<li><p>Project 2: analysis and interpretation of quantitative proteomics,
including mass spectrometry data (MS) and its processing.</p></li>
</ul>
<p>Each project will entail an introductory lecture, dedicated office
time for students to ask questions, oral presentations of the data
analysis and interpretation, and submission of a fully reproducible R
markdown report. The students will be able to amend their final
reports after the presentations, benefiting from in-class feedback and
discussion. The report will contain an introduction, a description of
the experimental design of the data, a detailed description of the
data analysis steps, discussion and interpretation of the results,
including a biological interpretation, as well as references and
contribution statement of all authors.</p>
<p>The projects will be done in groups, with different groups for each
project. The number of students per group will be set once the total
number of students is known.</p>
<p>Tentative schedule for the course.</p>
<ul>
<li>Short presentations: shell (see <a href="sec-cli.html#sec-cli">10</a>) and Rmd (chapter
<a href="sec-rmd.html#sec-rmd">1</a>).</li>
<li>Linear models (chapter <a href="sec-linmod.html#sec-linmod">2</a>).</li>
<li>Project 1: lectures on HTS (chapter <a href="sec-hts.html#sec-hts">3</a>) and RNA-Seq
(chapter <a href="sec-rnaseq.html#sec-rnaseq">4</a>).</li>
<li>Project 1 (cont’d): Gene set enrichment (chapter <a href="sec-gsea.html#sec-gsea">6</a>).</li>
<li>QA session for project 1</li>
<li>Project 1 presentations</li>
<li>Submission of the 1st report</li>
<li>Project 2: lectures on mass spectrometry (chapter <a href="sec-ms.html#sec-ms">7</a>) and
quantitative proteomics (chapter <a href="sec-prot.html#sec-prot">8</a>).</li>
<li>QA session for project 2</li>
<li>Project 2 presentations</li>
<li>Submission of the 2nd report</li>
<li>Oral exams based on the submitted reports.</li>
</ul>
<p>The course will be taught in English. The reports and oral exams can
be either in English or in French. The final mark will be based on the
submitted reports and the oral exam, tentatively about 50% each.</p>
<div id="pre-requisites" class="section level2 unnumbered">
<h2>Pre-requisites<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('pre-requisites')" onmouseout="reset_tooltip('pre-requisites-tooltip')"><span class="tooltiptext" id="pre-requisites-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Students taking this course should be comfortable with data analysis
and visualisation in R. Formal pre-requisites for students taking the
class are the <a href="https://UCLouvain-CBIO.github.io/WSBIM1207">WSBIM1207</a>
and <a href="https://UCLouvain-CBIO.github.io/WSBIM1322">WSBIM1322</a>.</p>
<p>Software requirements are documented in the <em>Setup</em> section below.</p>
</div>
<div id="about-this-course-material" class="section level2 unnumbered">
<h2>About this course material<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('about-this-course-material')" onmouseout="reset_tooltip('about-this-course-material-tooltip')"><span class="tooltiptext" id="about-this-course-material-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>This material is written in R markdown <span class="citation">(<label for="tufte-mn-1" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-1" class="margin-toggle">Allaire et al. 2024<span class="marginnote">Allaire, JJ, Yihui Xie, Christophe Dervieux, Jonathan McPherson, Javier Luraschi, Kevin Ushey, Aron Atkins, et al. 2024. <em>Rmarkdown: Dynamic Documents for r</em>. <a href="https://github.com/rstudio/rmarkdown">https://github.com/rstudio/rmarkdown</a>.</span>)</span> and compiled as a
book using <code>knitr</code> <span class="citation">(<label for="tufte-mn-2" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-2" class="margin-toggle">Xie 2024b<span class="marginnote">Xie, Yihui. 2024b. <em>Knitr: A General-Purpose Package for Dynamic Report Generation in r</em>. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.</span>)</span> <code>bookdown</code> <span class="citation">(<label for="tufte-mn-3" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-3" class="margin-toggle">Xie 2024a<span class="marginnote">Xie, Yihui. 2024a. <em>Bookdown: Authoring Books and Technical Documents with r Markdown</em>. <a href="https://github.com/rstudio/bookdown">https://github.com/rstudio/bookdown</a>.</span>)</span>. The source
code is publicly available in a Github repository
<a href="https://github.com/uclouvain-cbio/WSBIM2122">https://github.com/uclouvain-cbio/WSBIM2122</a>
and the compiled material can be read at <a href="http://bit.ly/WSBIM2122" class="uri">http://bit.ly/WSBIM2122</a>.</p>
<p>Contributions to this material are welcome. The best way to contribute
or contact the maintainers is by means of pull requests and
<a href="https://github.com/uclouvain-cbio/WSBIM2122/issues">issues</a>. Please
familiarise yourself with the <a href="https://github.com/UCLouvain-CBIO/WSBIM2122/blob/master/CONDUCT.md">code of
conduct</a>. By
participating in this project you agree to abide by its terms.</p>
</div>
<div id="citation" class="section level2 unnumbered">
<h2>Citation<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('citation')" onmouseout="reset_tooltip('citation-tooltip')"><span class="tooltiptext" id="citation-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>If you use this course, please cite it as</p>
<blockquote>
<p>Laurent Gatto and Axelle Loriot. <em>UCLouvain-CBIO/WSBIM2122: Omics
data analysis.</em> <a href="https://github.com/UCLouvain-CBIO/WSBIM2122" class="uri">https://github.com/UCLouvain-CBIO/WSBIM2122</a>.</p>
</blockquote>
</div>
<div id="license" class="section level2 unnumbered">
<h2>License<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('license')" onmouseout="reset_tooltip('license-tooltip')"><span class="tooltiptext" id="license-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>This material is licensed under the <a href="https://creativecommons.org/licenses/by-sa/4.0/">Creative Commons
Attribution-ShareAlike 4.0
License</a>.</p>
</div>
<div id="setup" class="section level2 unnumbered">
<h2>Setup<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('setup')" onmouseout="reset_tooltip('setup-tooltip')"><span class="tooltiptext" id="setup-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>We will be using the <a href="https://www.r-project.org/">R environment for statistical
computing</a> as main data science language.
We will also use the
<a href="https://www.rstudio.com/products/RStudio/">RStudio</a> interface to
interact with R and write scripts and reports. Both R and RStudio are
easy to install and works on all major operating systems.</p>
<p>Once R and RStudio are installed, a set of packages will need to be
installed. See section <a href="sec-info.html#sec-setup2">12.4</a> for details.</p>
<p>The <code>rWSBIM2122</code> package provides some pre-formatted data used in this
course. It can be installed with</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="index.html#cb1-1" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"UCLouvain-CBIO/rWSBIM2122"</span>)</span></code></pre></div>
<p>and then loaded with</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="index.html#cb2-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"rWSBIM2122"</span>)</span></code></pre></div>

</div>
</div>
<p style="text-align: center;">
<a href="sec-rmd.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2024-11-29
 using 
R version 4.4.1 (2024-06-14)
</p>
</div>
</div>



</body>
</html>
