<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 4 Differential expression analysis | Omics Data Analysis" />
<meta property="og:type" content="book" />

<meta property="og:description" content="Course material for the Omics Data Analysis (WSBIM2122) course at UCLouvain." />


<meta name="author" content="Laurent Gatto" />

<meta name="date" content="2024-11-01" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="Course material for the Omics Data Analysis (WSBIM2122) course at UCLouvain.">

<title>Chapter 4 Differential expression analysis | Omics Data Analysis</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>
<script>
function copy_link(id) {
  var dummy = document.createElement('input'),
  text = window.location.href.split(/[?#]/)[0] + '#' + id;
  document.body.appendChild(dummy);
  dummy.value = text;
  dummy.select();
  document.execCommand('copy');
  document.body.removeChild(dummy);
  
  var tooltip = document.getElementById(id + '-tooltip');
  tooltip.innerHTML = 'Copied!';
}

function reset_tooltip(id) {
  var tooltip = document.getElementById(id);
  tooltip.innerHTML = 'Copy link';
}
</script>
<link href="libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="libs/datatables-binding-0.33/datatables.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>
<script src="libs/plotly-binding-4.10.4/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-2.11.1/plotly-latest.min.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }

code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">Omics Data Analysis<p><p class="author">Laurent Gatto</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html" id="toc-preamble">Preamble</a>
<a href="sec-rmd.html" id="toc-sec-rmd"><span class="toc-section-number">1</span> R markdown</a>
<a href="sec-linmod.html" id="toc-sec-linmod"><span class="toc-section-number">2</span> Linear models</a>
<a href="sec-hts.html" id="toc-sec-hts"><span class="toc-section-number">3</span> High throughput sequencing</a>
<a id="active-page" href="sec-rnaseq.html" id="toc-sec-rnaseq"><span class="toc-section-number">4</span> Differential expression analysis</a><ul class="toc-sections">
<li class="toc"><a href="#theory-behind-deseq2"> Theory behind DESeq2</a></li>
<li class="toc"><a href="#running-deseq2"> Running DESeq2</a></li>
</ul>
<a href="sec-design.html" id="toc-sec-design"><span class="toc-section-number">5</span> More complex experimental design</a>
<a href="sec-gsea.html" id="toc-sec-gsea"><span class="toc-section-number">6</span> Enrichment analyses</a>
<a href="sec-ms.html" id="toc-sec-ms"><span class="toc-section-number">7</span> Mass spectrometry</a>
<a href="sec-prot.html" id="toc-sec-prot"><span class="toc-section-number">8</span> Quantitative proteomics data analysis</a>
<a href="sec-ccl.html" id="toc-sec-ccl"><span class="toc-section-number">9</span> Conclusions</a>
<a href="sec-cli.html" id="toc-sec-cli"><span class="toc-section-number">10</span> Annex: the shell</a>
<a href="sec-git.html" id="toc-sec-git"><span class="toc-section-number">11</span> Annex: Collaborating with Git and GitHub</a>
<a href="sec-info.html" id="toc-sec-info"><span class="toc-section-number">12</span> Course and session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="sec-rnaseq" class="section level1" number="4">
<h1>
<span class="header-section-number">Chapter 4</span> Differential expression analysis</h1>
<p><strong>Learning Objectives</strong></p>
<p>The goal of this chapter is</p>
<ul>
<li>to understand the different theorical concepts behind a differential expression analysis</li>
<li>to provide a real-life example of DE analysis analysis running DESeq2 step-by-step</li>
</ul>
<div id="theory-behind-deseq2" class="section level2" number="4.1">
<h2>
<span class="header-section-number">4.1</span> Theory behind DESeq2<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('theory-behind-deseq2')" onmouseout="reset_tooltip('theory-behind-deseq2-tooltip')"><span class="tooltiptext" id="theory-behind-deseq2-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>A large number of computational methods have been developed for differential expression analysis
<span class="citation">(<label for="tufte-mn-7" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-7" class="margin-toggle">Seyednasrollah, Laiho, and Elo 2015<span class="marginnote">Seyednasrollah, F, A Laiho, and LA Elo. 2015. <span>“Comparison of Software Packages for Detecting Differential Expression in RNA-Seq Studies.”</span> <em>Brief Bioinform</em> 1 (16): 59. <a href="https://doi.org/10.1093/bib/bbt086">https://doi.org/10.1093/bib/bbt086</a>.</span>)</span>, differing slightly in their methodology. Here we will present DESeq2,
a widely used bioconductor package dedicated to this type of analysis. For more information read
the original paper <span class="citation">(<label for="tufte-mn-8" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-8" class="margin-toggle">Love, Huber, and Anders 2014<span class="marginnote">Love, M, W Huber, and S Anders. 2014. <span>“Moderated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.”</span> <em>Genome Biology</em> 15 (5): 550–58. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.</span>)</span> and the
<a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">DESeq2 vignette</a>.</p>
<p>The starting point of the analysis is a count matrix, and the goal is to identify
genes that are differentially expressed between samples.</p>
<p>The different steps of the analysis are illustrated in the figure below. Briefly,
DESeq2 starts by estimating scaling factors. Then, it estimates the gene-wise
dispersions and shrinks these estimates to generate more accurate estimates of dispersion
to model the counts. Finally, DESeq2 fits a generalized linear model, performs hypothesis
testing and generates a list of differentially expressed genes.</p>
<p><img src="figs/deseq2_steps.png" width="60%" style="display: block; margin: auto;"></p>
<div id="size-factor-estimation" class="section level3" number="4.1.1">
<h3>
<span class="header-section-number">4.1.1</span> Size factor estimation<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('size-factor-estimation')" onmouseout="reset_tooltip('size-factor-estimation-tooltip')"><span class="tooltiptext" id="size-factor-estimation-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>DESeq2 expects as an input a matrix of raw counts (un-normalised counts).
These counts are supposed to reflect gene abundance (what we are interested in),
but they are also dependent on other less interesting factors such as gene length,
sequencing biases, sequencing depth or library composition. DESeq2 will estimate
scaling factors that will be used internally to account for the “uninteresting”
factors rendering the expression levels more comparable between samples.</p>
<ul>
<li><strong>Gene length</strong></li>
</ul>
<p>As illustrated in the example below, gene 1 and gene 2 have similar levels of expression,
but many more reads map to gene 2 than to gene 1. This might not be related to biology
but it could just reflect the fact that gene 2 is longer. Taking into account gene lengths
in the normalisation procedure is important when the purpose is to compare gene expression
levels within the same sample. However, for differential expression analysis, as gene
expression levels are compared between samples, it is not necessary to use gene lengths to
estimate the scaling factors (it is even not recommended). It was just mentioned here for
information because many RNAseq common normalisation methods such as TPM (transcripts per
million), FPKM (fragments per million), or RPKM (reads per million) take into account gene
lengths.</p>
<p><img src="figs/read_length.png" width="100%" style="display: block; margin: auto;"></p>
<ul>
<li><strong>Sequencing depth</strong></li>
</ul>
<p>Each sequencing experiment will produce a certain number of reads
expected to be typically around tens of millions. A fraction of the
raw sequencing reads will be discarded during the quality control, the
alignment and the counting processes, which implies that the total
number of reads for each sample will be different.</p>
<p>As shown in the following example, all genes seem to be expressed at
higher levels in sample 1 than in sample 2, but this is likely because
sample 1 has twice more reads than sample 2. Accounting for sequencing
depth is necessary for differential expression analysis as samples are
compared with each other.</p>
<p><img src="figs/Sequencing_depth.png" width="100%" style="display: block; margin: auto;"></p>
<ul>
<li><strong>Library composition</strong></li>
</ul>
<p>The library composition might also be different across samples. To
illustrate this, let’s imagine a basic cell expressing only 2 genes
(genes 1 and 2) and assume that a drug treatment induces a strong
expression of gene 3. If the normalisation was done using total number
of reads only, then the counts of gene 1 would be divided by 15 in
control cells, while it would be divided by 165 in treated cells. This
would lead to the misleading conclusion that the treatment has reduced
11 times the expression of gene 1. In this case, the library
composition has changed but not the expression level of gene 1.</p>
<p><img src="figs/library_composition.png" width="100%" style="display: block; margin: auto;"></p>
<p>In a real dataset, a few highly differentially expressed genes,
differences in the number of genes expressed between samples, or
presence of contaminations can skew library composition. Accounting
for it is highly recommended for accurate comparison of expression
between samples <span class="citation">(<label for="tufte-mn-9" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-9" class="margin-toggle">Anders and Huber 2010<span class="marginnote">Anders, S, and W Huber. 2010. <span>“Differential Expression Analysis for Sequence Count Data.”</span> <em>Genome Biology</em> 11 (27): 10. <a href="https://doi.org/10.1186/gb-2010-11-10-r106">https://doi.org/10.1186/gb-2010-11-10-r106</a>.</span>)</span>.</p>
<ul>
<li><strong>DESeq2 normalisation method</strong></li>
</ul>
<p>DESeq2 will estimate size factors in a way that takes into account both
library size and library composition, using the median of ratios method:</p>
<p><img src="figs/SF_formula.png" width="60%" style="display: block; margin: auto;"></p>
<p>Let’s try to understand what is behind this formula.</p>
<p><strong>Step 1</strong>: DESeq2 creates a pseudo-reference sample by calculating a
row-wise geometric mean (for each gene). Geometric mean is used instead
of classical mean because it uses log values. It is hence more robust
as it is less influenced by extreme values.</p>
<p><strong>Step 2</strong>: For every gene in every sample, ratios of
counts/pseudo-reference sample are calculated.</p>
<p><strong>Step 3</strong>: The median value of all ratios for a given sample is taken
as the scale factor for that sample. Importantly, the method is based
on the assumption that the majority of genes are not differentially
expressed, which implies that rare genes that are really up-regulated
or down-regulated should not influence the median. Furthermore, the
median is calculated skipping genes with a geometric mean of
zero. This will hence automatically eliminate genes expressed in some
samples but not in others and will help to focus the scaling factor on
housekeeping genes.</p>
<p><strong>Step 4</strong>: Normalised counts can be obtained by dividing each raw count
values in a given sample by that sample’s scaling factor.</p>
<p><img src="figs/SF_steps.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div id="count-modeling" class="section level3" number="4.1.2">
<h3>
<span class="header-section-number">4.1.2</span> Count modeling<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('count-modeling')" onmouseout="reset_tooltip('count-modeling-tooltip')"><span class="tooltiptext" id="count-modeling-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Let’s first have a look at the counts distribution for a typical
RNAseq sample:</p>
<p><img src="figs/distribution_of_counts.png" width="80%" style="display: block; margin: auto;"></p>
<p>It is obvious that the count data is not normally distributed. Counts
are integer values, always positive, and we observe a large number of
genes with low counts (or counts about zero), and a few number of
genes with a very high count level.</p>
<p>As seen in the <a href="http://bit.ly/WSBIM1322">WSBIM1322 course</a> with the
example of the coin toss, count data are often modelised by a binomial
distribution with parameters n and p where p is the discrete
probability distribution of the number of successes in a sequence of n
independent experiments. In an RNAseq experiment, p would be the
probability of getting a read associated to a particular gene given
that n total number of reads were sequenced in the
experiment. However, when n is large and p is low, Poisson
distribution is used instead of binomial. It describes typically the
distribution of a rare event in a large population, which fits better
to RNAseq. Indeed, for each sample, the total number of reads tends to
be in millions, while the counts per gene can vary considerably but
tend to be in tens, hundreds or thousands. Therefore, the chance of a
given read to be mapped to any specific gene is extremely small.</p>
<p>The Poisson distribution has only one parameter indicating its
expected mean. Its variance and all other properties follow from
it. In particular, one key assumption of the Poisson distribution is
that the variance equals the mean.</p>
<p><span class="math display">\[K_{ij} \sim P(µ_{ij} = \sigma^2_{ij})\]</span></p>
<p>Applying a Poisson distribution to
Rnaseq counts holds true when comparing technical replicates from a
same sample, where the variance only reflects the counting noise. But
when comparing biological replicates, counting noise is not the only
source of variance. The observed count values for each gene within
biological replicates fluctuate more importantly, due to the
combination of biological and technical factors: inter-individual
variations in gene expression, sample purity, cell reponses to
environment (e.g. heat-shock)… Due to this overdispersion, the
Poisson distribution doesn’t fit that well to RNAseq counts.</p>
<p>Actually, RNAseq counts are better modelised by an alternative
distribution, the negative-binomial. It is derived from the Poisson
distribution but the negative-binomial distribution has, in addition
to the mean parameter, an extra parameter <span class="math inline">\(α\)</span> called the “dispersion”
parameter to model this “extra” variance that is empirically observed
in RNA-Seq experiments.</p>
<p><span class="math display">\[K_{ij} \sim NB(mean = µ_{ij},  dispersion = \alpha_i)\]</span></p>
<p>The dispersion parameter <span class="math inline">\(\alpha_i\)</span> defines the relationship between
the variance of the observed count and its mean value<label for="tufte-sn-7" class="margin-toggle sidenote-number">7</label><input type="checkbox" id="tufte-sn-7" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">7</span> Note that as dispersion parameter gets lower and lower,
the variance converges to the same value as the mean, and the negative
binomial turns into a Poisson distribution.</span>.</p>
<p><span class="math display">\[VAR(K_{ij}) = µ_{ij} + \alpha_i.µ_{ij}^2\]</span></p>
<p>Having modelised counts by a negative-binomial distribution, next step
is to estimate, for each gene, the two parameters of the distribution
(mean and dispersion). The mean will be estimated easily from the
observed normalised counts in both conditions, but the dispersion is
not that trivial to estimate accurately.</p>
</div>
<div id="dispersion-estimation" class="section level3" number="4.1.3">
<h3>
<span class="header-section-number">4.1.3</span> Dispersion estimation<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('dispersion-estimation')" onmouseout="reset_tooltip('dispersion-estimation-tooltip')"><span class="tooltiptext" id="dispersion-estimation-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Dispersion is a measure of variability in the data (<span class="math inline">\(α = CV^2\)</span>).
A gene with a dispersion value of 0.04 means 20% variation
around the expected mean. Estimate the dispersion for each gene would
be quite straightforward if we had for each condition, hundreds of
replicates. Of course, this is not feasible for economic reasons, and
experiments are often done on only 3 replicates. But how to
estimate dispersion reliably based on such a little number of samples?
To overcome this problem, DESeq2 makes the assumption that genes of
similar expression levels have similar dispersions and it will use
information coming from other genes expressed at similar level.</p>
<p><strong>Step1: Dispersion for each gene is estimated separately.</strong> An
initial estimation of dispersion for each gene is first estimated
using maximum likelihood estimation. In other words, given the count
values of the replicates, the most likely estimate of dispersion is
calculated. For each gene, the dispersion estimate is plotted in
function of the mean expression level (mean counts of replicates).
This produce the so-called “dispersion plot” where each gene is
represented by a black dot.</p>
<p><img src="figs/dispersion_plot.png" width="80%" style="display: block; margin: auto;"></p>
<p>Note that the dispersion plot highlights an intrinsic feature of
RNAseq data: genes with low read counts show substantially higher
dispersion than highly expressed genes.</p>
<p><strong>Step 2: A curve is fitted to gene-wise dispersion estimates</strong>. A
curve is fitted (displayed as a red line in the dispersion plot),
which represents the estimate for the expected dispersion value for
genes of a given expression strength. The idea behind fitting a curve
to the data is that different genes will have different scales of
biological variability, but, over all genes, there will be a
distribution of reasonable estimates of dispersion.</p>
<p><strong>Step 3: Shrinkage of gene-wise dispersion estimates toward the
values predicted by the curve</strong>. Initial gene-wise dispersion
estimates will be shrinked (by an empirical Bayes approach) towards
this fitted curve to obtain the final dispersion estimates. The strength
of the shrinkage for each gene will depend on how close the dispersion
values are from the curve, and on the number of samples available (the
more sample (as the sample size increases, the shrinkage decreases in
strength, and eventually becomes negligible).
adjusted dispersion values are represented by the blue dots in the
dispersion plot. For a certain number of genes, the adjusted
dispersion will be significantly increased and this will limit the
number of false-positive that could arise from an underestimated
dispersion. Dispersion estimates that are slightly above the curve are
also shrunk toward the curve. However, genes with extremely high
dispersion<label for="tufte-sn-8" class="margin-toggle sidenote-number">8</label><input type="checkbox" id="tufte-sn-8" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">8</span> genes with extremely high dispersion are those for which the adjusted
dispersion is more than 2 residual standard deviations above the curve.</span> values are not. In fact DESeq2 assumes that
these genes might not follow the modeling assumptions and could have
higher variability than others for biological or technical
reasons. For these genes, shrinking the values toward the curve could
result in false positives. These genes are shown surrounded by blue
circles in the dispersion plot.</p>
<p>Dispersion shrinkage is particularly important to reduce false positives in the
differential expression analysis.</p>
<p><img src="figs/disp_shrinkage.png" width="80%" style="display: block; margin: auto;"></p>
</div>
<div id="generalized-linear-model" class="section level3" number="4.1.4">
<h3>
<span class="header-section-number">4.1.4</span> Generalized linear model<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('generalized-linear-model')" onmouseout="reset_tooltip('generalized-linear-model-tooltip')"><span class="tooltiptext" id="generalized-linear-model-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>DESeq2 fits a generalized linear model of the form: <span class="math display">\[log2(q_{ij}) = \Sigma x_j.β_i\]</span></p>
<p>where the parameter <span class="math inline">\(q_{ij}\)</span> is proportional to the expected true concentration of
gene i for sample j:</p>
<p><span class="math display">\[ q_{ij} = \frac{µ_{ij}}{SizeFactor_{j}} \]</span>
and <span class="math inline">\(β_i\)</span> represents the log2FC between conditions. <span class="math inline">\(β_i\)</span> coefficients are computed from the data.</p>
<p>In the case of a simple design with one condition (a treatment effect for example), the model
can be written</p>
<p><span class="math display">\[log2(q_{ij}) = \beta_0 + \beta_1.x_j + \epsilon\]</span></p>
<p><span class="math inline">\(\beta_0\)</span> is the log2 expression level in the reference (control samples)</p>
<p><span class="math inline">\(\beta_1\)</span> is the log2FC between treated and control cells</p>
<p><span class="math inline">\(x_j\)</span> = 0 if sample j is the control sample</p>
<p><span class="math inline">\(x_j\)</span> = 1 if sample j is the treated sample</p>
</div>
<div id="hypothesis-testing" class="section level3" number="4.1.5">
<h3>
<span class="header-section-number">4.1.5</span> Hypothesis testing<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('hypothesis-testing')" onmouseout="reset_tooltip('hypothesis-testing-tooltip')"><span class="tooltiptext" id="hypothesis-testing-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>The logFC are computed from the data using the GLM, and these are
associated to standard errors that depend on the variance of the
counts.</p>
<p><img src="figs/z_stat.png" width="80%" style="display: block; margin: auto;"></p>
<p>The ultimate goal of a test for differential expression is to decide
whether, for a given gene, an observed difference in read counts is
significant, that is, whether it is greater than what would be
expected just due to natural random variation.</p>
<p>The null hypothesis <span class="math inline">\(H_0\)</span> is that there is no differential expression across
the sample groups, which is the same as saying that the log2FC = 0.
A statistical test, the <strong>Wald test</strong>, will determine whether the data
provides sufficient evidence to conclude that this value is really
different from zero.</p>
<p>For the Wald test, the log2 fold-change is divided by its standard
error, resulting in a z-statistic. The z-statistic is compared to a
standard normal distribution, and a p-value is computed reporting the
probability that a z-statistic at least as extreme as the observed
value would be selected at random. In principle, if this p-value is
small (below a certain cutoff) the null hypothesis is rejected.</p>
</div>
<div id="multiple-testing-correction" class="section level3" number="4.1.6">
<h3>
<span class="header-section-number">4.1.6</span> Multiple testing correction<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('multiple-testing-correction')" onmouseout="reset_tooltip('multiple-testing-correction-tooltip')"><span class="tooltiptext" id="multiple-testing-correction-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Recall that a pvalue of 0.05 means that there is only 5% chance of
getting this data if no real difference existed (if <span class="math inline">\(H_0\)</span> was really
true). In other words, choosing a cut off of 0.05 means there is 5%
chance that the wrong decision is made (resulting in a false
positive). But remember the problematic of multiple testing seen in
chapter 7 from <a href="http://bit.ly/WSBIM1322">WSBIM1322 course</a>.</p>
<p>In a typical RNAseq differential expression analysis, we might have
about 20,000 genes to test and usually only a fraction of genes is
really differentially expressed. Imagine a drug treatment that
modifies the expression of about 1000 genes, but that has no impact on
the other ones. The first histogram shows how the distribution of
pvalues for truly modified genes (<span class="math inline">\(H_0\)</span> is false) would look like:
most of the pvalues would be very small. Using a pvalue cutoff of 0.05
should permit to identify most of these differentially expressed
genes. The second histogram shows the distribution of pvalues for
unmodified genes (<span class="math inline">\(H_0\)</span> is true). Here the p-values are uniformly
distributed between 0 and 1, and we can see that 5% of these genes
appear to be significant even though this is only by chance as the
drug had no real effect on them. But 5% of 19000 genes means … 950
false positive genes! Hence, pvalues obtained from the Wald test must
be corrected for multiple testing to avoid excess false positives.</p>
<p>By default DESeq2 uses Benjamini-Hochberg method to adjust pvalues.
The third histogram bellow illustrates the principle behind this
<strong>False discovery rate (FDR)</strong> adjustment. As differential expression
analysis is done on the whole set of genes, the resulting pvalues will
have a distribution corresponding to the combination of both
histograms. Most of the p-values are uniformly distributed between 0
and 1 but there is a spike to the left close to zero, due to those
p-values for which <span class="math inline">\(H_0\)</span> is false. The correction approach helps to
estimate how many of the significant values are actually false
positives. It tries to find the height where the p-value distribution
flattens out (corresponding to the red line) and incorporates this
height value into the calculation of FDR adjusted p-values.</p>
<p>Choosing a cut off of 0.05 for padjusted values now implies that 5% of
significant tests (but not 5% of all tests as before) will result in false
positives.</p>
<p><img src="figs/pval_histograms.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div id="independent-filtering" class="section level3" number="4.1.7">
<h3>
<span class="header-section-number">4.1.7</span> Independent filtering<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('independent-filtering')" onmouseout="reset_tooltip('independent-filtering-tooltip')"><span class="tooltiptext" id="independent-filtering-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Multiple testing adjustment tends to be associated with a loss of power.
To counteract this effect, one possibility is to filter out those tests from
the procedure that have no, or little chance of showing significant evidence,
without even looking at their test statistic. Genes with very low counts are
typically not likely to be significant due to high dispersion. However, these
genes have an influence on the multiple testing adjustment, whose performance
improves if such genes are removed. By removing the weakly-expressed genes from
the input to the FDR procedure, more significant genes can be found among those
that are kept, and this improves the power of the test. This approach is known
as independent filtering.</p>
<p>DESeq2 uses as filtering criterion the mean of normalised counts. Genes with a
mean expression value under a certain threshold are removed. Such filtering is
permissible only if the filter criterion is independent of the actual test statistic,
otherwise, the filtering would invalidate the test and consequently the assumptions
of the FDR procedure. This is why filtering is done on the average expression over
all samples, irrespective of biological condition: this filter is blind to the assignment
of samples to the treatment and control group and hence independent.</p>
<p>The mean expression threshold used by DESeq2 for independentfiltering is defined
automatically by the software. It is chosen in a way that maximizes the number of
genes which will have a significant p-adjusted value.
The threshold chosen (the vertical line) is the lowest quantile for which the number of
rejections is within 1 residual standard deviation to the peak of the curve.</p>
<p><img src="figs/IF_threshold.png" width="100%" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="running-deseq2" class="section level2" number="4.2">
<h2>
<span class="header-section-number">4.2</span> Running DESeq2<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('running-deseq2')" onmouseout="reset_tooltip('running-deseq2-tooltip')"><span class="tooltiptext" id="running-deseq2-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Let’s start by installing the
<a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a> package.</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="sec-rnaseq.html#cb107-1" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"DESeq2"</span>))</span>
<span id="cb107-2"><a href="sec-rnaseq.html#cb107-2" tabindex="-1"></a>    BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"DESeq2"</span>)</span>
<span id="cb107-3"><a href="sec-rnaseq.html#cb107-3" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb107-4"><a href="sec-rnaseq.html#cb107-4" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>We will run a DESeq2 analysis using real data.</p>
<div id="construct-deseqdataset-object" class="section level3" number="4.2.1">
<h3>
<span class="header-section-number">4.2.1</span> Construct DESeqDataSet object<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('construct-deseqdataset-object')" onmouseout="reset_tooltip('construct-deseqdataset-object-tooltip')"><span class="tooltiptext" id="construct-deseqdataset-object-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Let’s first load the count matrix and the sample metadata.
This dataset corresponds to RNAseq data from a cell line
treated or not by a siRNA.</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="sec-rnaseq.html#cb108-1" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"wsbim2122_data/deseq2/counts.rda"</span>)</span>
<span id="cb108-2"><a href="sec-rnaseq.html#cb108-2" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"wsbim2122_data/deseq2/coldata.rda"</span>)</span>
<span id="cb108-3"><a href="sec-rnaseq.html#cb108-3" tabindex="-1"></a>coldata</span></code></pre></div>
<pre><code>##          Cell       Type Condition
## sample1 Cell1 Epithelial      mock
## sample2 Cell1 Epithelial      mock
## sample3 Cell1 Epithelial      mock
## sample4 Cell1 Epithelial        KD
## sample5 Cell1 Epithelial        KD
## sample6 Cell1 Epithelial        KD</code></pre>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="sec-rnaseq.html#cb110-1" tabindex="-1"></a><span class="fu">head</span>(counts)</span></code></pre></div>
<pre><code>##                 sample1 sample2 sample3 sample4 sample5 sample6
## ENSG00000223972       0       0       0       0       0       1
## ENSG00000227232      14      28      17      40      16      13
## ENSG00000278267       8       4       3       1       1       6
## ENSG00000243485       0       0       0       0       0       0
## ENSG00000284332       0       0       0       0       0       0
## ENSG00000237613       0       0       0       0       0       0</code></pre>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="sec-rnaseq.html#cb112-1" tabindex="-1"></a><span class="fu">dim</span>(counts)</span></code></pre></div>
<pre><code>## [1] 58735     6</code></pre>
<p>Using these data, we will start by creating a
<a href="https://www.rdocumentation.org/packages/DESeq2/versions/1.12.3/topics/DESeqDataSet-class">DESeqDataSet</a>,
which is a subclass of
<a href="https://www.rdocumentation.org/packages/SummarizedExperiment/versions/1.2.3/topics/RangedSummarizedExperiment-class">RangedSummarizedExperiment</a>
used by the DESeq2 package to store the read counts and the
intermediate estimated quantities during statistical analysis. The
DESeqDataSet class enforces non-negative integer values in the count
matrix stored as the first element in the assay list. In addition, a
formula which specifies the design of the experiment (the variables
that will be used in modeling) must be provided.</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="sec-rnaseq.html#cb114-1" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> counts,</span>
<span id="cb114-2"><a href="sec-rnaseq.html#cb114-2" tabindex="-1"></a>                              <span class="at">colData =</span> coldata,</span>
<span id="cb114-3"><a href="sec-rnaseq.html#cb114-3" tabindex="-1"></a>                              <span class="at">design =</span> <span class="sc">~</span> Condition)</span></code></pre></div>
<pre><code>## converting counts to integer mode</code></pre>
<pre><code>## Warning in DESeqDataSet(se, design = design, ignoreRank): some variables in
## design formula are characters, converting to factors</code></pre>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="sec-rnaseq.html#cb117-1" tabindex="-1"></a>dds</span></code></pre></div>
<pre><code>## class: DESeqDataSet 
## dim: 58735 6 
## metadata(1): version
## assays(1): counts
## rownames(58735): ENSG00000223972 ENSG00000227232 ... ENSG00000277475
##   ENSG00000268674
## rowData names(0):
## colnames(6): sample1 sample2 ... sample5 sample6
## colData names(3): Cell Type Condition</code></pre>
<p>As for SummarizedExperiments (see chapter 3 from <a href="http://bit.ly/WSBIM1322">WSBIM1322 course</a>):</p>
<ul>
<li>The Quantitative data can be accessed with <code>assay()</code>.</li>
<li>The sample (columns) metadata can be access with the <code>colData()</code>
function.</li>
<li>The features (rows) metadata can be access with the <code>rowData()</code>
column.</li>
<li>Additional metadata describing the overall experiment can be
accessed with <code>metadata()</code>.</li>
</ul>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Access the count data from the dds object and plot the count distributions
of each sample.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-9" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-9', 'sol-start-9')"></span>
</p>
<div id="sol-body-9" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="sec-rnaseq.html#cb119-1" tabindex="-1"></a><span class="fu">as_tibble</span>(<span class="fu">assay</span>(dds), <span class="at">rownames =</span> <span class="st">"gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb119-2"><a href="sec-rnaseq.html#cb119-2" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">names_to =</span> <span class="st">"sample"</span>, <span class="at">values_to =</span> <span class="st">"counts"</span>, <span class="at">cols =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">7</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb119-3"><a href="sec-rnaseq.html#cb119-3" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log2</span>(counts <span class="sc">+</span> <span class="dv">1</span>), <span class="at">fill =</span> sample)) <span class="sc">+</span></span>
<span id="cb119-4"><a href="sec-rnaseq.html#cb119-4" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">20</span>) <span class="sc">+</span></span>
<span id="cb119-5"><a href="sec-rnaseq.html#cb119-5" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> sample)</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-78-1.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>The count matrix contains many rows with only zeros. We can start by removing
these rows of the dds object as anyway no statistical test can be done on
these genes that have no variation in their counts (equal to 0 in all samples).</p>
<p>It is not necessary to pre-filter low count genes before running
the DESeq2 functions, but by removing rows containing only zeros, we reduce the
memory size of the <code>dds</code> object and increase the speed of the analysis.
Additional filtering to improve power will be applied later.</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="sec-rnaseq.html#cb120-1" tabindex="-1"></a>dds <span class="ot">&lt;-</span> dds[<span class="fu">rowSums</span>(<span class="fu">assay</span>(dds)) <span class="sc">&gt;</span> <span class="dv">0</span>, ]</span>
<span id="cb120-2"><a href="sec-rnaseq.html#cb120-2" tabindex="-1"></a><span class="fu">dim</span>(dds)</span></code></pre></div>
<pre><code>## [1] 32102     6</code></pre>
</div>
<div id="run-deseq2" class="section level3" number="4.2.2">
<h3>
<span class="header-section-number">4.2.2</span> Run DESeq2<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('run-deseq2')" onmouseout="reset_tooltip('run-deseq2-tooltip')"><span class="tooltiptext" id="run-deseq2-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>The standard differential expression analysis steps are wrapped into a single
function <code>DESeq()</code>. This function will automatically run the following other functions:</p>
<ul>
<li><p><code>estimateSizeFactors()</code> (estimation of size factors)</p></li>
<li><p><code>estimateDispersions()</code> (estimation of dispersion)</p></li>
<li><p><code>nbinomWaldTest()</code> (Negative Binomial GLM fitting and Wald statistics)</p></li>
</ul>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="sec-rnaseq.html#cb122-1" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span></code></pre></div>
<pre><code>## estimating size factors</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## fitting model and testing</code></pre>
</div>
<div id="pca" class="section level3" number="4.2.3">
<h3>
<span class="header-section-number">4.2.3</span> PCA<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('pca')" onmouseout="reset_tooltip('pca-tooltip')"><span class="tooltiptext" id="pca-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Before anything else, a good practice is to explore the data
and perform quality controls checks.</p>
<p>It is highly recommended to starts by a PCA to assess overall similarity between
the samples:</p>
<ul>
<li><p>Which samples are similar/different to each other?</p></li>
<li><p>Does this fit to the expectation from the experiment’s design?</p></li>
<li><p>Are they any sample outliers which may need to be explored further?</p></li>
</ul>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Here are 3 examples of PCAs that correspond to different experimental designs.
How would you interprete these PCAs and what impact could they
have on the analysis?</p>
<p><img src="figs/PCA_paired_design.png" width="80%" style="display: block; margin: auto;"><img src="figs/PCA_batch_effect.png" width="80%" style="display: block; margin: auto;"><img src="figs/PCA_outlier.png" width="80%" style="display: block; margin: auto;"></p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>Remember that if one performs PCA directly on a matrix of normalised read counts,
the result typically depends only on the few most strongly expressed genes
because they show the largest absolute differences between samples.
A simple and often used strategy to avoid this is to take the logarithm of the
normalised count values plus a small pseudocount; however, now the genes with
low counts tend to dominate the results because, due to the strong Poisson noise
inherent to small count values, they show the strongest relative differences
between samples. As a solution, DESeq2 offers the regularized-logarithm
transformation <code>rlog()</code>.</p>
<p>For genes with high counts, the rlog transformation
differs not much from an ordinary log2 transformation. However for genes with
lower counts, the transformation moderates the variance across the mean,
shrunking the values towards the genes’ averages across all samples.
See <code>?rlog</code> for more details about the function.</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="sec-rnaseq.html#cb129-1" tabindex="-1"></a>rld <span class="ot">&lt;-</span> <span class="fu">rlogTransformation</span>(dds)</span></code></pre></div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ol style="list-style-type: decimal">
<li><p>ENSG00000198804 and ENSG00000248671 are two genes expressed respectively at a
very high and a very low level. Inspect the effect of the rlog transformation
on the counts of these genes. Compare to counts that would be transformed by a
classical log transformation (You can use the <code>normTransform()</code> function to
normalise and log-transform the dds counts)</p></li>
<li><p>Draw a scatter plot representing the log counts of sample1 versus the
log counts of sample2 (sample1 and sample2 are 2 control replicates).
Draw another scatter plot representing the “rlog transformed” counts of sample1
versus the “rlog transformed” counts of sample2. Compare the two plots.</p></li>
</ol>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-10" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-10', 'sol-start-10')"></span>
</p>
<div id="sol-body-10" class="solution-body" style="display: none;">
<p>Question 1</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="sec-rnaseq.html#cb130-1" tabindex="-1"></a><span class="fu">assay</span>(dds)[<span class="fu">c</span>(<span class="st">"ENSG00000198804"</span>, <span class="st">"ENSG00000248671"</span>),]</span></code></pre></div>
<pre><code>##                 sample1 sample2 sample3 sample4 sample5 sample6
## ENSG00000198804  143371  183617  178564  187194  151128  160159
## ENSG00000248671       1      16       1       5       3       8</code></pre>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="sec-rnaseq.html#cb132-1" tabindex="-1"></a><span class="co"># normalises and gives log2(n+1)</span></span>
<span id="cb132-2"><a href="sec-rnaseq.html#cb132-2" tabindex="-1"></a>log_dds <span class="ot">&lt;-</span> <span class="fu">normTransform</span>(dds)</span>
<span id="cb132-3"><a href="sec-rnaseq.html#cb132-3" tabindex="-1"></a></span>
<span id="cb132-4"><a href="sec-rnaseq.html#cb132-4" tabindex="-1"></a><span class="fu">assay</span>(log_dds)[<span class="fu">c</span>(<span class="st">"ENSG00000198804"</span>, <span class="st">"ENSG00000248671"</span>), ]</span></code></pre></div>
<pre><code>##                   sample1   sample2    sample3   sample4  sample5   sample6
## ENSG00000198804 17.589607 17.098256 17.3385798 17.564050 17.14591 17.270999
## ENSG00000248671  1.248376  3.725328  0.9472459  2.626638  1.95560  3.153798</code></pre>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="sec-rnaseq.html#cb134-1" tabindex="-1"></a><span class="fu">assay</span>(rld)[<span class="fu">c</span>(<span class="st">"ENSG00000198804"</span>, <span class="st">"ENSG00000248671"</span>), ]</span></code></pre></div>
<pre><code>##                   sample1   sample2   sample3   sample4   sample5  sample6
## ENSG00000198804 17.496502 17.189297 17.337713 17.480250 17.218320 17.29550
## ENSG00000248671  2.214236  2.511533  2.187179  2.318412  2.250167  2.39352</code></pre>
<p>Question 2</p>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb136-1"><a href="sec-rnaseq.html#cb136-1" tabindex="-1"></a>fig1 <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(<span class="fu">assay</span>(log_dds[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>])) <span class="sc">%&gt;%</span> </span>
<span id="cb136-2"><a href="sec-rnaseq.html#cb136-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> sample1, <span class="at">y =</span> sample2)) <span class="sc">+</span></span>
<span id="cb136-3"><a href="sec-rnaseq.html#cb136-3" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb136-4"><a href="sec-rnaseq.html#cb136-4" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">slope =</span> <span class="dv">1</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb136-5"><a href="sec-rnaseq.html#cb136-5" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"log2 transformation"</span>)</span>
<span id="cb136-6"><a href="sec-rnaseq.html#cb136-6" tabindex="-1"></a></span>
<span id="cb136-7"><a href="sec-rnaseq.html#cb136-7" tabindex="-1"></a>fig2 <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(<span class="fu">assay</span>(rld[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>])) <span class="sc">%&gt;%</span> </span>
<span id="cb136-8"><a href="sec-rnaseq.html#cb136-8" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> sample1, <span class="at">y =</span> sample2))<span class="sc">+</span></span>
<span id="cb136-9"><a href="sec-rnaseq.html#cb136-9" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb136-10"><a href="sec-rnaseq.html#cb136-10" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">slope =</span> <span class="dv">1</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb136-11"><a href="sec-rnaseq.html#cb136-11" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"rlogTransformation"</span>)</span>
<span id="cb136-12"><a href="sec-rnaseq.html#cb136-12" tabindex="-1"></a></span>
<span id="cb136-13"><a href="sec-rnaseq.html#cb136-13" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"patchwork"</span>)</span>
<span id="cb136-14"><a href="sec-rnaseq.html#cb136-14" tabindex="-1"></a>fig1 <span class="sc">|</span> fig2</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-83-1.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>The <code>plotPCA()</code> function can then be used on the transformed counts to generate a PCA.</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="sec-rnaseq.html#cb137-1" tabindex="-1"></a><span class="fu">plotPCA</span>(rld, <span class="at">intgroup =</span> <span class="st">"Condition"</span>)</span></code></pre></div>
<pre><code>## using ntop=500 top features by variance</code></pre>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-84-1.png" width="672"></p>
<p>Note that the argument <code>returnData = TRUE</code> can be used to obtain a dataframe of PC1 and PC2
for custom plotting.</p>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="sec-rnaseq.html#cb139-1" tabindex="-1"></a>pca_data <span class="ot">&lt;-</span> <span class="fu">plotPCA</span>(rld, <span class="at">intgroup =</span> <span class="st">"Condition"</span>, <span class="at">returnData =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## using ntop=500 top features by variance</code></pre>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="sec-rnaseq.html#cb141-1" tabindex="-1"></a><span class="fu">ggplot</span>(pca_data, <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, <span class="at">color =</span> Condition, <span class="at">label =</span> name)) <span class="sc">+</span></span>
<span id="cb141-2"><a href="sec-rnaseq.html#cb141-2" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb141-3"><a href="sec-rnaseq.html#cb141-3" tabindex="-1"></a>  <span class="fu">geom_text</span>()</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-85-1.png" width="672"></p>
</div>
<div id="inspecting-size-factors" class="section level3" number="4.2.4">
<h3>
<span class="header-section-number">4.2.4</span> Inspecting size factors<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('inspecting-size-factors')" onmouseout="reset_tooltip('inspecting-size-factors-tooltip')"><span class="tooltiptext" id="inspecting-size-factors-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>It is also advisable to investigate any systematic bias in the sequencing data, such as
whether one sample has been sequenced more deeply than others. One can extract size factors
using the <code>sizeFactors()</code> function.
Usually these size factors should vary around 1, indicating comparable sequencing depth.</p>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb142-1"><a href="sec-rnaseq.html#cb142-1" tabindex="-1"></a><span class="fu">sizeFactors</span>(dds)</span></code></pre></div>
<pre><code>##   sample1   sample2   sample3   sample4   sample5   sample6 
## 0.7268826 1.3086639 1.0773676 0.9660247 1.0421106 1.0126638</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Compare Size Factors to sequencing depth.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-11" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-11', 'sol-start-11')"></span>
</p>
<div id="sol-body-11" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="sec-rnaseq.html#cb144-1" tabindex="-1"></a>SF <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">sample =</span> <span class="fu">colnames</span>(dds), <span class="at">SF =</span> <span class="fu">sizeFactors</span>(dds)) <span class="sc">%&gt;%</span></span>
<span id="cb144-2"><a href="sec-rnaseq.html#cb144-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> sample, <span class="at">y =</span> SF)) <span class="sc">+</span></span>
<span id="cb144-3"><a href="sec-rnaseq.html#cb144-3" tabindex="-1"></a>  <span class="fu">geom_col</span>()</span>
<span id="cb144-4"><a href="sec-rnaseq.html#cb144-4" tabindex="-1"></a></span>
<span id="cb144-5"><a href="sec-rnaseq.html#cb144-5" tabindex="-1"></a>SD <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">sample =</span> <span class="fu">colnames</span>(dds), <span class="at">SeqDepth =</span> <span class="fu">colSums</span>(<span class="fu">assay</span>(dds))) <span class="sc">%&gt;%</span></span>
<span id="cb144-6"><a href="sec-rnaseq.html#cb144-6" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> sample, <span class="at">y =</span> SeqDepth)) <span class="sc">+</span></span>
<span id="cb144-7"><a href="sec-rnaseq.html#cb144-7" tabindex="-1"></a>  <span class="fu">geom_col</span>()</span>
<span id="cb144-8"><a href="sec-rnaseq.html#cb144-8" tabindex="-1"></a></span>
<span id="cb144-9"><a href="sec-rnaseq.html#cb144-9" tabindex="-1"></a>SF <span class="sc">/</span> SD</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/size_factor-1.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="dispersion-plot" class="section level3" number="4.2.5">
<h3>
<span class="header-section-number">4.2.5</span> Dispersion plot<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('dispersion-plot')" onmouseout="reset_tooltip('dispersion-plot-tooltip')"><span class="tooltiptext" id="dispersion-plot-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Plotting the dispersion estimates is a useful diagnostic.
This dispersion plot is typical, with the final dispersion, estimates shrunk from the
gene-wise estimates towards the fitted estimates.</p>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="sec-rnaseq.html#cb145-1" tabindex="-1"></a><span class="fu">plotDispEsts</span>(dds)</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/dispersion_plot-1.png" width="672"></p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ol style="list-style-type: decimal">
<li><p>Create a new fictive DESeqDataSet object containing only 2 replicates of each
condition instead of the 3 replicates that we have in our <code>dds</code> object.
Inspect its dispersion plot and compare it with the dispersion plot of <code>dds</code>.</p></li>
<li><p>Using the <code>dispersions()</code> function, you can acceed to the dispersion values.
Compare on a scatter plot the dispersions values obtained with 2 replicates
and with 3 replicates.</p></li>
</ol>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-12" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-12', 'sol-start-12')"></span>
</p>
<div id="sol-body-12" class="solution-body" style="display: none;">
<p>Question 1.</p>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="sec-rnaseq.html#cb146-1" tabindex="-1"></a>dds_2_replicates <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> counts[, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">6</span>)],</span>
<span id="cb146-2"><a href="sec-rnaseq.html#cb146-2" tabindex="-1"></a>                                         <span class="at">colData =</span> coldata[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">6</span>), ],</span>
<span id="cb146-3"><a href="sec-rnaseq.html#cb146-3" tabindex="-1"></a>                                         <span class="at">design =</span> <span class="sc">~</span> Condition)</span>
<span id="cb146-4"><a href="sec-rnaseq.html#cb146-4" tabindex="-1"></a><span class="co"># Keep only genes present in dds to be able to compare the 2 objects</span></span>
<span id="cb146-5"><a href="sec-rnaseq.html#cb146-5" tabindex="-1"></a>dds_2_replicates <span class="ot">&lt;-</span> dds_2_replicates[<span class="fu">rownames</span>(dds), ]</span>
<span id="cb146-6"><a href="sec-rnaseq.html#cb146-6" tabindex="-1"></a>dds_2_replicates <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds_2_replicates)</span>
<span id="cb146-7"><a href="sec-rnaseq.html#cb146-7" tabindex="-1"></a></span>
<span id="cb146-8"><a href="sec-rnaseq.html#cb146-8" tabindex="-1"></a><span class="fu">plotDispEsts</span>(dds_2_replicates) </span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/disp_exercice-1.png" width="672"></p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="sec-rnaseq.html#cb147-1" tabindex="-1"></a><span class="fu">plotDispEsts</span>(dds)</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/disp_exercice-2.png" width="672"></p>
<p>Question 2.</p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="sec-rnaseq.html#cb148-1" tabindex="-1"></a><span class="fu">tibble</span>(<span class="at">dispersions_3_replicates =</span> <span class="fu">dispersions</span>(dds), </span>
<span id="cb148-2"><a href="sec-rnaseq.html#cb148-2" tabindex="-1"></a>       <span class="at">dispersions_2_replicates =</span> <span class="fu">dispersions</span>(dds_2_replicates)) <span class="sc">%&gt;%</span> </span>
<span id="cb148-3"><a href="sec-rnaseq.html#cb148-3" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> dispersions_3_replicates, <span class="at">y =</span> dispersions_2_replicates)) <span class="sc">+</span></span>
<span id="cb148-4"><a href="sec-rnaseq.html#cb148-4" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb148-5"><a href="sec-rnaseq.html#cb148-5" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">slope =</span> <span class="dv">1</span>, <span class="at">color =</span> <span class="st">"red"</span>)</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-87-1.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="deseq2-results" class="section level3" number="4.2.6">
<h3>
<span class="header-section-number">4.2.6</span> DESeq2 results<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('deseq2-results')" onmouseout="reset_tooltip('deseq2-results-tooltip')"><span class="tooltiptext" id="deseq2-results-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>The <code>resultsNames()</code> function gives the names of results that can be extracted.
Note that by default, R will choose a reference level for factors based on alphabetical
order. Here KD condition was hence set arbitrarily as the reference level. This means
that the fold changes evaluated for every gene will correspond to their expression level
in mock cells versus their expression level in KD cells.</p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="sec-rnaseq.html#cb149-1" tabindex="-1"></a><span class="fu">resultsNames</span>(dds)</span></code></pre></div>
<pre><code>## [1] "Intercept"            "Condition_mock_vs_KD"</code></pre>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="sec-rnaseq.html#cb151-1" tabindex="-1"></a>dds<span class="sc">$</span>Condition</span></code></pre></div>
<pre><code>## [1] mock mock mock KD   KD   KD  
## Levels: KD mock</code></pre>
<p>In this case, it seems more logical to use mock cells as reference.
Reference levels can be changed using the <code>relevel()</code> function. But in this
case don’t forget to re-run <code>DESeq()</code> function after the re-leveling operation.</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="sec-rnaseq.html#cb153-1" tabindex="-1"></a>dds<span class="sc">$</span>Condition <span class="ot">&lt;-</span> <span class="fu">relevel</span>(dds<span class="sc">$</span>Condition, <span class="at">ref =</span> <span class="st">"mock"</span>)</span>
<span id="cb153-2"><a href="sec-rnaseq.html#cb153-2" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span>
<span id="cb153-3"><a href="sec-rnaseq.html#cb153-3" tabindex="-1"></a><span class="fu">resultsNames</span>(dds)</span></code></pre></div>
<pre><code>## [1] "Intercept"            "Condition_KD_vs_mock"</code></pre>
<p>Results can then be extracted using the <code>results()</code> function.</p>
<p>The <code>name</code> parameter must be an element of resultsNames(object) specifying the
samples to compare.</p>
<p>Note that the <code>contrast</code> argument of <code>results()</code> can also be used to
extract results of interest. In this case, the two following commands will give
the same results<label for="tufte-sn-9" class="margin-toggle sidenote-number">9</label><input type="checkbox" id="tufte-sn-9" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">9</span> Using the <code>name</code> argument is however necessary to extract specific
coefficients from more complex designs.</span>.</p>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="sec-rnaseq.html#cb155-1" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds,</span>
<span id="cb155-2"><a href="sec-rnaseq.html#cb155-2" tabindex="-1"></a>               <span class="at">name =</span> <span class="st">"Condition_KD_vs_mock"</span>)</span></code></pre></div>
<div class="sourceCode" id="cb156"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb156-1"><a href="sec-rnaseq.html#cb156-1" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds,</span>
<span id="cb156-2"><a href="sec-rnaseq.html#cb156-2" tabindex="-1"></a>               <span class="at">contrast =</span> <span class="fu">c</span>(<span class="st">"Condition"</span>, <span class="st">"KD"</span>, <span class="st">"mock"</span>))</span></code></pre></div>
<p>Let’s inspect the results and their signification.</p>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="sec-rnaseq.html#cb157-1" tabindex="-1"></a>res_tbl <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(res, <span class="at">rownames =</span> <span class="st">"ENSEMBL"</span>)</span>
<span id="cb157-2"><a href="sec-rnaseq.html#cb157-2" tabindex="-1"></a>res_tbl</span></code></pre></div>
<pre><code>## # A tibble: 32,102 × 7
##    ENSEMBL         baseMean log2FoldChange lfcSE   stat pvalue   padj
##    &lt;chr&gt;              &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
##  1 ENSG00000223972    0.165          0.964 4.08   0.236 0.813  NA    
##  2 ENSG00000227232   21.0            0.296 0.540  0.548 0.583   0.739
##  3 ENSG00000278267    4.13          -1.04  1.14  -0.906 0.365  NA    
##  4 ENSG00000238009    3.49          -1.69  1.22  -1.39  0.165  NA    
##  5 ENSG00000233750   10.2           -0.519 0.702 -0.739 0.460   0.642
##  6 ENSG00000268903   19.4           -0.630 0.606 -1.04  0.298   0.489
##  7 ENSG00000269981   18.1           -1.19  0.561 -2.12  0.0343  0.104
##  8 ENSG00000239906    0.229         -0.959 4.08  -0.235 0.814  NA    
##  9 ENSG00000241860    3.34          -0.982 1.23  -0.799 0.424  NA    
## 10 ENSG00000279928    0.255         -1.62  4.05  -0.399 0.690  NA    
## # ℹ 32,092 more rows</code></pre>
<p><strong>baseMean</strong>: The average of the normalised count values taken over all samples.</p>
<p><strong>log2FoldChange</strong>: Fold-change between the comparison and control groups,
reported on a logarithmic scale to base 2. In this case, as we are testing
the “Condition_KD_vs_mock” coefficient, positive log2FC indicates a gene up-regulated
in the KD condition compared to the mock condition, while a negative log2FC indicate
a down-regulation.</p>
<p><strong>lfcSE</strong>: The standard error estimate for the log2FoldChange estimate</p>
<p><strong>stat</strong>: The value of the test statistic for the gene</p>
<p><strong>pvalue</strong>: The pvalue of the test for the gene</p>
<p><strong>padj</strong>: pvalue adjusted for multiple testing</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ol style="list-style-type: decimal">
<li><p>Inspect the results table and identify the 5 “best genes” showing
the lowest padjusted value.</p></li>
<li><p>Calculate the mean expression level of these 5 “best genes” using
the function <code>counts()</code>. Compare with baseMean values.</p></li>
<li><p>Extract the ß coefficient of these 5 “best genes” from the GLM
using the function <code>coefficients()</code>. Compare with log2FoldChange values.</p></li>
<li><p>Using the function <code>counts()</code>, evaluate the mean expression levels
of these 5 “best genes” in mock cells. Compare with ß coefficients.</p></li>
<li><p>Evaluate the mean expression levels of these 5 “best genes”
in KD cells. Compare with ß coefficients.</p></li>
<li><p>How many genes have no padjusted value? Why?</p></li>
</ol>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-13" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-13', 'sol-start-13')"></span>
</p>
<div id="sol-body-13" class="solution-body" style="display: none;">
<ol style="list-style-type: decimal">
<li>Inspect the results table and identify the 5 “best genes” showing
the lowest padjusted value.</li>
</ol>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="sec-rnaseq.html#cb159-1" tabindex="-1"></a>res_tbl <span class="sc">%&gt;%</span></span>
<span id="cb159-2"><a href="sec-rnaseq.html#cb159-2" tabindex="-1"></a>  <span class="fu">arrange</span>(padj) <span class="sc">%&gt;%</span></span>
<span id="cb159-3"><a href="sec-rnaseq.html#cb159-3" tabindex="-1"></a>  <span class="fu">head</span>(<span class="dv">5</span>)</span></code></pre></div>
<pre><code>## # A tibble: 5 × 7
##   ENSEMBL         baseMean log2FoldChange lfcSE  stat    pvalue      padj
##   &lt;chr&gt;              &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
## 1 ENSG00000101856    1209.          -4.67 0.178 -26.2 8.24e-152 1.36e-147
## 2 ENSG00000177494     277.           6.25 0.351  17.8 6.25e- 71 5.17e- 67
## 3 ENSG00000136159     630.           2.24 0.154  14.6 3.32e- 48 1.83e- 44
## 4 ENSG00000175315     220.           4.29 0.300  14.3 2.20e- 46 9.11e- 43
## 5 ENSG00000128245    2692.           2.05 0.144  14.2 5.64e- 46 1.87e- 42</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Calculate the mean expression level of these 5 “best genes” using
the function <code>counts()</code>. Compare with baseMean values.</li>
</ol>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="sec-rnaseq.html#cb161-1" tabindex="-1"></a>best_genes <span class="ot">&lt;-</span> res_tbl <span class="sc">%&gt;%</span></span>
<span id="cb161-2"><a href="sec-rnaseq.html#cb161-2" tabindex="-1"></a>  <span class="fu">arrange</span>(padj) <span class="sc">%&gt;%</span></span>
<span id="cb161-3"><a href="sec-rnaseq.html#cb161-3" tabindex="-1"></a>  <span class="fu">head</span>(<span class="dv">5</span>) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(ENSEMBL)</span>
<span id="cb161-4"><a href="sec-rnaseq.html#cb161-4" tabindex="-1"></a></span>
<span id="cb161-5"><a href="sec-rnaseq.html#cb161-5" tabindex="-1"></a><span class="fu">rowMeans</span>(<span class="fu">counts</span>(dds[best_genes], <span class="at">normalize =</span> <span class="cn">TRUE</span>))</span></code></pre></div>
<pre><code>## ENSG00000101856 ENSG00000177494 ENSG00000136159 ENSG00000175315 ENSG00000128245 
##       1209.2283        277.1334        630.0444        219.6860       2692.2250</code></pre>
<ol start="3" style="list-style-type: decimal">
<li>Extract the ß coefficient of these 5 “best genes” from the GLM
using the function <code>coefficients()</code>. Compare with log2FoldChange values.</li>
</ol>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="sec-rnaseq.html#cb163-1" tabindex="-1"></a><span class="fu">coefficients</span>(dds[best_genes])</span></code></pre></div>
<pre><code>##                 Intercept Condition_KD_vs_mock
## ENSG00000101856 11.183974            -4.673040
## ENSG00000177494  2.843588             6.249730
## ENSG00000136159  7.780272             2.243621
## ENSG00000175315  4.413188             4.290890
## ENSG00000128245 10.032905             2.049342</code></pre>
<ol start="4" style="list-style-type: decimal">
<li>Using the function <code>counts()</code>, evaluate the mean expression levels
of these 5 “best genes” in mock cells. Compare with ß coefficients.</li>
</ol>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="sec-rnaseq.html#cb165-1" tabindex="-1"></a><span class="co"># log2 expression levels in mock cells</span></span>
<span id="cb165-2"><a href="sec-rnaseq.html#cb165-2" tabindex="-1"></a><span class="fu">log2</span>(<span class="fu">rowMeans</span>(<span class="fu">counts</span>(dds[best_genes, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">normalize =</span> <span class="cn">TRUE</span>)))</span></code></pre></div>
<pre><code>## ENSG00000101856 ENSG00000177494 ENSG00000136159 ENSG00000175315 ENSG00000128245 
##       11.184348        3.004525        7.773839        4.473037       10.033523</code></pre>
<ol start="5" style="list-style-type: decimal">
<li>Evaluate the mean expression levels of these 5 “best genes”
in KD cells. Compare with ß coefficients.</li>
</ol>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="sec-rnaseq.html#cb167-1" tabindex="-1"></a><span class="co"># log2 expression levels in KD cells</span></span>
<span id="cb167-2"><a href="sec-rnaseq.html#cb167-2" tabindex="-1"></a><span class="fu">log2</span>(<span class="fu">rowMeans</span>(<span class="fu">counts</span>(dds[best_genes, <span class="dv">4</span><span class="sc">:</span><span class="dv">6</span>], <span class="at">normalize =</span> <span class="cn">TRUE</span>)))</span></code></pre></div>
<pre><code>## ENSG00000101856 ENSG00000177494 ENSG00000136159 ENSG00000175315 ENSG00000128245 
##        6.512646        9.093396       10.024077        8.704469       12.082275</code></pre>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="sec-rnaseq.html#cb169-1" tabindex="-1"></a><span class="co"># log2 expression levels in KD cells evaluated from ß coefficients</span></span>
<span id="cb169-2"><a href="sec-rnaseq.html#cb169-2" tabindex="-1"></a><span class="fu">coefficients</span>(dds[best_genes])[,<span class="dv">1</span>] <span class="sc">+</span> <span class="fu">coefficients</span>(dds[best_genes])[,<span class="dv">2</span>]</span></code></pre></div>
<pre><code>## ENSG00000101856 ENSG00000177494 ENSG00000136159 ENSG00000175315 ENSG00000128245 
##        6.510934        9.093318       10.023893        8.704078       12.082248</code></pre>
<ol start="6" style="list-style-type: decimal">
<li>How many genes have no padjusted value? Why?</li>
</ol>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb171-1"><a href="sec-rnaseq.html#cb171-1" tabindex="-1"></a><span class="fu">summary</span>(res<span class="sc">$</span>padj)</span></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
##   0.000   0.039   0.320   0.382   0.687   1.000   15560</code></pre>
<p>Let’s inspect genes with no padjusted values</p>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb173-1"><a href="sec-rnaseq.html#cb173-1" tabindex="-1"></a>res_tbl <span class="sc">%&gt;%</span></span>
<span id="cb173-2"><a href="sec-rnaseq.html#cb173-2" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">is.na</span>(padj))</span></code></pre></div>
<pre><code>## # A tibble: 15,560 × 7
##    ENSEMBL         baseMean log2FoldChange lfcSE   stat pvalue  padj
##    &lt;chr&gt;              &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;
##  1 ENSG00000223972    0.165          0.964  4.08  0.236  0.813    NA
##  2 ENSG00000278267    4.13          -1.04   1.14 -0.906  0.365    NA
##  3 ENSG00000238009    3.49          -1.69   1.22 -1.39   0.165    NA
##  4 ENSG00000239906    0.229         -0.959  4.08 -0.235  0.814    NA
##  5 ENSG00000241860    3.34          -0.982  1.23 -0.799  0.424    NA
##  6 ENSG00000279928    0.255         -1.62   4.05 -0.399  0.690    NA
##  7 ENSG00000236679    1.28          -0.756  1.94 -0.390  0.696    NA
##  8 ENSG00000233653    0.165          0.964  4.08  0.236  0.813    NA
##  9 ENSG00000250575    1.28          -1.45   2.01 -0.720  0.471    NA
## 10 ENSG00000268663    0.127         -0.959  4.08 -0.235  0.814    NA
## # ℹ 15,550 more rows</code></pre>
<p>These genes all have a very low basemean and are the one that have been filtered
out by the independent filtering procedure.</p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="independent-filtering-exploration" class="section level3" number="4.2.7">
<h3>
<span class="header-section-number">4.2.7</span> Independent filtering exploration<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('independent-filtering-exploration')" onmouseout="reset_tooltip('independent-filtering-exploration-tooltip')"><span class="tooltiptext" id="independent-filtering-exploration-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>The filtering threshold that has been used to filter low count genes can be extracted
from the results metadata.</p>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb175-1"><a href="sec-rnaseq.html#cb175-1" tabindex="-1"></a><span class="fu">metadata</span>(res)<span class="sc">$</span>filterThreshold</span></code></pre></div>
<pre><code>## 48.46939% 
##  4.215383</code></pre>
<p>This means that genes whith basemean &lt; 4.215383 have been
filtered. This represents 48.46939% of all tested genes!</p>
<p>Remember that the filtering threshold has been fixed in a way that maximizes the
number of genes which will have a significant padjusted value.</p>
<p>We can plot the number of rejections over the basemean quantiles.
The threshold chosen (red vertical line) is the lowest quantile for which the number of
rejections is within 1 residual standard deviation to the peak of the curve.</p>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="sec-rnaseq.html#cb177-1" tabindex="-1"></a><span class="fu">as_tibble</span>(<span class="fu">metadata</span>(res)<span class="sc">$</span>filterNumRej) <span class="sc">%&gt;%</span></span>
<span id="cb177-2"><a href="sec-rnaseq.html#cb177-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> theta, <span class="at">y =</span> numRej)) <span class="sc">+</span></span>
<span id="cb177-3"><a href="sec-rnaseq.html#cb177-3" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb177-4"><a href="sec-rnaseq.html#cb177-4" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fl">0.484</span>,</span>
<span id="cb177-5"><a href="sec-rnaseq.html#cb177-5" tabindex="-1"></a>             <span class="at">color =</span> <span class="st">'red'</span>)</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-102-1.png" width="672"></p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ol style="list-style-type: decimal">
<li><p>Re-run the results() function on the same dds object,
but set the independent filtering parameter to FALSE.
Check how many genes have no padj?</p></li>
<li><p>Compare the number of significant genes obtained with and without
independent filtering</p></li>
<li><p>Imagine another way of filtering genes with very low counts</p></li>
</ol>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-14" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-14', 'sol-start-14')"></span>
</p>
<div id="sol-body-14" class="solution-body" style="display: none;">
<ol style="list-style-type: decimal">
<li>Re-run the results() function on the same dds object,
but set the independent filtering parameter to FALSE.
Check how many genes have no padj?</li>
</ol>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="sec-rnaseq.html#cb178-1" tabindex="-1"></a>res_no_IF <span class="ot">&lt;-</span> <span class="fu">results</span>(dds,</span>
<span id="cb178-2"><a href="sec-rnaseq.html#cb178-2" tabindex="-1"></a>                     <span class="at">name =</span> <span class="st">"Condition_KD_vs_mock"</span>,</span>
<span id="cb178-3"><a href="sec-rnaseq.html#cb178-3" tabindex="-1"></a>                     <span class="at">independentFiltering =</span> <span class="cn">FALSE</span>)</span>
<span id="cb178-4"><a href="sec-rnaseq.html#cb178-4" tabindex="-1"></a></span>
<span id="cb178-5"><a href="sec-rnaseq.html#cb178-5" tabindex="-1"></a>res_no_IF_tbl <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(res_no_IF, <span class="at">rownames =</span> <span class="st">"ENSEMBL"</span>)</span>
<span id="cb178-6"><a href="sec-rnaseq.html#cb178-6" tabindex="-1"></a></span>
<span id="cb178-7"><a href="sec-rnaseq.html#cb178-7" tabindex="-1"></a>res_no_IF_tbl <span class="sc">%&gt;%</span></span>
<span id="cb178-8"><a href="sec-rnaseq.html#cb178-8" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">is.na</span>(padj)) <span class="sc">%&gt;%</span> <span class="fu">nrow</span>()</span></code></pre></div>
<pre><code>## [1] 0</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Compare the number of significant genes obtained with and without
independent filtering</li>
</ol>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="sec-rnaseq.html#cb180-1" tabindex="-1"></a>res_tbl <span class="sc">%&gt;%</span> <span class="fu">filter</span>(padj <span class="sc">&lt;</span> .<span class="dv">05</span>) <span class="sc">%&gt;%</span> <span class="fu">nrow</span>()</span></code></pre></div>
<pre><code>## [1] 4436</code></pre>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="sec-rnaseq.html#cb182-1" tabindex="-1"></a>res_no_IF_tbl <span class="sc">%&gt;%</span> <span class="fu">filter</span>(padj <span class="sc">&lt;</span> .<span class="dv">05</span>) <span class="sc">%&gt;%</span> <span class="fu">nrow</span>()</span></code></pre></div>
<pre><code>## [1] 3772</code></pre>
<ol start="3" style="list-style-type: decimal">
<li>Imagine another way of filtering genes with very low counts</li>
</ol>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="sec-rnaseq.html#cb184-1" tabindex="-1"></a><span class="co"># filter the data to remove genes with few counts</span></span>
<span id="cb184-2"><a href="sec-rnaseq.html#cb184-2" tabindex="-1"></a>filtering_thr <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb184-3"><a href="sec-rnaseq.html#cb184-3" tabindex="-1"></a><span class="co"># keep genes with counts &gt; 5 in 3 or more samples</span></span>
<span id="cb184-4"><a href="sec-rnaseq.html#cb184-4" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">counts</span>(dds, <span class="at">normalized =</span> <span class="cn">TRUE</span>) <span class="sc">&gt;=</span> filtering_thr) <span class="sc">&gt;=</span><span class="dv">3</span></span>
<span id="cb184-5"><a href="sec-rnaseq.html#cb184-5" tabindex="-1"></a>dds_bis <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds[keep, ])</span>
<span id="cb184-6"><a href="sec-rnaseq.html#cb184-6" tabindex="-1"></a><span class="fu">dim</span>(dds_bis)</span></code></pre></div>
<pre><code>## [1] 16152     6</code></pre>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="sec-rnaseq.html#cb186-1" tabindex="-1"></a>res_bis <span class="ot">&lt;-</span> <span class="fu">results</span>(dds_bis,</span>
<span id="cb186-2"><a href="sec-rnaseq.html#cb186-2" tabindex="-1"></a>                   <span class="at">name =</span> <span class="st">"Condition_KD_vs_mock"</span>,</span>
<span id="cb186-3"><a href="sec-rnaseq.html#cb186-3" tabindex="-1"></a>                   <span class="at">independentFiltering =</span> <span class="cn">FALSE</span>)</span>
<span id="cb186-4"><a href="sec-rnaseq.html#cb186-4" tabindex="-1"></a>res_bis_tbl <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(res_bis, <span class="at">rownames =</span> <span class="st">"ENSEMBL"</span>) </span>
<span id="cb186-5"><a href="sec-rnaseq.html#cb186-5" tabindex="-1"></a></span>
<span id="cb186-6"><a href="sec-rnaseq.html#cb186-6" tabindex="-1"></a><span class="co"># Compare with previous analysis</span></span>
<span id="cb186-7"><a href="sec-rnaseq.html#cb186-7" tabindex="-1"></a><span class="co"># (Independant filtering threshold fixed automatically by DESeq2)</span></span>
<span id="cb186-8"><a href="sec-rnaseq.html#cb186-8" tabindex="-1"></a>res_tbl <span class="sc">%&gt;%</span> <span class="fu">filter</span>(padj <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code></pre></div>
<pre><code>## # A tibble: 4,436 × 7
##    ENSEMBL         baseMean log2FoldChange lfcSE  stat       pvalue        padj
##    &lt;chr&gt;              &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
##  1 ENSG00000237094    36.1          -1.12  0.400 -2.81 0.00502      0.0228     
##  2 ENSG00000228327    61.5          -1.12  0.308 -3.65 0.000264     0.00199    
##  3 ENSG00000187634     6.17         -4.16  1.29  -3.22 0.00127      0.00736    
##  4 ENSG00000187583    22.3           1.40  0.489  2.86 0.00430      0.0201     
##  5 ENSG00000187642    12.5           2.02  0.737  2.74 0.00613      0.0268     
##  6 ENSG00000176022   829.            0.765 0.137  5.60 0.0000000215 0.000000559
##  7 ENSG00000107404  1888.            1.08  0.212  5.10 0.000000334  0.00000645 
##  8 ENSG00000242485  1475.            0.369 0.146  2.53 0.0114       0.0441     
##  9 ENSG00000160072  1015.            1.01  0.213  4.76 0.00000197   0.0000304  
## 10 ENSG00000197785  1565.            0.543 0.148  3.65 0.000258     0.00195    
## # ℹ 4,426 more rows</code></pre>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="sec-rnaseq.html#cb188-1" tabindex="-1"></a>res_bis_tbl <span class="sc">%&gt;%</span> <span class="fu">filter</span>(padj <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code></pre></div>
<pre><code>## # A tibble: 4,532 × 7
##    ENSEMBL         baseMean log2FoldChange lfcSE  stat       pvalue        padj
##    &lt;chr&gt;              &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
##  1 ENSG00000237094    36.1          -1.12  0.381 -2.95 0.00321      0.0151     
##  2 ENSG00000228327    61.5          -1.12  0.295 -3.81 0.000139     0.00111    
##  3 ENSG00000187634     6.17         -4.16  1.25  -3.32 0.000901     0.00528    
##  4 ENSG00000187583    22.3           1.40  0.465  3.01 0.00263      0.0129     
##  5 ENSG00000187642    12.5           2.03  0.703  2.89 0.00391      0.0178     
##  6 ENSG00000176022   829.            0.765 0.138  5.56 0.0000000275 0.000000669
##  7 ENSG00000107404  1888.            1.08  0.214  5.05 0.000000430  0.00000770 
##  8 ENSG00000242485  1475.            0.369 0.148  2.50 0.0125       0.0455     
##  9 ENSG00000160072  1015.            1.01  0.214  4.73 0.00000226   0.0000328  
## 10 ENSG00000197785  1565.            0.543 0.150  3.61 0.000309     0.00218    
## # ℹ 4,522 more rows</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="p-values-distribution" class="section level3" number="4.2.8">
<h3>
<span class="header-section-number">4.2.8</span> p-values distribution<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('p-values-distribution')" onmouseout="reset_tooltip('p-values-distribution-tooltip')"><span class="tooltiptext" id="p-values-distribution-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Another useful diagnostic plot is the histogram of pvalues.
It can give you an immediate idea of the proportion of genes differentially expressed,
(the taller the is the left peak, the more p-values are close to 0 and therefore
significant). Furthermore, it gives an idea of how the test behaved across all your
hypotheses, and immediately diagnoses some potential problems.</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="sec-rnaseq.html#cb190-1" tabindex="-1"></a><span class="fu">hist</span>(res_tbl<span class="sc">$</span>pvalue)</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-106-1.png" width="672"></p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>What do you think about the pvalues histogram?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-15" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-15', 'sol-start-15')"></span>
</p>
<div id="sol-body-15" class="solution-body" style="display: none;">
<p>There is a peak at 0, but there is also a peak close to 0.8.
Assumption that the p-values near 1 are uniform doesn’t seem to be completely valid.
This implies that in principle, a false discovery rate shouldn’t be applied to
control these p-values.
Let’s try to figure out why the p-values show this behavior.</p>
<div class="sourceCode" id="cb191"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb191-1"><a href="sec-rnaseq.html#cb191-1" tabindex="-1"></a>res_tbl <span class="sc">%&gt;%</span></span>
<span id="cb191-2"><a href="sec-rnaseq.html#cb191-2" tabindex="-1"></a>  <span class="fu">filter</span>(pvalue <span class="sc">&gt;</span> <span class="fl">0.8</span> <span class="sc">&amp;</span> pvalue <span class="sc">&lt;</span> <span class="fl">0.85</span>) <span class="sc">%&gt;%</span></span>
<span id="cb191-3"><a href="sec-rnaseq.html#cb191-3" tabindex="-1"></a>  <span class="fu">head</span>(<span class="dv">10</span>)</span></code></pre></div>
<pre><code>## # A tibble: 10 × 7
##    ENSEMBL         baseMean log2FoldChange lfcSE   stat pvalue   padj
##    &lt;chr&gt;              &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
##  1 ENSG00000223972    0.165          0.964 4.08   0.236  0.813 NA    
##  2 ENSG00000239906    0.229         -0.959 4.08  -0.235  0.814 NA    
##  3 ENSG00000233653    0.165          0.964 4.08   0.236  0.813 NA    
##  4 ENSG00000225972   23.2           -0.119 0.543 -0.220  0.826  0.902
##  5 ENSG00000268663    0.127         -0.959 4.08  -0.235  0.814 NA    
##  6 ENSG00000229905    0.229         -0.959 4.08  -0.235  0.814 NA    
##  7 ENSG00000230368    0.173          0.964 4.08   0.236  0.813 NA    
##  8 ENSG00000283040    0.155         -0.959 4.08  -0.235  0.814 NA    
##  9 ENSG00000272438    0.160          0.964 4.08   0.236  0.813 NA    
## 10 ENSG00000260179    3.67          -0.230 1.12  -0.206  0.837 NA</code></pre>
<div class="sourceCode" id="cb193"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb193-1"><a href="sec-rnaseq.html#cb193-1" tabindex="-1"></a>res_tbl <span class="sc">%&gt;%</span></span>
<span id="cb193-2"><a href="sec-rnaseq.html#cb193-2" tabindex="-1"></a>  <span class="fu">filter</span>(pvalue <span class="sc">&gt;</span> <span class="fl">0.8</span> <span class="sc">&amp;</span> pvalue <span class="sc">&lt;</span> <span class="fl">0.85</span>) <span class="sc">%&gt;%</span></span>
<span id="cb193-3"><a href="sec-rnaseq.html#cb193-3" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(padj) <span class="sc">%&gt;%</span></span>
<span id="cb193-4"><a href="sec-rnaseq.html#cb193-4" tabindex="-1"></a>  <span class="fu">summary</span>()</span></code></pre></div>
<pre><code>##       padj      
##  Min.   :0.887  
##  1st Qu.:0.893  
##  Median :0.900  
##  Mean   :0.901  
##  3rd Qu.:0.909  
##  Max.   :0.918  
##  NA's   :4929</code></pre>
<p>We observe that most of the genes that have a pvalue around 0.8
(corresponding to genes belonging to the second peak) have a pvalue,
but no padjusted value. These are hence genes that were filtered out
by the independent filtering process. This means that DESeq2
calculates pvalue for every gene, but the pvalue adjustement is computed
later only, on genes that pass the independent filtering threshold.
If we plot again the pvalues histogram using filtered genes only,
then we observe that the pvalues behaviour is now much nicer!</p>
<div class="sourceCode" id="cb195"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb195-1"><a href="sec-rnaseq.html#cb195-1" tabindex="-1"></a>res_tbl <span class="sc">%&gt;%</span></span>
<span id="cb195-2"><a href="sec-rnaseq.html#cb195-2" tabindex="-1"></a>    <span class="fu">filter</span>(baseMean <span class="sc">&gt;</span> <span class="fu">metadata</span>(res)<span class="sc">$</span>filterThreshold) <span class="sc">%&gt;%</span></span>
<span id="cb195-3"><a href="sec-rnaseq.html#cb195-3" tabindex="-1"></a>    <span class="fu">pull</span>(pvalue) <span class="sc">%&gt;%</span></span>
<span id="cb195-4"><a href="sec-rnaseq.html#cb195-4" tabindex="-1"></a>    <span class="fu">hist</span>()</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-108-1.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="ma-plot" class="section level3" number="4.2.9">
<h3>
<span class="header-section-number">4.2.9</span> MA plot<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('ma-plot')" onmouseout="reset_tooltip('ma-plot-tooltip')"><span class="tooltiptext" id="ma-plot-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Another interesting plot is the MA-plot (“Mean-Average” plot), a
scatter plot of log2FC versus the mean of normalised counts. Genes
with a padjusted value lower than 0.05 are colored. The plot
highlights the fact that genes with low read counts show substantially
higher variability than highly expressed genes, resulting in a strong
log2FC even though are likely not really differentially expressed. In
the MA-plot, we hope to observe some genes located in the upper/lower
right quadrant of the plots (the most interesting candidates).</p>
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb196-1"><a href="sec-rnaseq.html#cb196-1" tabindex="-1"></a><span class="fu">plotMA</span>(res)</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-109-1.png" width="672"></p>
</div>
<div id="volcano-plot" class="section level3" number="4.2.10">
<h3>
<span class="header-section-number">4.2.10</span> Volcano plot<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('volcano-plot')" onmouseout="reset_tooltip('volcano-plot-tooltip')"><span class="tooltiptext" id="volcano-plot-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Drawing a volcano-plot is also informative to have a global view on the results.
Recall that the most interesting features are those towards the top corners given
that they have small pvalues and large fold-changes.</p>
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb197-1"><a href="sec-rnaseq.html#cb197-1" tabindex="-1"></a>res_tbl <span class="sc">%&gt;%</span></span>
<span id="cb197-2"><a href="sec-rnaseq.html#cb197-2" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(padj)) <span class="sc">%&gt;%</span></span>
<span id="cb197-3"><a href="sec-rnaseq.html#cb197-3" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> log2FoldChange, <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(padj),</span>
<span id="cb197-4"><a href="sec-rnaseq.html#cb197-4" tabindex="-1"></a>             <span class="at">color =</span> padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> <span class="fu">abs</span>(log2FoldChange) <span class="sc">&gt;</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb197-5"><a href="sec-rnaseq.html#cb197-5" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"gray"</span>, <span class="st">"red"</span>)) <span class="sc">+</span></span>
<span id="cb197-6"><a href="sec-rnaseq.html#cb197-6" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb197-7"><a href="sec-rnaseq.html#cb197-7" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>)) <span class="sc">+</span></span>
<span id="cb197-8"><a href="sec-rnaseq.html#cb197-8" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb197-9"><a href="sec-rnaseq.html#cb197-9" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb197-10"><a href="sec-rnaseq.html#cb197-10" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-110-1.png" width="672"></p>
</div>
<div id="count-plots" class="section level3" number="4.2.11">
<h3>
<span class="header-section-number">4.2.11</span> Count plots<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('count-plots')" onmouseout="reset_tooltip('count-plots-tooltip')"><span class="tooltiptext" id="count-plots-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>It may be also useful to check the validity of the analysis by simply assessing the expression
levels of the most highly differentially expressed genes.</p>
<div class="sourceCode" id="cb198"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb198-1"><a href="sec-rnaseq.html#cb198-1" tabindex="-1"></a>best_genes <span class="ot">&lt;-</span> res_tbl <span class="sc">%&gt;%</span></span>
<span id="cb198-2"><a href="sec-rnaseq.html#cb198-2" tabindex="-1"></a>  <span class="fu">arrange</span>(padj)  <span class="sc">%&gt;%</span></span>
<span id="cb198-3"><a href="sec-rnaseq.html#cb198-3" tabindex="-1"></a>  <span class="fu">head</span>(<span class="dv">6</span>)</span>
<span id="cb198-4"><a href="sec-rnaseq.html#cb198-4" tabindex="-1"></a></span>
<span id="cb198-5"><a href="sec-rnaseq.html#cb198-5" tabindex="-1"></a><span class="fu">as_tibble</span>(<span class="fu">counts</span>(dds[best_genes<span class="sc">$</span>ENSEMBL, ], <span class="at">normalize =</span> <span class="cn">TRUE</span>),</span>
<span id="cb198-6"><a href="sec-rnaseq.html#cb198-6" tabindex="-1"></a>          <span class="at">rownames =</span> <span class="st">'ENSEMBL'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb198-7"><a href="sec-rnaseq.html#cb198-7" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">names_to =</span> <span class="st">"sample"</span>, <span class="at">values_to =</span> <span class="st">"counts"</span>, <span class="sc">-</span>ENSEMBL) <span class="sc">%&gt;%</span>  </span>
<span id="cb198-8"><a href="sec-rnaseq.html#cb198-8" tabindex="-1"></a>  <span class="fu">left_join</span>(<span class="fu">as_tibble</span>(coldata, <span class="at">rownames =</span> <span class="st">"sample"</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb198-9"><a href="sec-rnaseq.html#cb198-9" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> sample, <span class="at">y =</span> counts, <span class="at">fill =</span> Condition)) <span class="sc">+</span></span>
<span id="cb198-10"><a href="sec-rnaseq.html#cb198-10" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">'identity'</span>, <span class="at">color =</span> <span class="st">"gray30"</span>) <span class="sc">+</span></span>
<span id="cb198-11"><a href="sec-rnaseq.html#cb198-11" tabindex="-1"></a>  <span class="fu">facet_wrap</span>( <span class="sc">~</span> ENSEMBL, <span class="at">scales =</span> <span class="st">"free"</span>, <span class="at">ncol =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb198-12"><a href="sec-rnaseq.html#cb198-12" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">7</span>, <span class="at">angle =</span> <span class="dv">90</span>),</span>
<span id="cb198-13"><a href="sec-rnaseq.html#cb198-13" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb198-14"><a href="sec-rnaseq.html#cb198-14" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">"right"</span>,</span>
<span id="cb198-15"><a href="sec-rnaseq.html#cb198-15" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">7</span>),</span>
<span id="cb198-16"><a href="sec-rnaseq.html#cb198-16" tabindex="-1"></a>        <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">7</span>))</span></code></pre></div>
<pre><code>## Joining with `by = join_by(sample)`</code></pre>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-111-1.png" width="672"></p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ol style="list-style-type: decimal">
<li>Identify and inspect counts of the genes plotted in red in the
following volcano-plot. These genes have a very large log2FC
(|log2FC| &gt; 5) but are far from bearing the lowest padjusted value
(their padjusted value is between 0.05 and 1e-5).</li>
</ol>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-112-1.png" width="672"></p>
<ol start="2" style="list-style-type: decimal">
<li><p>Compare counts with previous counts plots (counts of genes with
lowest pvalues). what is the most striking?</p></li>
<li><p>Using the <code>dispersions()</code> function, compare dispersion values for
both group of genes</p></li>
</ol>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-16" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-16', 'sol-start-16')"></span>
</p>
<div id="sol-body-16" class="solution-body" style="display: none;">
<ol style="list-style-type: decimal">
<li>Identify and inspect counts of the genes plotted in red in the
following volcano-plot. These genes have a very large log2FC
(|log2FC| &gt; 5) but are far from bearing the lowest padjusted value
(their padjusted value is between 0.05 and 1e-5).</li>
</ol>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb200-1"><a href="sec-rnaseq.html#cb200-1" tabindex="-1"></a>selected_genes <span class="ot">&lt;-</span> res_tbl <span class="sc">%&gt;%</span></span>
<span id="cb200-2"><a href="sec-rnaseq.html#cb200-2" tabindex="-1"></a>  <span class="fu">filter</span>(padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> padj <span class="sc">&gt;</span> <span class="fl">1e-5</span> <span class="sc">&amp;</span> <span class="fu">abs</span>(log2FoldChange) <span class="sc">&gt;</span> <span class="dv">5</span>)</span>
<span id="cb200-3"><a href="sec-rnaseq.html#cb200-3" tabindex="-1"></a></span>
<span id="cb200-4"><a href="sec-rnaseq.html#cb200-4" tabindex="-1"></a><span class="fu">as_tibble</span>(<span class="fu">counts</span>(dds[selected_genes<span class="sc">$</span>ENSEMBL, ], <span class="at">normalize =</span> <span class="cn">TRUE</span>),</span>
<span id="cb200-5"><a href="sec-rnaseq.html#cb200-5" tabindex="-1"></a>          <span class="at">rownames =</span> <span class="st">'ENSEMBL'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb200-6"><a href="sec-rnaseq.html#cb200-6" tabindex="-1"></a>  <span class="fu">gather</span>(sample, counts, <span class="sc">-</span>ENSEMBL) <span class="sc">%&gt;%</span></span>
<span id="cb200-7"><a href="sec-rnaseq.html#cb200-7" tabindex="-1"></a>  <span class="fu">left_join</span>(<span class="fu">as_tibble</span>(coldata, <span class="at">rownames =</span> <span class="st">"sample"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb200-8"><a href="sec-rnaseq.html#cb200-8" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> sample, <span class="at">y =</span> counts, <span class="at">fill =</span> Condition)) <span class="sc">+</span></span>
<span id="cb200-9"><a href="sec-rnaseq.html#cb200-9" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">'identity'</span>, <span class="at">color =</span> <span class="st">"gray30"</span>) <span class="sc">+</span></span>
<span id="cb200-10"><a href="sec-rnaseq.html#cb200-10" tabindex="-1"></a>  <span class="fu">facet_wrap</span>( <span class="sc">~</span> ENSEMBL, <span class="at">scales =</span> <span class="st">"free"</span>, <span class="at">ncol =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb200-11"><a href="sec-rnaseq.html#cb200-11" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">7</span>, <span class="at">angle =</span> <span class="dv">90</span>),</span>
<span id="cb200-12"><a href="sec-rnaseq.html#cb200-12" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb200-13"><a href="sec-rnaseq.html#cb200-13" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">"right"</span>,</span>
<span id="cb200-14"><a href="sec-rnaseq.html#cb200-14" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">7</span>),</span>
<span id="cb200-15"><a href="sec-rnaseq.html#cb200-15" tabindex="-1"></a>        <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">7</span>))</span></code></pre></div>
<pre><code>## Joining with `by = join_by(sample)`</code></pre>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-113-1.png" width="672"></p>
<ol start="2" style="list-style-type: decimal">
<li>Compare counts with previous counts plots (counts of genes with lowest pvalues).
what is the most striking?</li>
</ol>
<p>These genes have very low counts. Even if they show a strong log2FC, their variability is very high.
This will result in higher dispersion values, and larger pvalues.</p>
<ol start="3" style="list-style-type: decimal">
<li>Using <code>dispersions()</code> function, compare dispersion values for both
group of genes</li>
</ol>
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb202-1"><a href="sec-rnaseq.html#cb202-1" tabindex="-1"></a><span class="fu">dispersions</span>(dds[best_genes<span class="sc">$</span>ENSEMBL,])</span></code></pre></div>
<pre><code>## [1] 0.01718122 0.02025994 0.01427782 0.04048283 0.01433817 0.01637945</code></pre>
<div class="sourceCode" id="cb204"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb204-1"><a href="sec-rnaseq.html#cb204-1" tabindex="-1"></a><span class="fu">dispersions</span>(dds[selected_genes<span class="sc">$</span>ENSEMBL,])</span></code></pre></div>
<pre><code>## [1] 0.4404912 0.3507403 0.3052325 0.3471712 0.4573378</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="logfoldchange-shrinkage" class="section level3" number="4.2.12">
<h3>
<span class="header-section-number">4.2.12</span> Logfoldchange shrinkage<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('logfoldchange-shrinkage')" onmouseout="reset_tooltip('logfoldchange-shrinkage-tooltip')"><span class="tooltiptext" id="logfoldchange-shrinkage-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>The fold-changes values might sometimes be used for specific downstream analyses
(these values could for example be used for gene ranking or visualization).
But the Log2FC estimates provided by DESeq2 should not be used directly,
as some genes might be associated to large fold-changes reflecting a high
variability rather than true strong differences.</p>
<p>To overcome this problem, DESeq2 provides a shrinkage function called
<code>lfcShrink()</code>, to generate more accurate Log2foldchange estimates.
The function will shrink the fold-change values in a way that will depend
on counts and on gene dispersion values<label for="tufte-sn-10" class="margin-toggle sidenote-number">10</label><input type="checkbox" id="tufte-sn-10" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">10</span> The shrinkage is based on bayesian approach. See the paper
describing thedefault shrinkage estimator for more details <span class="citation">(<label for="tufte-mn-10" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-10" class="margin-toggle">Zhu, Ibrahim, and Love 2019<span class="marginnote">Zhu, A, G J Ibrahim, and M I Love. 2019. <span>“Heavy-Tailed Prior Distributions for Sequence Count Data: Removing the Noise and Preserving Large Differences.”</span> <em>Bioinformatics</em> 35 (1): 2084–92. <a href="https://doi.org/10.1093/bioinformatics/bty895">https://doi.org/10.1093/bioinformatics/bty895</a>.</span>)</span></span>. As seen in the following
MA-plots, the shrinkage is much stronger for lowly than for highly expressed genes.</p>
<p>This procedure improves interpretability of the fold-changes estimates.</p>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="sec-rnaseq.html#cb206-1" tabindex="-1"></a>res_shr <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds, <span class="at">coef =</span> <span class="st">"Condition_KD_vs_mock"</span>)</span>
<span id="cb206-2"><a href="sec-rnaseq.html#cb206-2" tabindex="-1"></a>res_shr_tbl <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(res_shr, <span class="at">rownames =</span> <span class="st">"ENSEMBL"</span>)</span>
<span id="cb206-3"><a href="sec-rnaseq.html#cb206-3" tabindex="-1"></a></span>
<span id="cb206-4"><a href="sec-rnaseq.html#cb206-4" tabindex="-1"></a>fig1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(res_tbl, <span class="fu">aes</span>(<span class="at">x =</span> baseMean, <span class="at">y =</span> log2FoldChange)) <span class="sc">+</span></span>
<span id="cb206-5"><a href="sec-rnaseq.html#cb206-5" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> res_tbl, <span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb206-6"><a href="sec-rnaseq.html#cb206-6" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb206-7"><a href="sec-rnaseq.html#cb206-7" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb206-8"><a href="sec-rnaseq.html#cb206-8" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"MA plot with original logFC values"</span>)</span>
<span id="cb206-9"><a href="sec-rnaseq.html#cb206-9" tabindex="-1"></a></span>
<span id="cb206-10"><a href="sec-rnaseq.html#cb206-10" tabindex="-1"></a>fig2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(res_shr_tbl, <span class="fu">aes</span>(<span class="at">x =</span> baseMean, <span class="at">y =</span> log2FoldChange)) <span class="sc">+</span></span>
<span id="cb206-11"><a href="sec-rnaseq.html#cb206-11" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> res_shr_tbl, <span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb206-12"><a href="sec-rnaseq.html#cb206-12" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb206-13"><a href="sec-rnaseq.html#cb206-13" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb206-14"><a href="sec-rnaseq.html#cb206-14" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"shrunked log2FoldChange"</span>) <span class="sc">+</span></span>
<span id="cb206-15"><a href="sec-rnaseq.html#cb206-15" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"MA plot with shrunked logFC values"</span>)</span>
<span id="cb206-16"><a href="sec-rnaseq.html#cb206-16" tabindex="-1"></a></span>
<span id="cb206-17"><a href="sec-rnaseq.html#cb206-17" tabindex="-1"></a>fig1 <span class="sc">|</span> fig2</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-115-1.png" width="672"></p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Genes ENSG00000183055 and ENSG00000161082, highlighted in the following MA
plots, have similar expression levels and log2FoldChanges before shrinkage.</p>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-116-1.png" width="672"></p>
<ol style="list-style-type: decimal">
<li><p>Compare DESeq2 results for these 2 genes before and after logFC shrinkage.
What was the impact on their log2FoldChange values? on their p-values?</p></li>
<li><p>Try to figure out why one of the 2 genes is more affected by the shrinkage
procedure.</p></li>
</ol>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-17" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-17', 'sol-start-17')"></span>
</p>
<div id="sol-body-17" class="solution-body" style="display: none;">
<ol style="list-style-type: decimal">
<li>Compare DESeq2 results for these 2 genes before and after logFC shrinkage.
What was the impact on their log2FoldChange values? on their p-values?</li>
</ol>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="sec-rnaseq.html#cb207-1" tabindex="-1"></a>res_tbl <span class="sc">%&gt;%</span> </span>
<span id="cb207-2"><a href="sec-rnaseq.html#cb207-2" tabindex="-1"></a>  <span class="fu">filter</span>(ENSEMBL <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"ENSG00000183055"</span>, <span class="st">"ENSG00000161082"</span>))</span></code></pre></div>
<pre><code>## # A tibble: 2 × 7
##   ENSEMBL         baseMean log2FoldChange lfcSE  stat    pvalue     padj
##   &lt;chr&gt;              &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1 ENSG00000183055     24.4           1.92 0.695  2.76 0.00572   0.0253  
## 2 ENSG00000161082     23.3           1.94 0.477  4.06 0.0000487 0.000481</code></pre>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb209-1"><a href="sec-rnaseq.html#cb209-1" tabindex="-1"></a>res_shr_tbl <span class="sc">%&gt;%</span> </span>
<span id="cb209-2"><a href="sec-rnaseq.html#cb209-2" tabindex="-1"></a>  <span class="fu">filter</span>(ENSEMBL <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"ENSG00000183055"</span>, <span class="st">"ENSG00000161082"</span>))</span></code></pre></div>
<pre><code>## # A tibble: 2 × 6
##   ENSEMBL         baseMean log2FoldChange lfcSE    pvalue     padj
##   &lt;chr&gt;              &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1 ENSG00000183055     24.4           1.25 0.906 0.00572   0.0253  
## 2 ENSG00000161082     23.3           1.69 0.507 0.0000487 0.000481</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Try to figure out why one of the 2 genes is more affected by the shrinkage
procedure.</li>
</ol>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="sec-rnaseq.html#cb211-1" tabindex="-1"></a><span class="fu">as_tibble</span>(<span class="fu">counts</span>(dds[<span class="fu">c</span>(<span class="st">"ENSG00000183055"</span>, <span class="st">"ENSG00000161082"</span>), ], <span class="at">normalize =</span> <span class="cn">TRUE</span>),</span>
<span id="cb211-2"><a href="sec-rnaseq.html#cb211-2" tabindex="-1"></a>          <span class="at">rownames =</span> <span class="st">'ENSEMBL'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb211-3"><a href="sec-rnaseq.html#cb211-3" tabindex="-1"></a>  <span class="fu">gather</span>(sample, counts, <span class="sc">-</span>ENSEMBL) <span class="sc">%&gt;%</span></span>
<span id="cb211-4"><a href="sec-rnaseq.html#cb211-4" tabindex="-1"></a>  <span class="fu">left_join</span>(<span class="fu">as_tibble</span>(coldata, <span class="at">rownames =</span> <span class="st">"sample"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb211-5"><a href="sec-rnaseq.html#cb211-5" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> sample, <span class="at">y =</span> counts, <span class="at">fill =</span> Condition)) <span class="sc">+</span></span>
<span id="cb211-6"><a href="sec-rnaseq.html#cb211-6" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">'identity'</span>, <span class="at">color =</span> <span class="st">"gray30"</span>) <span class="sc">+</span></span>
<span id="cb211-7"><a href="sec-rnaseq.html#cb211-7" tabindex="-1"></a>  <span class="fu">facet_wrap</span>( <span class="sc">~</span> ENSEMBL, <span class="at">scales =</span> <span class="st">"free"</span>, <span class="at">ncol =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb211-8"><a href="sec-rnaseq.html#cb211-8" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">7</span>, <span class="at">angle =</span> <span class="dv">90</span>),</span>
<span id="cb211-9"><a href="sec-rnaseq.html#cb211-9" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb211-10"><a href="sec-rnaseq.html#cb211-10" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">"right"</span>,</span>
<span id="cb211-11"><a href="sec-rnaseq.html#cb211-11" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">7</span>),</span>
<span id="cb211-12"><a href="sec-rnaseq.html#cb211-12" tabindex="-1"></a>        <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">7</span>))</span></code></pre></div>
<pre><code>## Joining with `by = join_by(sample)`</code></pre>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-118-1.png" width="672"></p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb213-1"><a href="sec-rnaseq.html#cb213-1" tabindex="-1"></a><span class="fu">dispersions</span>(dds[<span class="fu">c</span>(<span class="st">"ENSG00000183055"</span>, <span class="st">"ENSG00000161082"</span>),])</span></code></pre></div>
<pre><code>## [1] 0.28557090 0.09884484</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="adding-gene-annotations" class="section level3" number="4.2.13">
<h3>
<span class="header-section-number">4.2.13</span> Adding gene annotations<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('adding-gene-annotations')" onmouseout="reset_tooltip('adding-gene-annotations-tooltip')"><span class="tooltiptext" id="adding-gene-annotations-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>The result table only uses Ensembl gene IDs, but gene names may be
more informative.</p>
<p><code>AnnotationDbi</code> is an R package that provides an interface for connecting
and querying various annotation databases.</p>
<p>There are a many organism-specific <code>orgDb</code> packages, such as <code>org.Hs.eg.db</code>
for human and <code>org.Mm.eg.db</code> for mouse.</p>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb215-1"><a href="sec-rnaseq.html#cb215-1" tabindex="-1"></a><span class="co">#BiocManager::install(c("AnnotationDbi", "org.Hs.eg.db"))</span></span>
<span id="cb215-2"><a href="sec-rnaseq.html#cb215-2" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"AnnotationDbi"</span>)</span>
<span id="cb215-3"><a href="sec-rnaseq.html#cb215-3" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"org.Hs.eg.db"</span>)</span></code></pre></div>
<p>The <code>keytypes()</code> function can help us to see what information we can extract
from the database (ENSEMBL gene ID, gene symbol, gene name, …)</p>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="sec-rnaseq.html#cb216-1" tabindex="-1"></a><span class="fu">keytypes</span>(org.Hs.eg.db)</span></code></pre></div>
<pre><code>##  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
##  [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"    
## [11] "GENETYPE"     "GO"           "GOALL"        "IPI"          "MAP"         
## [16] "OMIM"         "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"        
## [21] "PMID"         "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"      
## [26] "UNIPROT"</code></pre>
<p>We can easily extract information from this database using the <code>mapIds()</code> function.</p>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="sec-rnaseq.html#cb218-1" tabindex="-1"></a><span class="co"># Retrieving the gene symbol of our "best genes"</span></span>
<span id="cb218-2"><a href="sec-rnaseq.html#cb218-2" tabindex="-1"></a><span class="fu">mapIds</span>(<span class="at">x =</span> org.Hs.eg.db,               </span>
<span id="cb218-3"><a href="sec-rnaseq.html#cb218-3" tabindex="-1"></a>       <span class="at">keys =</span> selected_genes<span class="sc">$</span>ENSEMBL,  </span>
<span id="cb218-4"><a href="sec-rnaseq.html#cb218-4" tabindex="-1"></a>       <span class="at">column =</span> <span class="st">"SYMBOL"</span>,              </span>
<span id="cb218-5"><a href="sec-rnaseq.html#cb218-5" tabindex="-1"></a>       <span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>)</span></code></pre></div>
<pre><code>## 'select()' returned 1:1 mapping between keys and columns</code></pre>
<pre><code>## ENSG00000127129 ENSG00000112499 ENSG00000171004 ENSG00000257859 ENSG00000157368 
##          "EDN2"       "SLC22A2"        "HS6ST2"        "CASC18"          "IL34"</code></pre>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb221-1"><a href="sec-rnaseq.html#cb221-1" tabindex="-1"></a><span class="co"># Retrieving the ENTREZID of our "best genes"</span></span>
<span id="cb221-2"><a href="sec-rnaseq.html#cb221-2" tabindex="-1"></a><span class="fu">mapIds</span>(<span class="at">x =</span> org.Hs.eg.db, </span>
<span id="cb221-3"><a href="sec-rnaseq.html#cb221-3" tabindex="-1"></a>       <span class="at">keys =</span> selected_genes<span class="sc">$</span>ENSEMBL,</span>
<span id="cb221-4"><a href="sec-rnaseq.html#cb221-4" tabindex="-1"></a>       <span class="at">column =</span> <span class="st">"ENTREZID"</span>, </span>
<span id="cb221-5"><a href="sec-rnaseq.html#cb221-5" tabindex="-1"></a>       <span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>)</span></code></pre></div>
<pre><code>## 'select()' returned 1:1 mapping between keys and columns</code></pre>
<pre><code>## ENSG00000127129 ENSG00000112499 ENSG00000171004 ENSG00000257859 ENSG00000157368 
##          "1907"          "6582"         "90161"     "101929110"        "146433"</code></pre>
<p>Let’s add gene symbols and ENTREZID in our table of results.</p>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="sec-rnaseq.html#cb224-1" tabindex="-1"></a>res_tbl <span class="ot">&lt;-</span> res_tbl <span class="sc">%&gt;%</span> </span>
<span id="cb224-2"><a href="sec-rnaseq.html#cb224-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene =</span> <span class="fu">mapIds</span>(org.Hs.eg.db, ENSEMBL, <span class="st">"SYMBOL"</span>, <span class="st">"ENSEMBL"</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb224-3"><a href="sec-rnaseq.html#cb224-3" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">ENTREZID =</span> <span class="fu">mapIds</span>(org.Hs.eg.db, ENSEMBL, <span class="st">"ENTREZID"</span>, <span class="st">"ENSEMBL"</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb224-4"><a href="sec-rnaseq.html#cb224-4" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(ENSEMBL, gene, ENTREZID, <span class="fu">everything</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb224-5"><a href="sec-rnaseq.html#cb224-5" tabindex="-1"></a>  <span class="fu">arrange</span>(padj)</span></code></pre></div>
<pre><code>## 'select()' returned 1:many mapping between keys and columns
## 'select()' returned 1:many mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb226-1"><a href="sec-rnaseq.html#cb226-1" tabindex="-1"></a>res_tbl</span></code></pre></div>
<pre><code>## # A tibble: 32,102 × 9
##    ENSEMBL         gene   ENTREZID baseMean log2FoldChange lfcSE  stat    pvalue
##    &lt;chr&gt;           &lt;chr&gt;  &lt;chr&gt;       &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;
##  1 ENSG00000101856 PGRMC1 10857       1209.          -4.67 0.178 -26.2 8.24e-152
##  2 ENSG00000177494 ZBED2  79413        277.           6.25 0.351  17.8 6.25e- 71
##  3 ENSG00000136159 NUDT15 55270        630.           2.24 0.154  14.6 3.32e- 48
##  4 ENSG00000175315 CST6   1474         220.           4.29 0.300  14.3 2.20e- 46
##  5 ENSG00000128245 YWHAH  7533        2692.           2.05 0.144  14.2 5.64e- 46
##  6 ENSG00000113272 THG1L  54974        381.           2.33 0.171  13.6 3.45e- 42
##  7 ENSG00000213853 EMP2   2013         506.           2.53 0.191  13.2 8.45e- 40
##  8 ENSG00000115107 STEAP3 55240        193.           3.44 0.271  12.7 7.18e- 37
##  9 ENSG00000008513 ST3GA… 6482        1419.           3.00 0.240  12.5 6.39e- 36
## 10 ENSG00000105855 ITGB8  3696        6134.           2.18 0.178  12.2 2.26e- 34
## # ℹ 32,092 more rows
## # ℹ 1 more variable: padj &lt;dbl&gt;</code></pre>
<p>Below, we store these results and the <code>DESeqDataSet</code> object for the
next chapter.</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="sec-rnaseq.html#cb228-1" tabindex="-1"></a><span class="fu">saveRDS</span>(dds, <span class="at">file =</span> <span class="st">"./data/dds.rds"</span>)</span>
<span id="cb228-2"><a href="sec-rnaseq.html#cb228-2" tabindex="-1"></a><span class="fu">saveRDS</span>(res_tbl, <span class="at">file =</span> <span class="st">"./data/res_tbl.rds"</span>)</span></code></pre></div>
</div>
<div id="further-links" class="section level3" number="4.2.14">
<h3>
<span class="header-section-number">4.2.14</span> Further links<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('further-links')" onmouseout="reset_tooltip('further-links-tooltip')"><span class="tooltiptext" id="further-links-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<ul>
<li><p><a href="https://bioconductor.org/packages/release/bioc/html/iSEE.html">iSEE</a> is a
Bioconductor package that provides an interactive Shiny-based graphical user
interface for exploring data stored in SummarizedExperiment objects.</p></li>
<li><p><a href="https://www.bioconductor.org/packages/release/bioc/html/Glimma.html">Glimma</a>
is a Bioconductor package that generates interactive visualisations for analysis
of RNA-sequencing data using output from limma, edgeR or DESeq2 packages in an
HTML page. The interactions are built on top of the popular static representations
of analysis results in order to provide additional information.</p></li>
</ul>
</div>
</div>
</div>
<p style="text-align: center;">
<a href="sec-hts.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-design.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2024-11-01
 using 
R version 4.4.1 (2024-06-14)
</p>
</div>
</div>



</body>
</html>
