<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 6 Enrichment analyses | Omics Data Analysis" />
<meta property="og:type" content="book" />

<meta property="og:description" content="Course material for the Omics Data Analysis (WSBIM2122) course at UCLouvain." />


<meta name="author" content="Laurent Gatto" />

<meta name="date" content="2024-11-08" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="Course material for the Omics Data Analysis (WSBIM2122) course at UCLouvain.">

<title>Chapter 6 Enrichment analyses | Omics Data Analysis</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>
<script>
function copy_link(id) {
  var dummy = document.createElement('input'),
  text = window.location.href.split(/[?#]/)[0] + '#' + id;
  document.body.appendChild(dummy);
  dummy.value = text;
  dummy.select();
  document.execCommand('copy');
  document.body.removeChild(dummy);
  
  var tooltip = document.getElementById(id + '-tooltip');
  tooltip.innerHTML = 'Copied!';
}

function reset_tooltip(id) {
  var tooltip = document.getElementById(id);
  tooltip.innerHTML = 'Copy link';
}
</script>
<link href="libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="libs/datatables-binding-0.33/datatables.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>
<script src="libs/plotly-binding-4.10.4/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-2.11.1/plotly-latest.min.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }

code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">Omics Data Analysis<p><p class="author">Laurent Gatto</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html" id="toc-preamble">Preamble</a>
<a href="sec-rmd.html" id="toc-sec-rmd"><span class="toc-section-number">1</span> R markdown</a>
<a href="sec-linmod.html" id="toc-sec-linmod"><span class="toc-section-number">2</span> Linear models</a>
<a href="sec-hts.html" id="toc-sec-hts"><span class="toc-section-number">3</span> High throughput sequencing</a>
<a href="sec-rnaseq.html" id="toc-sec-rnaseq"><span class="toc-section-number">4</span> Differential expression analysis</a>
<a href="sec-design.html" id="toc-sec-design"><span class="toc-section-number">5</span> More complex experimental design</a>
<a id="active-page" href="sec-gsea.html" id="toc-sec-gsea"><span class="toc-section-number">6</span> Enrichment analyses</a><ul class="toc-sections">
<li class="toc"><a href="#introduction"> Introduction</a></li>
<li class="toc"><a href="#gene-sets"> Gene sets</a></li>
<li class="toc"><a href="#over-representation-analysis-ora"> Over representation analysis (ORA)</a></li>
<li class="toc"><a href="#gene-set-enrichment-analysis-gsea"> Gene set enrichment analysis (GSEA)</a></li>
<li class="toc"><a href="#using-the-clusterprofiler-package"> Using the <code>clusterProfiler</code> package</a></li>
<li class="toc"><a href="#visualisation-of-enrichment-analyses"> Visualisation of enrichment analyses</a></li>
<li class="toc"><a href="#discussion"> Discussion</a></li>
</ul>
<a href="sec-ms.html" id="toc-sec-ms"><span class="toc-section-number">7</span> Mass spectrometry</a>
<a href="sec-prot.html" id="toc-sec-prot"><span class="toc-section-number">8</span> Quantitative proteomics data analysis</a>
<a href="sec-ccl.html" id="toc-sec-ccl"><span class="toc-section-number">9</span> Conclusions</a>
<a href="sec-cli.html" id="toc-sec-cli"><span class="toc-section-number">10</span> Annex: the shell</a>
<a href="sec-git.html" id="toc-sec-git"><span class="toc-section-number">11</span> Annex: Collaborating with Git and GitHub</a>
<a href="sec-info.html" id="toc-sec-info"><span class="toc-section-number">12</span> Course and session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="sec-gsea" class="section level1" number="6">
<h1>
<span class="header-section-number">Chapter 6</span> Enrichment analyses</h1>
<p><strong>Motivation</strong>: at the end of this chapter, the students will be able
to address the following points:</p>
<ul>
<li>What are the next steps after a differential expression analysis?</li>
<li>Why are enrichment analyses useful?</li>
<li>Understanding over representation analyses.</li>
<li>Understanding a gene set enrichment analysis.</li>
<li>Application of the <em><a href="https://bioconductor.org/packages/3.19/clusterProfiler">clusterProfiler</a></em> package.</li>
</ul>
<div id="introduction" class="section level2" number="6.1">
<h2>
<span class="header-section-number">6.1</span> Introduction<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('introduction')" onmouseout="reset_tooltip('introduction-tooltip')"><span class="tooltiptext" id="introduction-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<ul>
<li><p>Differential expression analysis is <strong>univariate</strong> - each gene is
tested on its own. This probably doesn’t reflect the underlying
biology - genes work in conjunction, not in isolation. One wouldn’t
expect that the effect of a drug, or a mutation, … would lead to
the perturbation of a single gene expression.</p></li>
<li><p>The univariate approach expects (or tests for) significant changes
in single genes. Moderate effects in many (related) genes cannot, by
definition be identified as statistically significant, even it such
an effect may be biologically more plausible that one of a few large
ones.</p></li>
<li><p>The goal of an enrichment analysis is to test for a group of related
genes, called <strong>gene sets</strong>, and test whether the genes within these
sets are enriched for differentially expression.</p></li>
</ul>
</div>
<div id="gene-sets" class="section level2" number="6.2">
<h2>
<span class="header-section-number">6.2</span> Gene sets<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('gene-sets')" onmouseout="reset_tooltip('gene-sets-tooltip')"><span class="tooltiptext" id="gene-sets-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>While nothing would stop a user to create their own gene sets, the
sets that are generally used stem from community-maintained resources.</p>
<div id="gene-ontology-go" class="section level3" number="6.2.1">
<h3>
<span class="header-section-number">6.2.1</span> Gene Ontology (GO)<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('gene-ontology-go')" onmouseout="reset_tooltip('gene-ontology-go-tooltip')"><span class="tooltiptext" id="gene-ontology-go-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>The <a href="http://geneontology.org/">Gene Ontology</a> <span class="citation">(<label for="tufte-mn-11" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-11" class="margin-toggle">Ashburner et al. 2000<span class="marginnote">Ashburner, M, C A Ball, J A Blake, D Botstein, H Butler, J M Cherry, A P Davis, et al. 2000. <span>“Gene Ontology: Tool for the Unification of Biology. The Gene Ontology Consortium.”</span> <em>Nat Genet</em> 25 (1): 25–29. <a href="https://doi.org/10.1038/75556">https://doi.org/10.1038/75556</a>.</span>)</span>
defines GO terms. These terms are based on a controlled vocabulary (to
resuse the same words in a consistent way) and relations<label for="tufte-sn-11" class="margin-toggle sidenote-number">11</label><input type="checkbox" id="tufte-sn-11" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">11</span> Example of relations between terms are, for example, <em>is_a</em>
(mitosis <em>is_a</em> cell cycle phase) or <em>part_of</em> (mitosis <em>part_of</em>
M phase of mitotic cell cycle).</span> that
define the directed links between terms. These relations define a
hierarchy between GO terms: all term can be represented as a (direct
acyclic) graph (DAG). There exist thus very general and very specific
terms.</p>
<p>These terms are classed into three categories, called <em>namespaces</em>:</p>
<ul>
<li>Molecular Function (MF): molecular activities of gene products</li>
<li>Cellular Component (CC): where gene products are active</li>
<li>Biological Process (BP): pathways and larger processes made up of
the activities of multiple gene products</li>
</ul>
<p>Here’s an example of GO term <code>GO:0007155</code>, that describes cell adhesion.</p>
<pre><code>A Term from the GO ontology: GO:0007155
 Label: cell adhesion
  The attachment of a cell, either to another cell or to an underlying
  substrate such as the extracellular matrix, via cell adhesion
  molecules.</code></pre>
<p>If you look at the GO term <code>GO:0007155</code> <a href="http://amigo.geneontology.org/amigo/term/GO:0007155">entry on the Gene Ontology
page</a>, you can
find more details and, if you scroll down, example genes that are
annotated with that term.</p>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:unnamed-chunk-142"></span>
<p class="caption marginnote shownote">
Figure 6.1: Gene ontology entry for the GO term for cell adhesion.
</p>
<img src="figs/go-screenshot.png" alt="Gene ontology entry for the GO term for cell adhesion." width="100%">
</div>
<p>The whole Gene Ontology is can be accessed in R with the
<em><a href="https://bioconductor.org/packages/3.19/GO.db">GO.db</a></em> package.</p>
<p>In the code chunk below, we query the <code>GO.db</code> package through the
<code>org.Hs.eg.db</code> interface. This <em>org.db</em>-type of packages for Homo
sapiens enables to perform various queries for human genes, such as
retrieving all gene symbols and ENTREZ identifiers (the <code>columns</code>
below) that are annotated with a GO term (the <code>keys</code> below) of
interest.</p>
<p>The GO term of interest here is focal adhesion:</p>
<pre><code>A cell-substrate junction that anchors the cell to the extracellular
matrix and that forms a point of termination of actin filaments. In
insects focal adhesion has also been referred to as hemi-adherens
junction (HAJ).</code></pre>
<div class="sourceCode" id="cb260"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb260-1"><a href="sec-gsea.html#cb260-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"GO.db"</span>)</span>
<span id="cb260-2"><a href="sec-gsea.html#cb260-2" tabindex="-1"></a>GO.db</span></code></pre></div>
<pre><code>## GODb object:
## | GOSOURCENAME: Gene Ontology
## | GOSOURCEURL: http://current.geneontology.org/ontology/go-basic.obo
## | GOSOURCEDATE: 2024-01-17
## | Db type: GODb
## | package: AnnotationDbi
## | DBSCHEMA: GO_DB
## | GOEGSOURCEDATE: 2024-Mar12
## | GOEGSOURCENAME: Entrez Gene
## | GOEGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA
## | DBSCHEMAVERSION: 2.1</code></pre>
<pre><code>## 
## Please see: help('select') for usage information</code></pre>
<div class="sourceCode" id="cb263"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb263-1"><a href="sec-gsea.html#cb263-1" tabindex="-1"></a><span class="fu">columns</span>(GO.db) <span class="do">## variables contained in GO.db</span></span></code></pre></div>
<pre><code>## [1] "DEFINITION" "GOID"       "ONTOLOGY"   "TERM"</code></pre>
<div class="sourceCode" id="cb265"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb265-1"><a href="sec-gsea.html#cb265-1" tabindex="-1"></a><span class="do">## Select the term of interest</span></span>
<span id="cb265-2"><a href="sec-gsea.html#cb265-2" tabindex="-1"></a><span class="fu">select</span>(GO.db, <span class="at">keys =</span> <span class="st">"GO:0005925"</span>,</span>
<span id="cb265-3"><a href="sec-gsea.html#cb265-3" tabindex="-1"></a>       <span class="at">columns =</span> <span class="fu">columns</span>(GO.db),</span>
<span id="cb265-4"><a href="sec-gsea.html#cb265-4" tabindex="-1"></a>       <span class="at">keytype =</span> <span class="st">"GOID"</span>)</span></code></pre></div>
<pre><code>## 'select()' returned 1:1 mapping between keys and columns</code></pre>
<pre><code>##         GOID
## 1 GO:0005925
##                                                                                                                                                                                                                   DEFINITION
## 1 A cell-substrate junction that anchors the cell to the extracellular matrix and that forms a point of termination of actin filaments. In insects focal adhesion has also been referred to as hemi-adherens junction (HAJ).
##   ONTOLOGY           TERM
## 1       CC focal adhesion</code></pre>
<p>Let’s now extract all human genes that are tagged with that GO term:</p>
<div class="sourceCode" id="cb268"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb268-1"><a href="sec-gsea.html#cb268-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"org.Hs.eg.db"</span>)</span>
<span id="cb268-2"><a href="sec-gsea.html#cb268-2" tabindex="-1"></a>GO_0005925 <span class="ot">&lt;-</span> AnnotationDbi<span class="sc">::</span><span class="fu">select</span>(org.Hs.eg.db,</span>
<span id="cb268-3"><a href="sec-gsea.html#cb268-3" tabindex="-1"></a>                                    <span class="at">keys =</span> <span class="st">"GO:0005925"</span>,</span>
<span id="cb268-4"><a href="sec-gsea.html#cb268-4" tabindex="-1"></a>                                    <span class="at">columns =</span> <span class="fu">c</span>(<span class="st">"ENTREZID"</span>, <span class="st">"SYMBOL"</span>),</span>
<span id="cb268-5"><a href="sec-gsea.html#cb268-5" tabindex="-1"></a>                                    <span class="at">keytype =</span> <span class="st">"GO"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb268-6"><a href="sec-gsea.html#cb268-6" tabindex="-1"></a>    as_tibble <span class="sc">%&gt;%</span></span>
<span id="cb268-7"><a href="sec-gsea.html#cb268-7" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">duplicated</span>(ENTREZID))</span></code></pre></div>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb270-1"><a href="sec-gsea.html#cb270-1" tabindex="-1"></a>GO_0005925</span></code></pre></div>
<pre><code>## # A tibble: 421 × 5
##    GO         EVIDENCE ONTOLOGY ENTREZID SYMBOL
##    &lt;chr&gt;      &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt; 
##  1 GO:0005925 HDA      CC       60       ACTB  
##  2 GO:0005925 HDA      CC       70       ACTC1 
##  3 GO:0005925 ISS      CC       71       ACTG1 
##  4 GO:0005925 HDA      CC       81       ACTN4 
##  5 GO:0005925 HDA      CC       87       ACTN1 
##  6 GO:0005925 IMP      CC       88       ACTN2 
##  7 GO:0005925 IMP      CC       89       ACTN3 
##  8 GO:0005925 HDA      CC       102      ADAM10
##  9 GO:0005925 HDA      CC       118      ADD1  
## 10 GO:0005925 HDA      CC       214      ALCAM 
## # ℹ 411 more rows</code></pre>
<p>We have 421 genes matching this GO term. There are
thus 421 genes in the GO_0005925 GO set.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Repeat the code above to extract the genes annotated with the
<code>GO:0005813</code> term for the centrosome:</p>
<pre><code>A structure comprised of a core structure (in most organisms, a pair
of centrioles) and peripheral material from which a
microtubule-based structure, such as a spindle apparatus, is
organized. Centrosomes occur close to the nucleus during interphase
in many eukaryotic cells, though in animal cells it changes
continually during the cell-division cycle.</code></pre>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-18" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-18', 'sol-start-18')"></span>
</p>
<div id="sol-body-18" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb273"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb273-1"><a href="sec-gsea.html#cb273-1" tabindex="-1"></a>GO_0005813 <span class="ot">&lt;-</span> AnnotationDbi<span class="sc">::</span><span class="fu">select</span>(org.Hs.eg.db,</span>
<span id="cb273-2"><a href="sec-gsea.html#cb273-2" tabindex="-1"></a>                     <span class="at">keys =</span> <span class="st">"GO:0005813"</span>,</span>
<span id="cb273-3"><a href="sec-gsea.html#cb273-3" tabindex="-1"></a>                     <span class="at">columns =</span> <span class="fu">c</span>(<span class="st">"ENTREZID"</span>, <span class="st">"SYMBOL"</span>),</span>
<span id="cb273-4"><a href="sec-gsea.html#cb273-4" tabindex="-1"></a>                     <span class="at">keytype =</span> <span class="st">"GO"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb273-5"><a href="sec-gsea.html#cb273-5" tabindex="-1"></a>    as_tibble <span class="sc">%&gt;%</span></span>
<span id="cb273-6"><a href="sec-gsea.html#cb273-6" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">duplicated</span>(ENTREZID))</span></code></pre></div>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb275"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb275-1"><a href="sec-gsea.html#cb275-1" tabindex="-1"></a>GO_0005813</span></code></pre></div>
<pre><code>## # A tibble: 637 × 5
##    GO         EVIDENCE ONTOLOGY ENTREZID SYMBOL
##    &lt;chr&gt;      &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt; 
##  1 GO:0005813 IDA      CC       35       ACADS 
##  2 GO:0005813 IDA      CC       324      APC   
##  3 GO:0005813 IDA      CC       328      APEX1 
##  4 GO:0005813 IDA      CC       402      ARL2  
##  5 GO:0005813 IDA      CC       403      ARL3  
##  6 GO:0005813 IDA      CC       468      ATF4  
##  7 GO:0005813 ISS      CC       472      ATM   
##  8 GO:0005813 IBA      CC       582      BBS1  
##  9 GO:0005813 IDA      CC       585      BBS4  
## 10 GO:0005813 IBA      CC       595      CCND1 
## # ℹ 627 more rows</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="kyoto-encyclopedia-of-genes-and-genomes-kegg" class="section level3" number="6.2.2">
<h3>
<span class="header-section-number">6.2.2</span> Kyoto Encyclopedia of Genes and Genomes (KEGG)<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('kyoto-encyclopedia-of-genes-and-genomes-kegg')" onmouseout="reset_tooltip('kyoto-encyclopedia-of-genes-and-genomes-kegg-tooltip')"><span class="tooltiptext" id="kyoto-encyclopedia-of-genes-and-genomes-kegg-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p><a href="https://www.genome.jp/kegg/pathway.html">KEGG</a> pathway is a
collection of manually drawn and curated pathway maps representing
current knowledge of the molecular interaction, reaction and relation
networks.</p>
<p>The figure below shows the <a href="https://www.genome.jp/kegg-bin/show_pathway?hsa04110">pathways for the cell
cycle</a> in
humans.</p>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:unnamed-chunk-143"></span>
<p class="caption marginnote shownote">
Figure 6.2: KEGG pathway for cell cycle.
</p>
<img src="figs/hsa04110.png" alt="KEGG pathway for cell cycle." width="100%">
</div>
<p>The <em><a href="https://bioconductor.org/packages/3.19/KEGGREST">KEGGREST</a></em> package provides a client interface to the KEGG server.</p>
</div>
<div id="reactome" class="section level3" number="6.2.3">
<h3>
<span class="header-section-number">6.2.3</span> Reactome<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('reactome')" onmouseout="reset_tooltip('reactome-tooltip')"><span class="tooltiptext" id="reactome-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Alike KEGG patway, <a href="https://reactome.org/">Reactome</a> is a free,
open-source, curated and peer-reviewed pathway database. The
Bioconductor <em><a href="https://bioconductor.org/packages/3.19/reactome.db">reactome.db</a></em> package provides
access to reactome maps and annotations within R.</p>
</div>
<div id="molecular-signatures-database-msigdb" class="section level3" number="6.2.4">
<h3>
<span class="header-section-number">6.2.4</span> Molecular Signatures Database (MSigDB)<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('molecular-signatures-database-msigdb')" onmouseout="reset_tooltip('molecular-signatures-database-msigdb-tooltip')"><span class="tooltiptext" id="molecular-signatures-database-msigdb-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>MSigDB is a collection of annotated gene sets for use with GSEA
software. The MSigDB gene sets are divided into 9 collections:</p>
<ul>
<li>Hallmark gene sets (H) are coherently expressed signatures derived
by aggregating many MSigDB gene sets to represent well-defined
biological states or processes.</li>
<li>Positional gene sets (C1) for each human chromosome and cytogenetic band.</li>
<li>Curated gene sets (C2) from online pathway databases, publications
in PubMed, and knowledge of domain experts.</li>
<li>Regulatory target gene sets (C3) based on gene target predictions
for microRNA seed sequences and predicted transcription factor
binding sites.</li>
<li>Computational gene sets (C4) defined by mining large collections of
cancer-oriented microarray data.</li>
<li>Ontology gene sets (C5) consist of genes annotated by the same
ontology term.</li>
<li>Oncogenic signature gene (C6) sets defined directly from microarray
gene expression data from cancer gene perturbations.</li>
<li>Immunologic signature gene sets (C7) defined directly from
microarray gene expression data from immunologic studies.</li>
<li>Cell type signature gene sets (C8) curated from cluster markers
identified in single-cell sequencing studies of human tissue.</li>
</ul>
<p>The <em><a href="https://CRAN.R-project.org/package=msigdbr">msigdbr</a></em> CRAN package provides the MSigDB
gene sets in a standard R data frame with key-value pairs.</p>
</div>
<div id="input-data" class="section level3" number="6.2.5">
<h3>
<span class="header-section-number">6.2.5</span> Input data<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('input-data')" onmouseout="reset_tooltip('input-data-tooltip')"><span class="tooltiptext" id="input-data-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>To illustrate enrichment analyses, we will use the DESeq2 results
stored in the <code>res_tbl</code> variable, computed in the previous
chapter.</p>
<p>We will focus on the genes that have an adjusted p-value (those that
have been tested) and that have unique ENTREZ gene identifiers.</p>
<div class="sourceCode" id="cb277"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb277-1"><a href="sec-gsea.html#cb277-1" tabindex="-1"></a>res_tbl <span class="ot">&lt;-</span> res_tbl <span class="sc">%&gt;%</span></span>
<span id="cb277-2"><a href="sec-gsea.html#cb277-2" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(ENTREZID),</span>
<span id="cb277-3"><a href="sec-gsea.html#cb277-3" tabindex="-1"></a>           <span class="sc">!</span><span class="fu">is.na</span>(padj),</span>
<span id="cb277-4"><a href="sec-gsea.html#cb277-4" tabindex="-1"></a>           <span class="sc">!</span><span class="fu">duplicated</span>(ENTREZID)) <span class="sc">%&gt;%</span></span>
<span id="cb277-5"><a href="sec-gsea.html#cb277-5" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">ENTREZID =</span> <span class="fu">as.character</span>(ENTREZID))</span>
<span id="cb277-6"><a href="sec-gsea.html#cb277-6" tabindex="-1"></a></span>
<span id="cb277-7"><a href="sec-gsea.html#cb277-7" tabindex="-1"></a>res_tbl</span></code></pre></div>
<pre><code>## # A tibble: 14,014 × 9
##    ENSEMBL         gene   ENTREZID baseMean log2FoldChange lfcSE  stat    pvalue
##    &lt;chr&gt;           &lt;chr&gt;  &lt;chr&gt;       &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;
##  1 ENSG00000101856 PGRMC1 10857       1209.          -4.67 0.178 -26.2 8.24e-152
##  2 ENSG00000177494 ZBED2  79413        277.           6.25 0.351  17.8 6.25e- 71
##  3 ENSG00000136159 NUDT15 55270        630.           2.24 0.154  14.6 3.32e- 48
##  4 ENSG00000175315 CST6   1474         220.           4.29 0.300  14.3 2.20e- 46
##  5 ENSG00000128245 YWHAH  7533        2692.           2.05 0.144  14.2 5.64e- 46
##  6 ENSG00000113272 THG1L  54974        381.           2.33 0.171  13.6 3.45e- 42
##  7 ENSG00000213853 EMP2   2013         506.           2.53 0.191  13.2 8.45e- 40
##  8 ENSG00000115107 STEAP3 55240        193.           3.44 0.271  12.7 7.18e- 37
##  9 ENSG00000008513 ST3GA… 6482        1419.           3.00 0.240  12.5 6.39e- 36
## 10 ENSG00000105855 ITGB8  3696        6134.           2.18 0.178  12.2 2.26e- 34
## # ℹ 14,004 more rows
## # ℹ 1 more variable: padj &lt;dbl&gt;</code></pre>
</div>
</div>
<div id="over-representation-analysis-ora" class="section level2" number="6.3">
<h2>
<span class="header-section-number">6.3</span> Over representation analysis (ORA)<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('over-representation-analysis-ora')" onmouseout="reset_tooltip('over-representation-analysis-ora-tooltip')"><span class="tooltiptext" id="over-representation-analysis-ora-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>An over representation analysis relies on the hypergeometric
distribution. The hypergeometric distribution is a discrete
probability distribution that describes the probability of <span class="math inline">\(k\)</span>
successes (random draws for which the object drawn has a specified
feature) in <span class="math inline">\(n\)</span> draws, without replacement, from a finite population
of size <span class="math inline">\(N\)</span> that contains exactly <span class="math inline">\(K\)</span> objects with that feature,
wherein each draw is either a success or a failure.</p>
<p><span class="math display">\[ P(X = k) = \frac{\binom{N}{k} \binom{N-K}{n-k}}{\binom{N}{n}} \]</span></p>
<p>where</p>
<ul>
<li>
<span class="math inline">\(N\)</span> is the population size,</li>
<li>
<span class="math inline">\(K\)</span> is the number of success states in the population,</li>
<li>
<span class="math inline">\(n\)</span> is the number of draws (i.e. quantity drawn in each trial),</li>
<li>
<span class="math inline">\(k\)</span> is the number of observed successes,</li>
<li>
<span class="math inline">\(\binom{n}{k}\)</span> is a binomial coefficient <span class="math inline">\(\frac{n!}{k! (n-k)!}\)</span>.</li>
</ul>
<p>The example used to describe the distribution is an urn contain <span class="math inline">\(N\)</span>
marbles of two colours. There are <span class="math inline">\(K\)</span> green and and <span class="math inline">\(N-K\)</span> red marbles
in the urn. Drawing a green marble is defined as success, and a red
marble failure. Using the formula above, we can compute the
probability to draw <span class="math inline">\(k\)</span> green marbles from the urn.</p>
<p>In the frame of an enrichment analysis <span class="citation">(<label for="tufte-mn-12" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-12" class="margin-toggle">Rivals et al. 2007<span class="marginnote">Rivals, I, L Personnaz, L Taing, and M C Potier. 2007. <span>“Enrichment or Depletion of a GO Category Within a Class of Genes: Which Test?”</span> <em>Bioinformatics</em> 23 (4): 401–7. <a href="https://doi.org/10.1093/bioinformatics/btl633">https://doi.org/10.1093/bioinformatics/btl633</a>.</span>)</span>, we use the
following formulation to calculate a probabiliy that we have more
green marbles than we would expect by change, i.e. there to be an over
representation of green marbles.</p>
<p><span class="math display">\[ p = 1 - \sum_{i = 0}^{k - 1} \frac{\binom{N}{k} \binom{N-K}{n-k}}{\binom{N}{n}} \]</span></p>
<p>To perform an over representation analysis, we thus need to define:</p>
<ul>
<li>among all the genes (called the universe), which ones are
differentially expressed (DE);</li>
<li>among all the genes, which ones are part of the gene set of
interest.</li>
</ul>
<p>And fill out the following table and count the number of DE genes that
are in the set of interest, the non-DE that are in the set, and the DE
and non-DE genes that are not in the set:</p>
<table>
<thead><tr class="header">
<th align="left"></th>
<th align="left">GO</th>
<th align="left">not_GO</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">DE</td>
<td align="left">n</td>
<td align="left">p</td>
</tr>
<tr class="even">
<td align="left">not_DE</td>
<td align="left">m</td>
<td align="left">q</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb279"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb279-1"><a href="sec-gsea.html#cb279-1" tabindex="-1"></a><span class="do">## DE and GO</span></span>
<span id="cb279-2"><a href="sec-gsea.html#cb279-2" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">intersect</span>(res_tbl<span class="sc">$</span>ENTREZID[res_tbl<span class="sc">$</span>padj <span class="sc">&lt;</span> <span class="fl">0.05</span>],</span>
<span id="cb279-3"><a href="sec-gsea.html#cb279-3" tabindex="-1"></a>                      GO_0005925<span class="sc">$</span>ENTREZID))</span>
<span id="cb279-4"><a href="sec-gsea.html#cb279-4" tabindex="-1"></a></span>
<span id="cb279-5"><a href="sec-gsea.html#cb279-5" tabindex="-1"></a><span class="do">## not DE and GO</span></span>
<span id="cb279-6"><a href="sec-gsea.html#cb279-6" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">intersect</span>(res_tbl<span class="sc">$</span>ENTREZID[res_tbl<span class="sc">$</span>padj <span class="sc">&gt;=</span> <span class="fl">0.05</span>],</span>
<span id="cb279-7"><a href="sec-gsea.html#cb279-7" tabindex="-1"></a>                      GO_0005925<span class="sc">$</span>ENTREZID))</span>
<span id="cb279-8"><a href="sec-gsea.html#cb279-8" tabindex="-1"></a></span>
<span id="cb279-9"><a href="sec-gsea.html#cb279-9" tabindex="-1"></a><span class="do">## DE and not GO</span></span>
<span id="cb279-10"><a href="sec-gsea.html#cb279-10" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">setdiff</span>(res_tbl<span class="sc">$</span>ENTREZID[res_tbl<span class="sc">$</span>padj <span class="sc">&lt;</span> <span class="fl">0.05</span>],</span>
<span id="cb279-11"><a href="sec-gsea.html#cb279-11" tabindex="-1"></a>                    GO_0005925<span class="sc">$</span>ENTREZID))</span>
<span id="cb279-12"><a href="sec-gsea.html#cb279-12" tabindex="-1"></a></span>
<span id="cb279-13"><a href="sec-gsea.html#cb279-13" tabindex="-1"></a><span class="do">## not DE not not GO</span></span>
<span id="cb279-14"><a href="sec-gsea.html#cb279-14" tabindex="-1"></a>q <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">setdiff</span>(res_tbl<span class="sc">$</span>ENTREZID[res_tbl<span class="sc">$</span>padj <span class="sc">&gt;=</span> <span class="fl">0.05</span>],</span>
<span id="cb279-15"><a href="sec-gsea.html#cb279-15" tabindex="-1"></a>                    GO_0005925<span class="sc">$</span>ENTREZID))</span></code></pre></div>
<div class="sourceCode" id="cb280"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb280-1"><a href="sec-gsea.html#cb280-1" tabindex="-1"></a>cont_mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(n, m, p, q), <span class="at">nrow =</span> <span class="dv">2</span>)</span>
<span id="cb280-2"><a href="sec-gsea.html#cb280-2" tabindex="-1"></a><span class="fu">rownames</span>(cont_mat) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"DE"</span>, <span class="st">"not_DE"</span>)</span>
<span id="cb280-3"><a href="sec-gsea.html#cb280-3" tabindex="-1"></a><span class="fu">colnames</span>(cont_mat) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GO"</span>, <span class="st">"not_GO"</span>)</span>
<span id="cb280-4"><a href="sec-gsea.html#cb280-4" tabindex="-1"></a>cont_mat</span></code></pre></div>
<pre><code>##         GO not_GO
## DE     158   4039
## not_DE 210   9607</code></pre>
<p>We can now apply a Fisher’s exact (or hypergeometric test) that will
test whether we can identify a statistically enrichment of DE genes in
the GO category.</p>
<div class="sourceCode" id="cb282"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb282-1"><a href="sec-gsea.html#cb282-1" tabindex="-1"></a><span class="fu">fisher.test</span>(cont_mat, <span class="at">alternative =</span> <span class="st">"greater"</span>)</span></code></pre></div>
<pre><code>## 
##  Fisher's Exact Test for Count Data
## 
## data:  cont_mat
## p-value = 6.091e-08
## alternative hypothesis: true odds ratio is greater than 1
## 95 percent confidence interval:
##  1.491543      Inf
## sample estimates:
## odds ratio 
##   1.789422</code></pre>
<p>Note that we could keep the default <code>alternative = "two.sided"</code> to
test to over or under representation.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Repeat the ORA analysis above the <a href="GO:0005813" class="uri">GO:0005813</a> term.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>This approach is straightfoward and very fast. Its major drawback
however is that we need to define a cutoff to differentiate DE from
non-DE genes. Setting this threshold might have a effect on the
results.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Try setting different DE genes and check if, in the cases above, this
has and effect on the GO terms of interest.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="gene-set-enrichment-analysis-gsea" class="section level2" number="6.4">
<h2>
<span class="header-section-number">6.4</span> Gene set enrichment analysis (GSEA)<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('gene-set-enrichment-analysis-gsea')" onmouseout="reset_tooltip('gene-set-enrichment-analysis-gsea-tooltip')"><span class="tooltiptext" id="gene-set-enrichment-analysis-gsea-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Gene set enrichment analysis refers to a broad family of tests. Here,
we will define the principles based on <span class="citation">(<label for="tufte-mn-13" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-13" class="margin-toggle">Subramanian et al. 2005<span class="marginnote">Subramanian, A, P Tamayo, V K Mootha, S Mukherjee, B L Ebert, M A Gillette, A Paulovich, et al. 2005. <span>“Gene Set Enrichment Analysis: A Knowledge-Based Approach for Interpreting Genome-Wide Expression Profiles.”</span> <em>Proc Natl Acad Sci U S A</em> 102 (43): 15545–50. <a href="https://doi.org/10.1073/pnas.0506580102">https://doi.org/10.1073/pnas.0506580102</a>.</span>)</span>, keeping in
mind that the exact implementation will differ in different tools.</p>
<p>The major advantage of GSEA approaches is that they don’t rely on
defining DE genes. The first step is to <strong>order</strong> the genes of
interest based on the statistics used; here, we will use the
p-values. This is already the case for our <code>res_tbl</code> table. We also
need to know which genes are in our set of interest.</p>
<div class="sourceCode" id="cb284"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb284-1"><a href="sec-gsea.html#cb284-1" tabindex="-1"></a>res_tbl<span class="sc">$</span>inGO <span class="ot">&lt;-</span> res_tbl<span class="sc">$</span>ENTREZID <span class="sc">%in%</span> GO_0005925<span class="sc">$</span>ENTREZID</span>
<span id="cb284-2"><a href="sec-gsea.html#cb284-2" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">select</span>(res_tbl, ENTREZID, padj, inGO)</span></code></pre></div>
<pre><code>## # A tibble: 14,014 × 3
##    ENTREZID      padj inGO 
##    &lt;chr&gt;        &lt;dbl&gt; &lt;lgl&gt;
##  1 10857    1.36e-147 FALSE
##  2 79413    5.17e- 67 FALSE
##  3 55270    1.83e- 44 FALSE
##  4 1474     9.11e- 43 FALSE
##  5 7533     1.87e- 42 FALSE
##  6 54974    9.51e- 39 FALSE
##  7 2013     2.00e- 36 FALSE
##  8 55240    1.48e- 33 FALSE
##  9 6482     1.17e- 32 FALSE
## 10 3696     3.75e- 31 TRUE 
## # ℹ 14,004 more rows</code></pre>
<p>We are now going to compute a score by traversing to ordered gene list
and count a positive score when we encounter a gene in the gene set,
and a -1 when the gene is not in the gene set. The positive score,
computed by <code>set_ratio()</code> below, is defined by <span class="math inline">\(\frac{n_{genes} - n_{genes~in~set}}{n_{genes~in~set}}\)</span> so that the sum of all genes in
the set and those not in the set becomes zero.</p>
<div class="sourceCode" id="cb286"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb286-1"><a href="sec-gsea.html#cb286-1" tabindex="-1"></a>set_ratio <span class="ot">&lt;-</span> <span class="cf">function</span>(tbl) {</span>
<span id="cb286-2"><a href="sec-gsea.html#cb286-2" tabindex="-1"></a>    (<span class="fu">nrow</span>(tbl) <span class="sc">-</span> <span class="fu">sum</span>(tbl<span class="sc">$</span>inGO))<span class="sc">/</span><span class="fu">sum</span>(tbl<span class="sc">$</span>inGO)</span>
<span id="cb286-3"><a href="sec-gsea.html#cb286-3" tabindex="-1"></a>}</span></code></pre></div>
<p>The cumulative sum of these scores along the ordered gene list becomes
the GSEA path which is plotted below, and the maximun score obtained
along this path is the GSEA score.</p>
<div class="sourceCode" id="cb287"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb287-1"><a href="sec-gsea.html#cb287-1" tabindex="-1"></a>gsea_path <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(res_tbl<span class="sc">$</span>inGO, <span class="fu">set_ratio</span>(res_tbl), <span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb287-2"><a href="sec-gsea.html#cb287-2" tabindex="-1"></a>gsea_path <span class="ot">&lt;-</span> <span class="fu">cumsum</span>(gsea_path)</span>
<span id="cb287-3"><a href="sec-gsea.html#cb287-3" tabindex="-1"></a>gsea_score <span class="ot">&lt;-</span> <span class="fu">max</span>(gsea_path)</span>
<span id="cb287-4"><a href="sec-gsea.html#cb287-4" tabindex="-1"></a><span class="fu">plot</span>(gsea_path, <span class="at">type =</span> <span class="st">"l"</span>,</span>
<span id="cb287-5"><a href="sec-gsea.html#cb287-5" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Ordered genes (padj)"</span>,</span>
<span id="cb287-6"><a href="sec-gsea.html#cb287-6" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"score"</span>,</span>
<span id="cb287-7"><a href="sec-gsea.html#cb287-7" tabindex="-1"></a>     <span class="at">main =</span> <span class="fu">paste0</span>(<span class="st">"GSEA score: "</span>, <span class="fu">round</span>(gsea_score)))</span>
<span id="cb287-8"><a href="sec-gsea.html#cb287-8" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> gsea_score, <span class="at">col =</span> <span class="st">"steelblue"</span>, <span class="at">lty =</span> <span class="st">"dotted"</span>)</span></code></pre></div>
<div class="figure">
<span style="display:block;" id="fig:gsea1"></span>
<p class="caption marginnote shownote">
Figure 6.3: Gene set enrichment analysis path for term <a href="GO:0005925" class="uri">GO:0005925</a>.
</p>
<img src="WSBIM2122_files/figure-html/gsea1-1.png" alt="Gene set enrichment analysis path for term GO:0005925." width="672">
</div>
<p>To be able to compute a statistical significance, we need to compute a
null distribution of GSEA scores, i.e. a distribution of scores
reflecting the absence of any enrichment. This is done by permuting
the samples in the original data, repeating the statistical analysis,
reorder the genes according to the new statistics (we used the
adjusted p-value above) and compute a new GSEA score.</p>
<div class="sourceCode" id="cb288"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb288-1"><a href="sec-gsea.html#cb288-1" tabindex="-1"></a>dds_tmp <span class="ot">&lt;-</span> dds</span>
<span id="cb288-2"><a href="sec-gsea.html#cb288-2" tabindex="-1"></a>dds_tmp<span class="sc">$</span>Condition <span class="ot">&lt;-</span> dds_tmp<span class="sc">$</span>Condition[<span class="fu">sample</span>(<span class="fu">ncol</span>(dds_tmp))]</span>
<span id="cb288-3"><a href="sec-gsea.html#cb288-3" tabindex="-1"></a><span class="fu">colData</span>(dds)</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 4 columns
##                Cell        Type Condition sizeFactor
##         &lt;character&gt; &lt;character&gt;  &lt;factor&gt;  &lt;numeric&gt;
## sample1       Cell1  Epithelial      mock   0.726883
## sample2       Cell1  Epithelial      mock   1.308664
## sample3       Cell1  Epithelial      mock   1.077368
## sample4       Cell1  Epithelial      KD     0.966025
## sample5       Cell1  Epithelial      KD     1.042111
## sample6       Cell1  Epithelial      KD     1.012664</code></pre>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="sec-gsea.html#cb290-1" tabindex="-1"></a><span class="fu">colData</span>(dds_tmp)</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 4 columns
##                Cell        Type Condition sizeFactor
##         &lt;character&gt; &lt;character&gt;  &lt;factor&gt;  &lt;numeric&gt;
## sample1       Cell1  Epithelial      KD     0.726883
## sample2       Cell1  Epithelial      KD     1.308664
## sample3       Cell1  Epithelial      mock   1.077368
## sample4       Cell1  Epithelial      KD     0.966025
## sample5       Cell1  Epithelial      mock   1.042111
## sample6       Cell1  Epithelial      mock   1.012664</code></pre>
<div class="sourceCode" id="cb292"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb292-1"><a href="sec-gsea.html#cb292-1" tabindex="-1"></a>ensembl_to_geneName <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">ENSEMBL =</span> <span class="fu">rownames</span>(dds)) <span class="sc">%&gt;%</span></span>
<span id="cb292-2"><a href="sec-gsea.html#cb292-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene =</span> AnnotationDbi<span class="sc">::</span><span class="fu">mapIds</span>(org.Hs.eg.db, ENSEMBL, <span class="st">"SYMBOL"</span>,</span>
<span id="cb292-3"><a href="sec-gsea.html#cb292-3" tabindex="-1"></a>                                      <span class="st">"ENSEMBL"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb292-4"><a href="sec-gsea.html#cb292-4" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">ENTREZID =</span> AnnotationDbi<span class="sc">::</span><span class="fu">mapIds</span>(org.Hs.eg.db, ENSEMBL, <span class="st">"ENTREZID"</span>,</span>
<span id="cb292-5"><a href="sec-gsea.html#cb292-5" tabindex="-1"></a>                                          <span class="st">"ENSEMBL"</span>))</span>
<span id="cb292-6"><a href="sec-gsea.html#cb292-6" tabindex="-1"></a></span>
<span id="cb292-7"><a href="sec-gsea.html#cb292-7" tabindex="-1"></a>res_tmp <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds_tmp) <span class="sc">%&gt;%</span></span>
<span id="cb292-8"><a href="sec-gsea.html#cb292-8" tabindex="-1"></a>    <span class="fu">results</span>(<span class="at">name =</span> <span class="st">"Condition_KD_vs_mock"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb292-9"><a href="sec-gsea.html#cb292-9" tabindex="-1"></a>    <span class="fu">as_tibble</span>(<span class="at">rownames =</span> <span class="st">"ENSEMBL"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb292-10"><a href="sec-gsea.html#cb292-10" tabindex="-1"></a>    <span class="fu">left_join</span>(ensembl_to_geneName) <span class="sc">%&gt;%</span></span>
<span id="cb292-11"><a href="sec-gsea.html#cb292-11" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">inGO =</span> ENTREZID <span class="sc">%in%</span> GO_0005925<span class="sc">$</span>ENTREZID) <span class="sc">%&gt;%</span></span>
<span id="cb292-12"><a href="sec-gsea.html#cb292-12" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(ENTREZID),</span>
<span id="cb292-13"><a href="sec-gsea.html#cb292-13" tabindex="-1"></a>           <span class="sc">!</span><span class="fu">is.na</span>(padj),</span>
<span id="cb292-14"><a href="sec-gsea.html#cb292-14" tabindex="-1"></a>           <span class="sc">!</span><span class="fu">duplicated</span>(ENTREZID)) <span class="sc">%&gt;%</span></span>
<span id="cb292-15"><a href="sec-gsea.html#cb292-15" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(ENSEMBL, ENTREZID, padj, inGO) <span class="sc">%&gt;%</span></span>
<span id="cb292-16"><a href="sec-gsea.html#cb292-16" tabindex="-1"></a>    <span class="fu">arrange</span>(padj)</span>
<span id="cb292-17"><a href="sec-gsea.html#cb292-17" tabindex="-1"></a></span>
<span id="cb292-18"><a href="sec-gsea.html#cb292-18" tabindex="-1"></a>res_tmp <span class="ot">&lt;-</span> res_tmp <span class="sc">%&gt;%</span></span>
<span id="cb292-19"><a href="sec-gsea.html#cb292-19" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">score =</span> <span class="fu">ifelse</span>(inGO, <span class="fu">set_ratio</span>(res_tmp), <span class="sc">-</span><span class="dv">1</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb292-20"><a href="sec-gsea.html#cb292-20" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">score =</span> <span class="fu">cumsum</span>(score))</span>
<span id="cb292-21"><a href="sec-gsea.html#cb292-21" tabindex="-1"></a></span>
<span id="cb292-22"><a href="sec-gsea.html#cb292-22" tabindex="-1"></a>res_tmp</span></code></pre></div>
<pre><code>## # A tibble: 14,332 × 5
##    ENSEMBL         ENTREZID    padj inGO  score
##    &lt;chr&gt;           &lt;chr&gt;      &lt;dbl&gt; &lt;lgl&gt; &lt;dbl&gt;
##  1 ENSG00000096433 3710     0.00256 FALSE  -1  
##  2 ENSG00000188157 375790   0.0601  FALSE  -2  
##  3 ENSG00000163359 1293     0.0601  FALSE  -3  
##  4 ENSG00000124942 79026    0.0601  TRUE   34.4
##  5 ENSG00000185567 113146   0.0601  FALSE  33.4
##  6 ENSG00000142197 9980     0.0601  FALSE  32.4
##  7 ENSG00000142798 3339     0.0654  TRUE   69.8
##  8 ENSG00000127603 23499    0.0654  FALSE  68.8
##  9 ENSG00000152377 6695     0.0654  FALSE  67.8
## 10 ENSG00000074054 23332    0.0685  TRUE  105. 
## # ℹ 14,322 more rows</code></pre>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb294-1"><a href="sec-gsea.html#cb294-1" tabindex="-1"></a><span class="fu">max</span>(res_tmp<span class="sc">$</span>score)</span></code></pre></div>
<pre><code>## [1] 521.4718</code></pre>
<div class="sourceCode" id="cb296"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb296-1"><a href="sec-gsea.html#cb296-1" tabindex="-1"></a><span class="fu">plot</span>(res_tmp<span class="sc">$</span>score, <span class="at">type =</span> <span class="st">"l"</span>)</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/gsea_res-1.png" width="672"></p>
<p>This approach is very time consuming, given that the statistical tests
for all the genes need to be recomputed at each permutation. In
addition, for experiments with limited number of samples, the number
of permutations would be limited.</p>
<p>We start by defining a function that will repeat the steps above and
return a list containing the GSEA score and path.</p>
<div class="sourceCode" id="cb297"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb297-1"><a href="sec-gsea.html#cb297-1" tabindex="-1"></a><span class="do">##' A function that assigns the permuted Conditions,</span></span>
<span id="cb297-2"><a href="sec-gsea.html#cb297-2" tabindex="-1"></a><span class="do">##' runs DESeq and a GSEA analysis</span></span>
<span id="cb297-3"><a href="sec-gsea.html#cb297-3" tabindex="-1"></a><span class="do">##'</span></span>
<span id="cb297-4"><a href="sec-gsea.html#cb297-4" tabindex="-1"></a><span class="do">##' @param x a DESeq object.</span></span>
<span id="cb297-5"><a href="sec-gsea.html#cb297-5" tabindex="-1"></a><span class="do">##' @param perm `character()` of length `ncol(x)` indicating the</span></span>
<span id="cb297-6"><a href="sec-gsea.html#cb297-6" tabindex="-1"></a><span class="do">##'     permutation to be tested.</span></span>
<span id="cb297-7"><a href="sec-gsea.html#cb297-7" tabindex="-1"></a><span class="do">##'</span></span>
<span id="cb297-8"><a href="sec-gsea.html#cb297-8" tabindex="-1"></a><span class="do">##' @return `numeric()` containing the GSEA path.</span></span>
<span id="cb297-9"><a href="sec-gsea.html#cb297-9" tabindex="-1"></a>gsea_perm <span class="ot">&lt;-</span> <span class="cf">function</span>(x, perm) {</span>
<span id="cb297-10"><a href="sec-gsea.html#cb297-10" tabindex="-1"></a>    <span class="do">## Set the Condition based on the permutation</span></span>
<span id="cb297-11"><a href="sec-gsea.html#cb297-11" tabindex="-1"></a>    x<span class="sc">$</span>Condition[perm] <span class="ot">&lt;-</span> <span class="st">"mock"</span></span>
<span id="cb297-12"><a href="sec-gsea.html#cb297-12" tabindex="-1"></a>    x<span class="sc">$</span>Condition[<span class="sc">-</span>perm] <span class="ot">&lt;-</span> <span class="st">"KD"</span></span>
<span id="cb297-13"><a href="sec-gsea.html#cb297-13" tabindex="-1"></a>    <span class="do">## Run DESeq2, extract and annotate results</span></span>
<span id="cb297-14"><a href="sec-gsea.html#cb297-14" tabindex="-1"></a>    <span class="fu">suppressMessages</span>(</span>
<span id="cb297-15"><a href="sec-gsea.html#cb297-15" tabindex="-1"></a>        tbl <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(x) <span class="sc">%&gt;%</span></span>
<span id="cb297-16"><a href="sec-gsea.html#cb297-16" tabindex="-1"></a>            <span class="fu">results</span>(<span class="at">name =</span> <span class="st">"Condition_KD_vs_mock"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb297-17"><a href="sec-gsea.html#cb297-17" tabindex="-1"></a>            <span class="fu">as_tibble</span>(<span class="at">rownames =</span> <span class="st">"ENSEMBL"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb297-18"><a href="sec-gsea.html#cb297-18" tabindex="-1"></a>            <span class="fu">left_join</span>(ensembl_to_geneName) <span class="sc">%&gt;%</span></span>
<span id="cb297-19"><a href="sec-gsea.html#cb297-19" tabindex="-1"></a>            <span class="fu">mutate</span>(<span class="at">inGO =</span> ENTREZID <span class="sc">%in%</span> GO_0005925<span class="sc">$</span>ENTREZID) <span class="sc">%&gt;%</span></span>
<span id="cb297-20"><a href="sec-gsea.html#cb297-20" tabindex="-1"></a>            <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(ENTREZID),</span>
<span id="cb297-21"><a href="sec-gsea.html#cb297-21" tabindex="-1"></a>                   <span class="sc">!</span><span class="fu">is.na</span>(padj),</span>
<span id="cb297-22"><a href="sec-gsea.html#cb297-22" tabindex="-1"></a>                   <span class="sc">!</span><span class="fu">duplicated</span>(ENTREZID)) <span class="sc">%&gt;%</span></span>
<span id="cb297-23"><a href="sec-gsea.html#cb297-23" tabindex="-1"></a>            dplyr<span class="sc">::</span><span class="fu">select</span>(ENSEMBL, ENTREZID, padj, inGO) <span class="sc">%&gt;%</span></span>
<span id="cb297-24"><a href="sec-gsea.html#cb297-24" tabindex="-1"></a>            <span class="fu">arrange</span>(padj)</span>
<span id="cb297-25"><a href="sec-gsea.html#cb297-25" tabindex="-1"></a>    )</span>
<span id="cb297-26"><a href="sec-gsea.html#cb297-26" tabindex="-1"></a>    <span class="do">## Compute GSEA results</span></span>
<span id="cb297-27"><a href="sec-gsea.html#cb297-27" tabindex="-1"></a>    tbl <span class="ot">&lt;-</span> tbl <span class="sc">%&gt;%</span></span>
<span id="cb297-28"><a href="sec-gsea.html#cb297-28" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">score =</span> <span class="fu">if_else</span>(inGO, <span class="fu">set_ratio</span>(tbl), <span class="sc">-</span><span class="dv">1</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb297-29"><a href="sec-gsea.html#cb297-29" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">score =</span> <span class="fu">cumsum</span>(score))</span>
<span id="cb297-30"><a href="sec-gsea.html#cb297-30" tabindex="-1"></a>    <span class="fu">return</span>(tbl<span class="sc">$</span>score)</span>
<span id="cb297-31"><a href="sec-gsea.html#cb297-31" tabindex="-1"></a>}</span></code></pre></div>
<p>We will execute this function on all unique permutations,
corresponding on the 10 first columns (including the actual design, in
the first column). Note that columns 11 to 20 are simply the opposite
of the first 10.</p>
<div class="sourceCode" id="cb298"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb298-1"><a href="sec-gsea.html#cb298-1" tabindex="-1"></a>(perms <span class="ot">&lt;-</span> <span class="fu">combn</span>(<span class="dv">6</span>, <span class="dv">3</span>))</span></code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14]
## [1,]    1    1    1    1    1    1    1    1    1     1     2     2     2     2
## [2,]    2    2    2    2    3    3    3    4    4     5     3     3     3     4
## [3,]    3    4    5    6    4    5    6    5    6     6     4     5     6     5
##      [,15] [,16] [,17] [,18] [,19] [,20]
## [1,]     2     2     3     3     3     4
## [2,]     4     5     4     4     5     5
## [3,]     6     6     5     6     6     6</code></pre>
<p>We not apply our function on every permutation.</p>
<div class="sourceCode" id="cb300"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb300-1"><a href="sec-gsea.html#cb300-1" tabindex="-1"></a>paths <span class="ot">&lt;-</span> <span class="fu">apply</span>(perms[, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>], <span class="dv">2</span>, <span class="cf">function</span>(p) <span class="fu">gsea_perm</span>(dds, p))</span></code></pre></div>
<p>And calculate the scores</p>
<div class="sourceCode" id="cb301"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb301-1"><a href="sec-gsea.html#cb301-1" tabindex="-1"></a>scores <span class="ot">&lt;-</span> <span class="fu">sapply</span>(paths, max)</span></code></pre></div>
<p>If we compare the actual score to the permutation scores, we see that
it is the largest one.</p>
<div class="sourceCode" id="cb302"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb302-1"><a href="sec-gsea.html#cb302-1" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(scores))</span>
<span id="cb302-2"><a href="sec-gsea.html#cb302-2" tabindex="-1"></a><span class="fu">rug</span>(scores)</span>
<span id="cb302-3"><a href="sec-gsea.html#cb302-3" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> scores[<span class="dv">1</span>])</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/gsea_perm_plot-1.png" width="672"></p>
<p>An empirical p-value can be computer by dividing the number of null
scores that are greater than the real score divided by the number of
permutations (the number of null scores). In our case, given that no
null scores are greater, the nominal p-value would be 0.</p>
<div class="sourceCode" id="cb303"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb303-1"><a href="sec-gsea.html#cb303-1" tabindex="-1"></a><span class="fu">sum</span>(scores <span class="sc">&gt;</span> scores[<span class="dv">1</span>])<span class="sc">/</span><span class="fu">length</span>(scores)</span></code></pre></div>
<pre><code>## [1] 0</code></pre>
<p>Below, we illustrate all GSEA path and confirm that the actual score
is indeed the largest one.</p>
<div class="sourceCode" id="cb305"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb305-1"><a href="sec-gsea.html#cb305-1" tabindex="-1"></a><span class="fu">plot</span>(paths[[<span class="dv">1</span>]], <span class="at">type =</span> <span class="st">"l"</span>,</span>
<span id="cb305-2"><a href="sec-gsea.html#cb305-2" tabindex="-1"></a>     <span class="at">ylim =</span> <span class="fu">c</span>(<span class="fu">min</span>(<span class="fu">unlist</span>(paths)), <span class="fu">max</span>(scores)),</span>
<span id="cb305-3"><a href="sec-gsea.html#cb305-3" tabindex="-1"></a>     <span class="at">col =</span> <span class="st">"steelblue"</span>)</span>
<span id="cb305-4"><a href="sec-gsea.html#cb305-4" tabindex="-1"></a><span class="fu">grid</span>()</span>
<span id="cb305-5"><a href="sec-gsea.html#cb305-5" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(paths))</span>
<span id="cb305-6"><a href="sec-gsea.html#cb305-6" tabindex="-1"></a>    <span class="fu">lines</span>(paths[[i]], <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">col =</span> <span class="st">"#00000060"</span>)</span></code></pre></div>
<div class="figure">
<span style="display:block;" id="fig:unnamed-chunk-147"></span>
<p class="caption marginnote shownote">
Figure 6.4: Representation of all GSEA paths: real path (blue) and random paths (dotted grey).
</p>
<img src="WSBIM2122_files/figure-html/unnamed-chunk-147-1.png" alt="Representation of all GSEA paths: real path (blue) and random paths (dotted grey)." width="672">
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Given that we have 10 permutations, how many null scores greater that
the actual GSEA score do we need for our results to become non
significant (assuming we set an alpha of 0.05)?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-19" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-19', 'sol-start-19')"></span>
</p>
<div id="sol-body-19" class="solution-body" style="display: none;">
<p>1 permutation score or more is enough to loose significativite, as a
single value greater than 2156 produces a p-value
of 1/10 = 0.1.</p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>There exist various approaches to the GSEA analysis, that apply
different permutation approaches to increate the number of possible
permutations and reduce the running time.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Repeat the GSEA analysis above the <a href="GO:0005813" class="uri">GO:0005813</a> term.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>While there is not need to define a list of differentially expressed
genes for the GSEA analysis, the genes need to be ordered, which can
be done in different ways and lead to different results.</p>
</div>
<div id="using-the-clusterprofiler-package" class="section level2" number="6.5">
<h2>
<span class="header-section-number">6.5</span> Using the <code>clusterProfiler</code> package<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('using-the-clusterprofiler-package')" onmouseout="reset_tooltip('using-the-clusterprofiler-package-tooltip')"><span class="tooltiptext" id="using-the-clusterprofiler-package-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>I practice, the analyses presented above are executed using any of
<a href="http://bioconductor.org/packages/release/BiocViews.html#___GeneSetEnrichment">the very
many</a>
packages that are available.
Here, we will demonstrate <em><a href="https://bioconductor.org/packages/3.19/clusterProfiler">clusterProfiler</a></em><label for="tufte-sn-12" class="margin-toggle sidenote-number">12</label><input type="checkbox" id="tufte-sn-12" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">12</span> <a href="https://yulab-smu.top/clusterProfiler-book/" class="uri">https://yulab-smu.top/clusterProfiler-book/</a></span>,
that itself relies on other packages to perform the GSEA-related computations.</p>
<div class="sourceCode" id="cb306"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb306-1"><a href="sec-gsea.html#cb306-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"clusterProfiler"</span>)</span></code></pre></div>
<p>These dedicated packages will perform either of the two tests (or
variations thereof) to all sets, adjust the computed p-values for
multiplicity, and provide additional details and visualisations for
further exploration of the results.</p>
<div id="ora-using-go-sets" class="section level3" number="6.5.1">
<h3>
<span class="header-section-number">6.5.1</span> ORA using GO sets<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('ora-using-go-sets')" onmouseout="reset_tooltip('ora-using-go-sets-tooltip')"><span class="tooltiptext" id="ora-using-go-sets-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<div class="sourceCode" id="cb307"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb307-1"><a href="sec-gsea.html#cb307-1" tabindex="-1"></a>de_genes <span class="ot">&lt;-</span> res_tbl<span class="sc">$</span>ENTREZID[res_tbl<span class="sc">$</span>padj <span class="sc">&lt;</span> <span class="fl">0.05</span>]</span>
<span id="cb307-2"><a href="sec-gsea.html#cb307-2" tabindex="-1"></a></span>
<span id="cb307-3"><a href="sec-gsea.html#cb307-3" tabindex="-1"></a>go_ora <span class="ot">&lt;-</span> <span class="fu">enrichGO</span>(<span class="at">gene =</span> de_genes,</span>
<span id="cb307-4"><a href="sec-gsea.html#cb307-4" tabindex="-1"></a>                   <span class="at">universe =</span> res_tbl<span class="sc">$</span>ENTREZID,</span>
<span id="cb307-5"><a href="sec-gsea.html#cb307-5" tabindex="-1"></a>                   <span class="at">OrgDb =</span> org.Hs.eg.db,</span>
<span id="cb307-6"><a href="sec-gsea.html#cb307-6" tabindex="-1"></a>                   <span class="at">ont =</span> <span class="st">"CC"</span>,</span>
<span id="cb307-7"><a href="sec-gsea.html#cb307-7" tabindex="-1"></a>                   <span class="at">readable =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb307-8"><a href="sec-gsea.html#cb307-8" tabindex="-1"></a>    as_tibble</span>
<span id="cb307-9"><a href="sec-gsea.html#cb307-9" tabindex="-1"></a>go_ora</span></code></pre></div>
<pre><code>## # A tibble: 21 × 12
##    ID    Description GeneRatio BgRatio RichFactor FoldEnrichment zScore   pvalue
##    &lt;chr&gt; &lt;chr&gt;       &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;          &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt;
##  1 GO:0… cytosolic … 70/3862   113/12…      0.619           1.94   6.86 4.79e-11
##  2 GO:0… cytosolic … 40/3862   56/120…      0.714           2.23   6.34 1.52e- 9
##  3 GO:0… ribosomal … 92/3862   185/12…      0.497           1.56   5.22 3.30e- 7
##  4 GO:0… ribosome    105/3862  226/12…      0.465           1.45   4.71 3.20e- 6
##  5 GO:0… focal adhe… 158/3862  368/12…      0.429           1.34   4.58 4.93e- 6
##  6 GO:0… cell-subst… 159/3862  374/12…      0.425           1.33   4.44 9.01e- 6
##  7 GO:0… large ribo… 57/3862   114/12…      0.5             1.56   4.15 4.48e- 5
##  8 GO:0… neuron to … 103/3862  235/12…      0.438           1.37   3.93 8.04e- 5
##  9 GO:0… cell surfa… 176/3862  435/12…      0.405           1.27   3.86 9.04e- 5
## 10 GO:0… postsynapt… 95/3862   215/12…      0.442           1.38   3.87 1.04e- 4
## # ℹ 11 more rows
## # ℹ 4 more variables: p.adjust &lt;dbl&gt;, qvalue &lt;dbl&gt;, geneID &lt;chr&gt;, Count &lt;int&gt;</code></pre>
<p>The <code>GeneRatio</code> and <code>BgRatio</code> correspond respectively to</p>
<ul>
<li><p>the ration between the number of genes of interest in the gene set
and the number of (unique) genes-of-interest:
<span class="math inline">\(\frac{|genes\_of\_interest~in~gene\_set|}{|genes\_of\_interest|}\)</span></p></li>
<li><p>the ratio between the number of genes of interest in the gene set
and size of the universe, i.e. the number of (unique) genes across
all gene sets: <span class="math inline">\(\frac{|genes~in~gene\_set|}{|universe|}\)</span></p></li>
</ul>
<!-- Example from: https://www.biostars.org/p/220465/ --><p>Below, we have 12 genes of interest, <code>"a"</code>, to <code>"l"</code>:</p>
<pre><code>##  [1] "a" "b" "c" "d" "e" "f" "g" "h" "i" "j" "k" "l"</code></pre>
<p>And two gene sets, <code>"X"</code> and <code>"Y"</code>:</p>
<pre><code>##    geneset entrez_gene
## 1        X           a
## 2        X           b
## 3        X           c
## 4        X           d
## 5        X           e
## 6        X           f
## 7        X           g
## 8        X           h
## 9        X           i
## 10       X           j</code></pre>
<pre><code>##    geneset entrez_gene
## 1        Y           b
## 2        Y           c
## 3        Y           d
## 4        Y           e
## 5        Y           f
## 6        Y           g
## 7        Y           h
## 8        Y           i
## 9        Y           j
## 10       Y           k
## 11       Y           l
## 12       Y           m
## 13       Y           n
## 14       Y           o
## 15       Y           p
## 16       Y           q
## 17       Y           r
## 18       Y           s
## 19       Y           t
## 20       Y           u
## 21       Y           v
## 22       Y           w
## 23       Y           x
## 24       Y           y
## 25       Y           z</code></pre>
<p>We get:</p>
<pre><code>##   GeneRatio BgRatio                geneID Count
## X     10/12   10/26   a/b/c/d/e/f/g/h/i/j    10
## Y     11/12   25/26 b/c/d/e/f/g/h/i/j/k/l    11</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Repeat the ORA analysis above using all GO namespaces. See <code>?enrichGO</code>
for details.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="gsea-using-go-sets" class="section level3" number="6.5.2">
<h3>
<span class="header-section-number">6.5.2</span> GSEA using GO sets<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('gsea-using-go-sets')" onmouseout="reset_tooltip('gsea-using-go-sets-tooltip')"><span class="tooltiptext" id="gsea-using-go-sets-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>The functions that perform GSEA in <code>clusterProfiler</code> require the genes
to be ordered in decreasing order. This indicates that the p-values
can’t be used without transformations. One could also use the log2
fold-change or the (absolute value) test statistics.</p>
<div class="sourceCode" id="cb313"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb313-1"><a href="sec-gsea.html#cb313-1" tabindex="-1"></a>ordered_genes <span class="ot">&lt;-</span> <span class="fu">abs</span>(res_tbl<span class="sc">$</span>stat)</span>
<span id="cb313-2"><a href="sec-gsea.html#cb313-2" tabindex="-1"></a><span class="fu">names</span>(ordered_genes) <span class="ot">&lt;-</span> res_tbl<span class="sc">$</span>ENTREZID</span>
<span id="cb313-3"><a href="sec-gsea.html#cb313-3" tabindex="-1"></a>ordered_genes <span class="ot">&lt;-</span> <span class="fu">sort</span>(ordered_genes, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb313-4"><a href="sec-gsea.html#cb313-4" tabindex="-1"></a></span>
<span id="cb313-5"><a href="sec-gsea.html#cb313-5" tabindex="-1"></a>go_gsea <span class="ot">&lt;-</span> <span class="fu">gseGO</span>(<span class="at">gene =</span> ordered_genes,</span>
<span id="cb313-6"><a href="sec-gsea.html#cb313-6" tabindex="-1"></a>                 <span class="at">OrgDb =</span> org.Hs.eg.db,</span>
<span id="cb313-7"><a href="sec-gsea.html#cb313-7" tabindex="-1"></a>                 <span class="at">scoreType =</span> <span class="st">"pos"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb313-8"><a href="sec-gsea.html#cb313-8" tabindex="-1"></a>    as_tibble</span>
<span id="cb313-9"><a href="sec-gsea.html#cb313-9" tabindex="-1"></a>go_gsea</span></code></pre></div>
<pre><code>## # A tibble: 542 × 11
##    ID        Description setSize enrichmentScore   NES   pvalue p.adjust  qvalue
##    &lt;chr&gt;     &lt;chr&gt;         &lt;int&gt;           &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;
##  1 GO:00308… epithelial…     441           0.479  1.40 1.44e-10  7.61e-7 5.87e-7
##  2 GO:00015… blood vess…     493           0.459  1.34 3.63e- 9  9.59e-6 7.41e-6
##  3 GO:00485… blood vess…     434           0.460  1.35 1.60e- 8  2.81e-5 2.17e-5
##  4 GO:00015… angiogenes…     376           0.463  1.35 4.57e- 8  6.03e-5 4.66e-5
##  5 GO:00343… cell junct…     499           0.448  1.31 6.76e- 8  7.14e-5 5.52e-5
##  6 GO:00708… response t…     493           0.445  1.30 1.91e- 7  1.68e-4 1.30e-4
##  7 GO:00435… regulation…     394           0.457  1.33 2.90e- 7  1.83e-4 1.41e-4
##  8 GO:00430… positive r…     396           0.453  1.32 2.73e- 7  1.83e-4 1.41e-4
##  9 GO:00713… cellular r…     472           0.447  1.31 3.11e- 7  1.83e-4 1.41e-4
## 10 GO:00455… negative r…     426           0.450  1.32 3.62e- 7  1.91e-4 1.48e-4
## # ℹ 532 more rows
## # ℹ 3 more variables: rank &lt;int&gt;, leading_edge &lt;chr&gt;, core_enrichment &lt;chr&gt;</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Repeat the GSEA analysis above using all GO namespaces. See
<code>?enrichGO</code> for details.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Compare and discuss the use of these different ways to order the genes
for the GSEA analysis.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Compare the GO and GSEA results. Which GO terms are shared or unique
to each approach?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-20" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-20', 'sol-start-20')"></span>
</p>
<div id="sol-body-20" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb315"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb315-1"><a href="sec-gsea.html#cb315-1" tabindex="-1"></a>go_cc <span class="ot">&lt;-</span> <span class="fu">full_join</span>(go_ora <span class="sc">%&gt;%</span></span>
<span id="cb315-2"><a href="sec-gsea.html#cb315-2" tabindex="-1"></a>                   <span class="fu">filter</span>(p.adjust <span class="sc">&lt;</span> <span class="fl">0.01</span>) <span class="sc">%&gt;%</span></span>
<span id="cb315-3"><a href="sec-gsea.html#cb315-3" tabindex="-1"></a>                   dplyr<span class="sc">::</span><span class="fu">select</span>(ID, p.adjust),</span>
<span id="cb315-4"><a href="sec-gsea.html#cb315-4" tabindex="-1"></a>                 go_gsea <span class="sc">%&gt;%</span></span>
<span id="cb315-5"><a href="sec-gsea.html#cb315-5" tabindex="-1"></a>                   <span class="fu">filter</span>(p.adjust <span class="sc">&lt;</span> <span class="fl">0.01</span>) <span class="sc">%&gt;%</span></span>
<span id="cb315-6"><a href="sec-gsea.html#cb315-6" tabindex="-1"></a>                   dplyr<span class="sc">::</span><span class="fu">select</span>(ID, p.adjust),</span>
<span id="cb315-7"><a href="sec-gsea.html#cb315-7" tabindex="-1"></a>                 <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"ID"</span> <span class="ot">=</span> <span class="st">"ID"</span>))</span>
<span id="cb315-8"><a href="sec-gsea.html#cb315-8" tabindex="-1"></a></span>
<span id="cb315-9"><a href="sec-gsea.html#cb315-9" tabindex="-1"></a><span class="do">## shared</span></span>
<span id="cb315-10"><a href="sec-gsea.html#cb315-10" tabindex="-1"></a><span class="fu">na.omit</span>(go_cc)</span></code></pre></div>
<pre><code>## # A tibble: 0 × 3
## # ℹ 3 variables: ID &lt;chr&gt;, p.adjust.x &lt;dbl&gt;, p.adjust.y &lt;dbl&gt;</code></pre>
<div class="sourceCode" id="cb317"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb317-1"><a href="sec-gsea.html#cb317-1" tabindex="-1"></a><span class="do">## ORA only</span></span>
<span id="cb317-2"><a href="sec-gsea.html#cb317-2" tabindex="-1"></a><span class="fu">filter</span>(go_cc, <span class="fu">is.na</span>(p.adjust.y))</span></code></pre></div>
<pre><code>## # A tibble: 13 × 3
##    ID           p.adjust.x p.adjust.y
##    &lt;chr&gt;             &lt;dbl&gt;      &lt;dbl&gt;
##  1 GO:0022626 0.0000000323         NA
##  2 GO:0022625 0.000000512          NA
##  3 GO:0044391 0.0000740            NA
##  4 GO:0005840 0.000539             NA
##  5 GO:0005925 0.000663             NA
##  6 GO:0030055 0.00101              NA
##  7 GO:0015934 0.00431              NA
##  8 GO:0098984 0.00676              NA
##  9 GO:0009986 0.00676              NA
## 10 GO:0099572 0.00686              NA
## 11 GO:0014069 0.00686              NA
## 12 GO:0022627 0.00686              NA
## 13 GO:0098857 0.00772              NA</code></pre>
<div class="sourceCode" id="cb319"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb319-1"><a href="sec-gsea.html#cb319-1" tabindex="-1"></a><span class="do">## GSEA only</span></span>
<span id="cb319-2"><a href="sec-gsea.html#cb319-2" tabindex="-1"></a><span class="fu">filter</span>(go_cc, <span class="fu">is.na</span>(p.adjust.x))</span></code></pre></div>
<pre><code>## # A tibble: 170 × 3
##    ID         p.adjust.x  p.adjust.y
##    &lt;chr&gt;           &lt;dbl&gt;       &lt;dbl&gt;
##  1 GO:0030855         NA 0.000000761
##  2 GO:0001568         NA 0.00000959 
##  3 GO:0048514         NA 0.0000281  
##  4 GO:0001525         NA 0.0000603  
##  5 GO:0034330         NA 0.0000714  
##  6 GO:0070848         NA 0.000168   
##  7 GO:0043549         NA 0.000183   
##  8 GO:0043068         NA 0.000183   
##  9 GO:0071363         NA 0.000183   
## 10 GO:0045596         NA 0.000191   
## # ℹ 160 more rows</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="using-kegg-sets" class="section level3" number="6.5.3">
<h3>
<span class="header-section-number">6.5.3</span> Using KEGG sets<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('using-kegg-sets')" onmouseout="reset_tooltip('using-kegg-sets-tooltip')"><span class="tooltiptext" id="using-kegg-sets-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>The <code>enrichKEGG()</code> and <code>gseKEGG()</code> functions can be used to perform
the ORA and GSEA analyses againt the KEGG sets.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Perform ORA and GSEA analysis using the KEGG set.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="using-msigdb-sets" class="section level3" number="6.5.4">
<h3>
<span class="header-section-number">6.5.4</span> Using MSigDb sets<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('using-msigdb-sets')" onmouseout="reset_tooltip('using-msigdb-sets-tooltip')"><span class="tooltiptext" id="using-msigdb-sets-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>We will use the hallmark gene set as an example here, but any set
described above can be utilised.</p>
<div class="sourceCode" id="cb321"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb321-1"><a href="sec-gsea.html#cb321-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"msigdbr"</span>)</span>
<span id="cb321-2"><a href="sec-gsea.html#cb321-2" tabindex="-1"></a>msig_h <span class="ot">&lt;-</span> <span class="fu">msigdbr</span>(<span class="at">species =</span> <span class="st">"Homo sapiens"</span>, <span class="at">category =</span> <span class="st">"H"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb321-3"><a href="sec-gsea.html#cb321-3" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(gs_name, entrez_gene) <span class="sc">%&gt;%</span></span>
<span id="cb321-4"><a href="sec-gsea.html#cb321-4" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">ont =</span> gs_name, <span class="at">gene =</span> entrez_gene)</span>
<span id="cb321-5"><a href="sec-gsea.html#cb321-5" tabindex="-1"></a>msig_h</span></code></pre></div>
<pre><code>## # A tibble: 8,209 × 2
##    ont                    gene
##    &lt;chr&gt;                 &lt;int&gt;
##  1 HALLMARK_ADIPOGENESIS    19
##  2 HALLMARK_ADIPOGENESIS 11194
##  3 HALLMARK_ADIPOGENESIS 10449
##  4 HALLMARK_ADIPOGENESIS    33
##  5 HALLMARK_ADIPOGENESIS    34
##  6 HALLMARK_ADIPOGENESIS    35
##  7 HALLMARK_ADIPOGENESIS    47
##  8 HALLMARK_ADIPOGENESIS    50
##  9 HALLMARK_ADIPOGENESIS    51
## 10 HALLMARK_ADIPOGENESIS   112
## # ℹ 8,199 more rows</code></pre>
<p>The <code>msig_h</code> table can now be used with the <code>enricher()</code> and <code>GSEA()</code>
functions from the <code>clusterProfiler</code> package to perform ORA and GSEA
analyses on the MSigDB hallmark gene set.</p>
<div class="sourceCode" id="cb323"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb323-1"><a href="sec-gsea.html#cb323-1" tabindex="-1"></a>msig_ora <span class="ot">&lt;-</span> <span class="fu">enricher</span>(<span class="at">gene =</span> de_genes,</span>
<span id="cb323-2"><a href="sec-gsea.html#cb323-2" tabindex="-1"></a>                     <span class="at">universe =</span> res_tbl<span class="sc">$</span>ENTREZID,</span>
<span id="cb323-3"><a href="sec-gsea.html#cb323-3" tabindex="-1"></a>                     <span class="at">TERM2GENE =</span> msig_h) <span class="sc">%&gt;%</span></span>
<span id="cb323-4"><a href="sec-gsea.html#cb323-4" tabindex="-1"></a>    as_tibble</span>
<span id="cb323-5"><a href="sec-gsea.html#cb323-5" tabindex="-1"></a>msig_ora</span></code></pre></div>
<pre><code>## # A tibble: 3 × 12
##   ID      Description GeneRatio BgRatio RichFactor FoldEnrichment zScore  pvalue
##   &lt;chr&gt;   &lt;chr&gt;       &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;          &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;
## 1 HALLMA… HALLMARK_H… 82/1212   164/33…      0.5             1.38   3.77 1.46e-4
## 2 HALLMA… HALLMARK_M… 34/1212   58/3347      0.586           1.62   3.58 3.79e-4
## 3 HALLMA… HALLMARK_A… 53/1212   106/33…      0.5             1.38   3.00 2.17e-3
## # ℹ 4 more variables: p.adjust &lt;dbl&gt;, qvalue &lt;dbl&gt;, geneID &lt;chr&gt;, Count &lt;int&gt;</code></pre>
<div class="sourceCode" id="cb325"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb325-1"><a href="sec-gsea.html#cb325-1" tabindex="-1"></a>msig_gsea <span class="ot">&lt;-</span> <span class="fu">GSEA</span>(ordered_genes, <span class="at">TERM2GENE =</span> msig_h, <span class="at">scoreType =</span> <span class="st">"pos"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb325-2"><a href="sec-gsea.html#cb325-2" tabindex="-1"></a>    as_tibble</span>
<span id="cb325-3"><a href="sec-gsea.html#cb325-3" tabindex="-1"></a>msig_gsea</span></code></pre></div>
<pre><code>## # A tibble: 23 × 11
##    ID         Description setSize enrichmentScore   NES  pvalue p.adjust  qvalue
##    &lt;chr&gt;      &lt;chr&gt;         &lt;int&gt;           &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;
##  1 HALLMARK_… HALLMARK_E…     128           0.554  1.56 6.24e-8  3.12e-6 1.58e-6
##  2 HALLMARK_… HALLMARK_H…     164           0.509  1.45 2.01e-6  5.03e-5 2.54e-5
##  3 HALLMARK_… HALLMARK_M…      58           0.605  1.62 1.31e-5  1.64e-4 8.29e-5
##  4 HALLMARK_… HALLMARK_P…     178           0.489  1.40 1.30e-5  1.64e-4 8.29e-5
##  5 HALLMARK_… HALLMARK_G…     175           0.477  1.36 4.85e-5  4.85e-4 2.45e-4
##  6 HALLMARK_… HALLMARK_T…     160           0.482  1.37 9.87e-5  8.23e-4 4.16e-4
##  7 HALLMARK_… HALLMARK_U…     122           0.492  1.38 4.63e-4  3.30e-3 1.67e-3
##  8 HALLMARK_… HALLMARK_K…     124           0.486  1.36 5.72e-4  3.58e-3 1.81e-3
##  9 HALLMARK_… HALLMARK_X…     134           0.478  1.35 8.29e-4  4.33e-3 2.19e-3
## 10 HALLMARK_… HALLMARK_E…     152           0.470  1.33 8.66e-4  4.33e-3 2.19e-3
## # ℹ 13 more rows
## # ℹ 3 more variables: rank &lt;int&gt;, leading_edge &lt;chr&gt;, core_enrichment &lt;chr&gt;</code></pre>
</div>
</div>
<div id="visualisation-of-enrichment-analyses" class="section level2" number="6.6">
<h2>
<span class="header-section-number">6.6</span> Visualisation of enrichment analyses<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('visualisation-of-enrichment-analyses')" onmouseout="reset_tooltip('visualisation-of-enrichment-analyses-tooltip')"><span class="tooltiptext" id="visualisation-of-enrichment-analyses-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>The <code>clusterProfiler</code> documentation provides a chapter on the
<a href="http://yulab-smu.top/clusterProfiler-book/chapter12.html">visualization of functional enrichment
results</a>.</p>
<p>Another useful visualisation, that links the enrichment results back
to the whole set of results is to highlight the genes in a particular
set of interest on the volcano plot.</p>
<div class="sourceCode" id="cb327"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb327-1"><a href="sec-gsea.html#cb327-1" tabindex="-1"></a>sel <span class="ot">&lt;-</span> res_tbl<span class="sc">$</span>ENTREZID <span class="sc">%in%</span> GO_0005925<span class="sc">$</span>ENTREZID</span>
<span id="cb327-2"><a href="sec-gsea.html#cb327-2" tabindex="-1"></a><span class="fu">plot</span>(res_tbl<span class="sc">$</span>log2FoldChange, <span class="sc">-</span><span class="fu">log10</span>(res_tbl<span class="sc">$</span>padj))</span>
<span id="cb327-3"><a href="sec-gsea.html#cb327-3" tabindex="-1"></a><span class="fu">points</span>(res_tbl<span class="sc">$</span>log2FoldChange[sel],</span>
<span id="cb327-4"><a href="sec-gsea.html#cb327-4" tabindex="-1"></a>     <span class="sc">-</span><span class="fu">log10</span>(res_tbl<span class="sc">$</span>padj)[sel],</span>
<span id="cb327-5"><a href="sec-gsea.html#cb327-5" tabindex="-1"></a>     <span class="at">col =</span> <span class="st">"red"</span>)</span>
<span id="cb327-6"><a href="sec-gsea.html#cb327-6" tabindex="-1"></a><span class="fu">grid</span>()</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-149-1.png" width="672"></p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Extract the Entrez ids from the top hit in the <code>msig_gsea</code> results
above (available in <code>core_enrichment</code>) and visualise the results on a
volcano plot.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-21" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-21', 'sol-start-21')"></span>
</p>
<div id="sol-body-21" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb328"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb328-1"><a href="sec-gsea.html#cb328-1" tabindex="-1"></a>entrez_ids <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(msig_gsea<span class="sc">$</span>core_enrichment[[<span class="dv">1</span>]], <span class="st">"/"</span>)[[<span class="dv">1</span>]]</span>
<span id="cb328-2"><a href="sec-gsea.html#cb328-2" tabindex="-1"></a></span>
<span id="cb328-3"><a href="sec-gsea.html#cb328-3" tabindex="-1"></a>res_tbl <span class="sc">%&gt;%</span></span>
<span id="cb328-4"><a href="sec-gsea.html#cb328-4" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">inGO =</span> ENTREZID <span class="sc">%in%</span> entrez_ids) <span class="sc">%&gt;%</span></span>
<span id="cb328-5"><a href="sec-gsea.html#cb328-5" tabindex="-1"></a>    <span class="fu">arrange</span>(inGO) <span class="sc">%&gt;%</span> <span class="do">## to plot inGO points last (on top)</span></span>
<span id="cb328-6"><a href="sec-gsea.html#cb328-6" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> log2FoldChange,</span>
<span id="cb328-7"><a href="sec-gsea.html#cb328-7" tabindex="-1"></a>               <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(padj),</span>
<span id="cb328-8"><a href="sec-gsea.html#cb328-8" tabindex="-1"></a>               <span class="at">colour =</span> inGO)) <span class="sc">+</span></span>
<span id="cb328-9"><a href="sec-gsea.html#cb328-9" tabindex="-1"></a>    <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="WSBIM2122_files/figure-html/unnamed-chunk-150-1.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>Finally, the
<a href="https://bioconductor.org/packages/release/bioc/html/pathview.html">pathview</a>
package allows to annotate KEEG pathways, colouring the genes/nodes
based on quantiative values such as log fold-changes to visualise the
impact or a specific experiment on a pathway of interest.</p>
</div>
<div id="discussion" class="section level2" number="6.7">
<h2>
<span class="header-section-number">6.7</span> Discussion<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('discussion')" onmouseout="reset_tooltip('discussion-tooltip')"><span class="tooltiptext" id="discussion-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Enrichment analyses are powerful techniques, as they allow to
integrate thousands of univariate results into biologically driven
sets.</p>
<p>Their interpretation isn’t however always straightforward when
different methods (ORA and GSEA) and implementations (different GSEA
permutation strategies) or user-set parameters (DE cutoffs in ORA,
ordering criteria in GSEA) provide different results.</p>

</div>
</div>
<p style="text-align: center;">
<a href="sec-design.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-ms.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2024-11-08
 using 
R version 4.4.1 (2024-06-14)
</p>
</div>
</div>



</body>
</html>
