<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 1 R markdown | Omics Data Analysis" />
<meta property="og:type" content="book" />

<meta property="og:description" content="Course material for the Omics Data Analysis (WSBIM2122) course at UCLouvain." />


<meta name="author" content="Laurent Gatto" />

<meta name="date" content="2024-11-01" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="Course material for the Omics Data Analysis (WSBIM2122) course at UCLouvain.">

<title>Chapter 1 R markdown | Omics Data Analysis</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>
<script>
function copy_link(id) {
  var dummy = document.createElement('input'),
  text = window.location.href.split(/[?#]/)[0] + '#' + id;
  document.body.appendChild(dummy);
  dummy.value = text;
  dummy.select();
  document.execCommand('copy');
  document.body.removeChild(dummy);
  
  var tooltip = document.getElementById(id + '-tooltip');
  tooltip.innerHTML = 'Copied!';
}

function reset_tooltip(id) {
  var tooltip = document.getElementById(id);
  tooltip.innerHTML = 'Copy link';
}
</script>
<link href="libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="libs/datatables-binding-0.33/datatables.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>
<script src="libs/plotly-binding-4.10.4/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-2.11.1/plotly-latest.min.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }

code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">Omics Data Analysis<p><p class="author">Laurent Gatto</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html" id="toc-preamble">Preamble</a>
<a id="active-page" href="sec-rmd.html" id="toc-sec-rmd"><span class="toc-section-number">1</span> R markdown</a><ul class="toc-sections">
<li class="toc"><a href="#knitting-from-within-r"> Knitting from within R</a></li>
<li class="toc"><a href="#caching"> Caching</a></li>
<li class="toc"><a href="#writing-slides-in-rmarkdown"> Writing slides in Rmarkdown</a></li>
<li class="toc"><a href="#citations"> Citations</a></li>
<li class="toc"><a href="#references"> References</a></li>
</ul>
<a href="sec-linmod.html" id="toc-sec-linmod"><span class="toc-section-number">2</span> Linear models</a>
<a href="sec-hts.html" id="toc-sec-hts"><span class="toc-section-number">3</span> High throughput sequencing</a>
<a href="sec-rnaseq.html" id="toc-sec-rnaseq"><span class="toc-section-number">4</span> Differential expression analysis</a>
<a href="sec-design.html" id="toc-sec-design"><span class="toc-section-number">5</span> More complex experimental design</a>
<a href="sec-gsea.html" id="toc-sec-gsea"><span class="toc-section-number">6</span> Enrichment analyses</a>
<a href="sec-ms.html" id="toc-sec-ms"><span class="toc-section-number">7</span> Mass spectrometry</a>
<a href="sec-prot.html" id="toc-sec-prot"><span class="toc-section-number">8</span> Quantitative proteomics data analysis</a>
<a href="sec-ccl.html" id="toc-sec-ccl"><span class="toc-section-number">9</span> Conclusions</a>
<a href="sec-cli.html" id="toc-sec-cli"><span class="toc-section-number">10</span> Annex: the shell</a>
<a href="sec-git.html" id="toc-sec-git"><span class="toc-section-number">11</span> Annex: Collaborating with Git and GitHub</a>
<a href="sec-info.html" id="toc-sec-info"><span class="toc-section-number">12</span> Course and session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="sec-rmd" class="section level1" number="1">
<h1>
<span class="header-section-number">Chapter 1</span> R markdown</h1>
<p>We have already seen how to implement <a href="https://uclouvain-cbio.github.io/WSBIM1207/sec-rr.html">reproducible
report</a> using
the
<a href="https://uclouvain-cbio.github.io/WSBIM1207/sec-rr.html#knitr-and-rmarkdown">Rmarkdown</a>
<a href="http://rmarkdown.rstudio.com/">package</a> in the WSBIM1207 cours.</p>
<p>Here, we are going to see a few additional capabilities</p>
<div id="knitting-from-within-r" class="section level2" number="1.1">
<h2>
<span class="header-section-number">1.1</span> Knitting from within R<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('knitting-from-within-r')" onmouseout="reset_tooltip('knitting-from-within-r-tooltip')"><span class="tooltiptext" id="knitting-from-within-r-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>When compiling Rmd documents with many code chunks, it becomes
difficult to quickly locate bugs in case of errors:</p>
<pre><code>label: unnamed-chunk-67
  |...........................................................           |  85%
   inline R code fragments

  |............................................................          |  85%
label: unnamed-chunk-68
  |............................................................          |  86%
  ordinary text without R code


label: unnamed-chunk-69
  |.............................................................         |  87%
  ordinary text without R code


label: unnamed-chunk-70
  |..............................................................        |  88%
  ordinary text without R code

  |..............................................................        |  89%
label: unnamed-chunk-71

   inline R code fragments</code></pre>
<p>It is recommended to <strong>uniquely</strong> name each code chunk. These labels
are then reported in the knitting progress output, and it becomes
trivial to identify the code chunk that produces the error.</p>
<pre><code>label: si
  |....................................................................  |  98%
  ordinary text without R code

  |..................................................................... |  98%
label: pkgs (with options)
List of 4
 $ eval   : logi TRUE
 $ echo   : logi FALSE
 $ results: chr "markup"
 $ comment: chr ""

  |..................................................................... |  99%
  ordinary text without R code

  |......................................................................|  99%
label: setup (with options)
List of 1
 $ eval: logi FALSE

  |......................................................................| 100%
  ordinary text without R code</code></pre>
<p>There will be cases when the <code>Knit</code> button in RStudio will not be
available. A typical example is when you use R on a server through the
terminal. In such cases, you can directly use the
<code>rmarkdown::render()</code> function:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="sec-rmd.html#cb5-1" tabindex="-1"></a>rmarkdown<span class="sc">::</span><span class="fu">render</span>(<span class="st">"report.Rmd"</span>)</span></code></pre></div>
<p>It is also possibly to define the output format, which will override
the output format defined in the header:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="sec-rmd.html#cb6-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"rmarkdown"</span>)</span>
<span id="cb6-2"><a href="sec-rmd.html#cb6-2" tabindex="-1"></a><span class="fu">render</span>(<span class="st">"report.Rmd"</span>, <span class="at">output_format =</span> <span class="fu">pdf_document</span>())</span>
<span id="cb6-3"><a href="sec-rmd.html#cb6-3" tabindex="-1"></a><span class="fu">render</span>(<span class="st">"report.Rmd"</span>, <span class="at">output_format =</span> <span class="fu">html_document</span>())</span></code></pre></div>
<p>Note that the <code>pdf_document()</code> and <code>html_document()</code> functions are
also defined in the <code>rmarkdown</code> package. Make sure to either load the
package or prefix it all functions.</p>
</div>
<div id="caching" class="section level2" number="1.2">
<h2>
<span class="header-section-number">1.2</span> Caching<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('caching')" onmouseout="reset_tooltip('caching-tooltip')"><span class="tooltiptext" id="caching-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Caching refers to the general technique of storing a result once its
computed so that next time, one the result is needed, it can be read
back instead of repeating the computation. This is useful when the
computation take much more time that reading the results from disk.</p>
<p>In Rmarkdown, this can be done by setting the <code>cache</code> option to <code>TRUE</code>
in the code chunk header:</p>
<pre><code>```{r long_computation, cache=TRUE}
res &lt;- long_computation()
```</code></pre>
<p>This can be easily be tested by simulating a long computation with the <code>Sys.sleep(time)</code>
function, where <code>time</code> is the time interval to suspend execution for, in seconds.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li><p>Create a new <code>Rmd</code> document with a single code chunk calling
<code>Sys.sleep(5)</code>. Compile it twice and compare the compilation times.</p></li>
<li><p>The <code>system.time()</code> function can be used to measure the time to run
a fonction in R. Instead of manually comparing the compilation
timings, wrap a call to <code>render()</code> in <code>system.time()</code> to let R
record the compilation timings. Hint: you can set <code>render(quiet =   TRUE)</code> to remove the knitting messages.</p></li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>Once you code chunk is cached, it will never be executed unless the
content of the code chunk is modified. In such a case, the code will
be executed again and the new results will be stored.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Modifiy the content of the cached code chunk and re-render the Rmd
file to confirm that the chunk is executed.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="writing-slides-in-rmarkdown" class="section level2" number="1.3">
<h2>
<span class="header-section-number">1.3</span> Writing slides in Rmarkdown<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('writing-slides-in-rmarkdown')" onmouseout="reset_tooltip('writing-slides-in-rmarkdown-tooltip')"><span class="tooltiptext" id="writing-slides-in-rmarkdown-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>So far, we have used Rmarkdown to produce documents, in pdf or in
html. Rmarkdown can be used to produce various types of documents
using different templates. In particular, when generating slides for a
presentation about data analysis, it might be convenient to use and
Rmd template to make sure that all figures and results presented are
genuine, and can readily be reproduced.</p>
<p>Producing other types of document is only a matter of installing the
appropriate template and defining the type of document in the Rmd
header. In RStudio, this is simply a matter of choosing a template
from the list of pre-installed ones when creation a new Rmd document:</p>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:unnamed-chunk-2"></span>
<p class="caption marginnote shownote">
Figure 1.1: Choosing and Rmd template.
</p>
<img src="figs/rmd_template.png" alt="Choosing and Rmd template." width="80%">
</div>
<p>For the slides, choose the
<a href="https://bookdown.org/yihui/rmarkdown/xaringan.html"><code>xaringan</code></a>
template called <em>Ninja Presentation</em>.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Create a new <code>xaringan</code> slide deck, compile it and go through the
slides to learn about its capabilities. You can also consult the
<a href="https://bookdown.org/yihui/rmarkdown/xaringan.html"><code>xaringan</code>
chapter</a> in the <em>R
Markdown: The Definitive Guide</em> book.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="citations" class="section level2" number="1.4">
<h2>
<span class="header-section-number">1.4</span> Citations<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('citations')" onmouseout="reset_tooltip('citations-tooltip')"><span class="tooltiptext" id="citations-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>One important aspect of writing scientific documents is to cite one’s
sources. The packages and papers that one reads and cites as part of a
scientific report, paper or thesis quickly becomes too large to manage
manually. It is therefore recommeded to use store references in a
unique file, ideally managed by a dedicated software.</p>
<p>In markdown, that file type is often a BibTex file (extension <code>.bib</code>)
that contain a set of entries such as these ones for R itself, the
<code>dplyr</code> package, or the <code>DESeq</code> paper (see chapter <a href="sec-rnaseq.html#sec-rnaseq">4</a>).</p>
<pre><code>@Manual{R-base,
  title = {R: A Language and Environment for Statistical
           Computing},
  author = {{R Core Team}},
  organization = {R Foundation for Statistical Computing},
  address = {Vienna, Austria},
  year = {2019},
  url = {https://www.R-project.org},
}

@Manual{dplyr,
    title = {dplyr: A Grammar of Data Manipulation},
    author = {Hadley Wickham and Romain François and Lionel {
             Henry} and Kirill Müller},
    year = {2020},
    note = {R package version 1.0.2},
    url = {https://CRAN.R-project.org/package=dplyr},
}

@Article{Love:2014,
  title = {Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2},
  author = {Michael I. Love and Wolfgang Huber and Simon Anders},
  year = {2014},
  journal = {Genome Biology},
  doi = {10.1186/s13059-014-0550-8},
  volume = {15},
  issue = {12},
  pages = {550},
}</code></pre>
<p>If the above entries are stored in the file <code>refs.bib</code>, then you’ll
need to define that file in the Rmd header as</p>
<pre><code>---
bibliography: references.bib
---</code></pre>
<p>and then cite each reference using the unique keys with <code>@R-base</code> or
<code>[@R-base]</code> to put the citation between parenthesis. To cite multiple
entries, separate these by semicolons,
e.g. <code>[@R-base;@dplyr;@Love:2014]</code>.</p>
<p>When compiled, the bibliography will be inserted at the end of the
document under the section <em>References</em>.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Create a simple markdown file and a BibTex file using the references
above and cite them.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>A useful function to get the BibTex entries of R packages or their
associated scientific papers is to use the <code>citation()</code> function:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="sec-rmd.html#cb10-1" tabindex="-1"></a><span class="fu">citation</span>()</span></code></pre></div>
<pre><code>## To cite R in publications use:
## 
##   R Core Team (2024). _R: A Language and Environment for Statistical
##   Computing_. R Foundation for Statistical Computing, Vienna, Austria.
##   &lt;https://www.R-project.org/&gt;.
## 
## A BibTeX entry for LaTeX users is
## 
##   @Manual{,
##     title = {R: A Language and Environment for Statistical Computing},
##     author = {{R Core Team}},
##     organization = {R Foundation for Statistical Computing},
##     address = {Vienna, Austria},
##     year = {2024},
##     url = {https://www.R-project.org/},
##   }
## 
## We have invested a lot of time and effort in creating R, please cite it
## when using it for data analysis. See also 'citation("pkgname")' for
## citing R packages.</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="sec-rmd.html#cb12-1" tabindex="-1"></a><span class="fu">citation</span>(<span class="at">package =</span> <span class="st">"SummarizedExperiment"</span>)</span></code></pre></div>
<pre><code>## To cite package 'SummarizedExperiment' in publications use:
## 
##   Morgan M, Obenchain V, Hester J, Pagès H (2024).
##   _SummarizedExperiment: SummarizedExperiment container_.
##   doi:10.18129/B9.bioc.SummarizedExperiment
##   &lt;https://doi.org/10.18129/B9.bioc.SummarizedExperiment&gt;, R package
##   version 1.34.0,
##   &lt;https://bioconductor.org/packages/SummarizedExperiment&gt;.
## 
## A BibTeX entry for LaTeX users is
## 
##   @Manual{,
##     title = {SummarizedExperiment: SummarizedExperiment container},
##     author = {Martin Morgan and Valerie Obenchain and Jim Hester and Hervé Pagès},
##     year = {2024},
##     note = {R package version 1.34.0},
##     url = {https://bioconductor.org/packages/SummarizedExperiment},
##     doi = {10.18129/B9.bioc.SummarizedExperiment},
##   }</code></pre>
<p>For more details about bibliographies and citations, see</p>
<ul>
<li><a href="https://bookdown.org/yihui/rmarkdown-cookbook/bibliography.html" class="uri">https://bookdown.org/yihui/rmarkdown-cookbook/bibliography.html</a></li>
<li><a href="https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#Bibliographies" class="uri">https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#Bibliographies</a></li>
</ul>
</div>
<div id="references" class="section level2" number="1.5">
<h2>
<span class="header-section-number">1.5</span> References<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('references')" onmouseout="reset_tooltip('references-tooltip')"><span class="tooltiptext" id="references-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<ul>
<li><p><a href="https://bookdown.org/yihui/rmarkdown">R Markdown: The Definitive
Guide</a>, Yihui Xie,
J. J. Allaire, Garrett Grolemund.</p></li>
<li><p><a href="https://bookdown.org/yihui/rmarkdown-cookbook/">R Markdown
Cookbook</a>, Yihui
Xie, Christophe Dervieux, Emily Riederer.</p></li>
</ul>
</div>
</div>
<p style="text-align: center;">
<a href="index.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-linmod.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2024-11-01
 using 
R version 4.4.1 (2024-06-14)
</p>
</div>
</div>



</body>
</html>
